<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.550">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>INVESTIGATING BATCH EFFECT AND CELL FATE DYNAMICS IN THE DROSOPHILA MELANOGASTER INTESTINE USING SINGLE-CELL RNA VELOCITY</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="main_analyses_files/libs/clipboard/clipboard.min.js"></script>
<script src="main_analyses_files/libs/quarto-html/quarto.js"></script>
<script src="main_analyses_files/libs/quarto-html/popper.min.js"></script>
<script src="main_analyses_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="main_analyses_files/libs/quarto-html/anchor.min.js"></script>
<link href="main_analyses_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="main_analyses_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="main_analyses_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="main_analyses_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="main_analyses_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">


</head>

<body>

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">
<div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
  <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#the-quality-of-single-cell-rna-sequencing-data-of-the-drosophila-intestine" id="toc-the-quality-of-single-cell-rna-sequencing-data-of-the-drosophila-intestine" class="nav-link active" data-scroll-target="#the-quality-of-single-cell-rna-sequencing-data-of-the-drosophila-intestine">3.2 The quality of single-cell RNA-sequencing data of the Drosophila intestine</a></li>
  <li><a href="#exploring-data-structure" id="toc-exploring-data-structure" class="nav-link" data-scroll-target="#exploring-data-structure">3.3 Exploring data structure</a></li>
  <li><a href="#rna-velocity-quality-control" id="toc-rna-velocity-quality-control" class="nav-link" data-scroll-target="#rna-velocity-quality-control">3.4 RNA velocity quality control</a></li>
  <li><a href="#rna-velocity-estimation-approaches-affect-phase-portraits" id="toc-rna-velocity-estimation-approaches-affect-phase-portraits" class="nav-link" data-scroll-target="#rna-velocity-estimation-approaches-affect-phase-portraits">3.6 RNA velocity estimation approaches affect phase portraits</a></li>
  <li><a href="#rna-velocity-recapitulates-trajectories-with-certain-cell-types" id="toc-rna-velocity-recapitulates-trajectories-with-certain-cell-types" class="nav-link" data-scroll-target="#rna-velocity-recapitulates-trajectories-with-certain-cell-types">3.7 RNA velocity recapitulates trajectories with certain cell types</a></li>
  <li><a href="#abundance-in-cell-types-is-different-in-conditions" id="toc-abundance-in-cell-types-is-different-in-conditions" class="nav-link" data-scroll-target="#abundance-in-cell-types-is-different-in-conditions">3.8 Abundance in cell types is different in conditions</a></li>
  </ul>
</nav>
</div>
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">INVESTIGATING BATCH EFFECT AND CELL FATE DYNAMICS IN THE DROSOPHILA MELANOGASTER INTESTINE USING SINGLE-CELL RNA VELOCITY</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<p>Code to reproduce the figures from master thesis Author: Sviatoslav Kharuk</p>
<section id="the-quality-of-single-cell-rna-sequencing-data-of-the-drosophila-intestine" class="level1">
<h1>3.2 The quality of single-cell RNA-sequencing data of the Drosophila intestine</h1>
<div id="cell-3" class="cell" data-execution_count="1">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co"># scseq env on cluster </span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="co"># import libraries</span></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> scanpy <span class="im">as</span> sc</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> numpy <span class="im">as</span> np</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> matplotlib.pyplot <span class="im">as</span> plt</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> scvelo <span class="im">as</span> scv</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> re</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> seaborn <span class="im">as</span> sns</span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> matplotlib</span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> scanpy.external <span class="im">as</span> sce</span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> anndata <span class="im">as</span> ad</span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> scipy</span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-15"><a href="#cb1-15" aria-hidden="true" tabindex="-1"></a>sc.settings.verbosity <span class="op">=</span> <span class="dv">3</span> <span class="co"># verbosity: errors (0), warnings (1), info (2), hints (3)  ##! better to set 3</span></span>
<span id="cb1-16"><a href="#cb1-16" aria-hidden="true" tabindex="-1"></a><span class="co"># verbosity is the amount of information that is printed to the screen. </span></span>
<span id="cb1-17"><a href="#cb1-17" aria-hidden="true" tabindex="-1"></a>sc.logging.print_header() <span class="co"># print header info</span></span>
<span id="cb1-18"><a href="#cb1-18" aria-hidden="true" tabindex="-1"></a>sc.settings.set_figure_params(dpi<span class="op">=</span><span class="dv">140</span>, transparent<span class="op">=</span><span class="va">True</span>, frameon<span class="op">=</span><span class="va">False</span>) <span class="co"># set dpi and background color for figures</span></span>
<span id="cb1-19"><a href="#cb1-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-20"><a href="#cb1-20" aria-hidden="true" tabindex="-1"></a>scv.logging.print_version() <span class="co"># show version  </span></span>
<span id="cb1-21"><a href="#cb1-21" aria-hidden="true" tabindex="-1"></a>scv.settings.presenter_view <span class="op">=</span> <span class="va">True</span>  <span class="co"># set max width size for presenter view</span></span>
<span id="cb1-22"><a href="#cb1-22" aria-hidden="true" tabindex="-1"></a>scv.set_figure_params(<span class="st">'scvelo'</span>)</span>
<span id="cb1-23"><a href="#cb1-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-24"><a href="#cb1-24" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> warnings</span>
<span id="cb1-25"><a href="#cb1-25" aria-hidden="true" tabindex="-1"></a>warnings.filterwarnings(<span class="st">"ignore"</span>, category<span class="op">=</span><span class="pp">DeprecationWarning</span>)</span>
<span id="cb1-26"><a href="#cb1-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-27"><a href="#cb1-27" aria-hidden="true" tabindex="-1"></a><span class="co"># try to get rid of the warnings</span></span>
<span id="cb1-28"><a href="#cb1-28" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> warnings</span>
<span id="cb1-29"><a href="#cb1-29" aria-hidden="true" tabindex="-1"></a>warnings.filterwarnings(<span class="st">"ignore"</span>, category<span class="op">=</span><span class="pp">DeprecationWarning</span>, module<span class="op">=</span><span class="st">"scvelo.plotting.utils"</span>)</span>
<span id="cb1-30"><a href="#cb1-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-31"><a href="#cb1-31" aria-hidden="true" tabindex="-1"></a><span class="co"># import pertpy as pt</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>scanpy==1.9.6 anndata==0.10.8 umap==0.5.5 numpy==1.26.2 scipy==1.11.4 pandas==2.2.2 scikit-learn==1.1.3 statsmodels==0.14.1 igraph==0.11.6 pynndescent==0.5.11
Running scvelo 0.3.1 (python 3.10.14) on 2024-12-01 14:46.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>ERROR: XMLRPC request failed [code: -32500]
RuntimeError: PyPI no longer supports 'pip search' (or XML-RPC search). Please use https://pypi.org/search (via a browser) instead. See https://warehouse.pypa.io/api-reference/xml-rpc.html#deprecated-methods for more information.</code></pre>
</div>
</div>
<div id="cell-4" class="cell" data-execution_count="2">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> statannotations.Annotator <span class="im">import</span> Annotator</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-error">
<pre><code>ModuleNotFoundError: No module named 'statannotations'</code></pre>
</div>
</div>
<div id="cell-5" class="cell" data-execution_count="2">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> functions <span class="im">import</span> <span class="op">*</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>OMP: Info #276: omp_set_nested routine deprecated, please use omp_set_max_active_levels instead.</code></pre>
</div>
</div>
<div id="cell-6" class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>adata <span class="op">=</span> sc.read_h5ad(<span class="st">"/g/huber/users/peidli/data/DECODE/version_1/SP_merged_filtered_feature_bc_matrix.h5ad"</span>)   </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="cell-7" class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="co">### RELABELING, based on instructions from Siamak </span><span class="al">###</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>adata.obs.perturbation <span class="op">=</span> adata.obs.perturbation.astype(<span class="bu">str</span>)</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>adata.obs.perturbation <span class="op">=</span> adata.obs.perturbation.replace({<span class="st">"fkh 00845"</span>: <span class="st">"GataE"</span>})</span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>adata <span class="op">=</span> adata[<span class="op">~</span>adata.obs.perturbation.isin([<span class="st">"Rab35.W"</span>, <span class="st">"Rab35.DN"</span>, <span class="st">"Rab35.CA"</span>, <span class="st">"FBNR control"</span>, <span class="st">"Arm.CA"</span>, <span class="st">"dcr, APCi"</span>, <span class="st">"Rab35i"</span>])].copy()</span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>adata <span class="op">=</span> adata[<span class="op">~</span>adata.obs.sample_name.isin([<span class="st">"TX32_1"</span>, <span class="st">"TX34_A"</span>])].copy()</span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a>adata.obs[<span class="st">"perturbation"</span>][adata.obs.sample_name.isin([<span class="st">"TX29_1"</span>, <span class="st">"TX30_1"</span>, <span class="st">"TX31_1"</span>, <span class="st">"TX33_1"</span>])] <span class="op">=</span> <span class="st">"control"</span></span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a>adata.obs[<span class="st">"perturbation"</span>][adata.obs.sample_name.isin([<span class="st">"TX39_6"</span>])] <span class="op">=</span> <span class="st">"Crp"</span></span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a>adata.obs.perturbation.replace({</span>
<span id="cb9-9"><a href="#cb9-9" aria-hidden="true" tabindex="-1"></a>            <span class="st">"control (None)"</span>: <span class="st">"control"</span>,</span>
<span id="cb9-10"><a href="#cb9-10" aria-hidden="true" tabindex="-1"></a>            <span class="st">"control (esgts, cas9/+)"</span>: <span class="st">"control"</span>,</span>
<span id="cb9-11"><a href="#cb9-11" aria-hidden="true" tabindex="-1"></a>        }, inplace<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb9-12"><a href="#cb9-12" aria-hidden="true" tabindex="-1"></a>adata.obs.perturbation.replace({</span>
<span id="cb9-13"><a href="#cb9-13" aria-hidden="true" tabindex="-1"></a>            <span class="st">"Dome"</span>: <span class="st">"dome"</span>,</span>
<span id="cb9-14"><a href="#cb9-14" aria-hidden="true" tabindex="-1"></a>            <span class="st">"stat92E 00701"</span>: <span class="st">"Stat92E"</span>,</span>
<span id="cb9-15"><a href="#cb9-15" aria-hidden="true" tabindex="-1"></a>            <span class="st">"APC 1_2 gRNA"</span>: <span class="st">"APC1+2 gRNA"</span>,</span>
<span id="cb9-16"><a href="#cb9-16" aria-hidden="true" tabindex="-1"></a>            <span class="st">"Notch 1184 gRNA"</span>: <span class="st">"Notch gRNA"</span>,</span>
<span id="cb9-17"><a href="#cb9-17" aria-hidden="true" tabindex="-1"></a>            <span class="st">"USP 00341"</span>: <span class="st">"usp"</span>,</span>
<span id="cb9-18"><a href="#cb9-18" aria-hidden="true" tabindex="-1"></a>            <span class="st">"Rab11 (02478)"</span>: <span class="st">"Rab11"</span>,</span>
<span id="cb9-19"><a href="#cb9-19" aria-hidden="true" tabindex="-1"></a>            <span class="st">"rab11"</span>: <span class="st">"Rab11"</span>,</span>
<span id="cb9-20"><a href="#cb9-20" aria-hidden="true" tabindex="-1"></a>            <span class="st">"xbp1"</span>: <span class="st">"Xbp1"</span>,</span>
<span id="cb9-21"><a href="#cb9-21" aria-hidden="true" tabindex="-1"></a>        }, inplace<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="cell-8" class="cell" data-execution_count="4">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="co"># subset the data to 3' Notch + control</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>adata_notch_control_3p <span class="op">=</span> adata[((adata.obs[<span class="st">"perturbation"</span>] <span class="op">==</span> <span class="st">"control"</span>) <span class="op">&amp;</span></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>                                ((adata.obs[<span class="st">"sample_name"</span>] <span class="op">==</span> <span class="st">"TX22_A"</span>) <span class="op">|</span></span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>                                (adata.obs[<span class="st">"sample_name"</span>] <span class="op">==</span> <span class="st">"TX23_A"</span>))) <span class="op">|</span></span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>                                ((adata.obs[<span class="st">"perturbation"</span>] <span class="op">==</span> <span class="st">"Notch gRNA"</span>) <span class="op">&amp;</span></span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a>                                ((adata.obs[<span class="st">"sample_name"</span>] <span class="op">==</span> <span class="st">"TX22_D"</span>) <span class="op">|</span></span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a>                                 (adata.obs[<span class="st">"sample_name"</span>] <span class="op">==</span> <span class="st">"TX23_D"</span>)))].copy()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="cell-9" class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="co"># calcuate the QC metrics</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a><span class="co"># mito genes</span></span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>adata_notch_control_3p.var[<span class="st">"mt"</span>] <span class="op">=</span> adata_notch_control_3p.var_names.<span class="bu">str</span>.startswith(<span class="st">"mt:"</span>) <span class="co"># # create new columns and store the information whether the gene is mitochondrial or not (True or False).</span></span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a><span class="co"># calculate the metrics</span></span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a>sc.pp.calculate_qc_metrics(adata_notch_control_3p, qc_vars<span class="op">=</span>[<span class="st">"mt"</span>], inplace <span class="op">=</span> <span class="va">True</span>, percent_top <span class="op">=</span> <span class="va">None</span>, log1p<span class="op">=</span><span class="va">False</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="cell-10" class="cell" data-execution_count="7">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> matplotlib.lines <span class="im">import</span> Line2D</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>palette<span class="op">=</span>{<span class="st">"Notch gRNA"</span>: <span class="st">"#1f77b4"</span>, <span class="st">"control"</span>: <span class="st">"#d62728"</span>}</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a>plt.figure(figsize<span class="op">=</span>(<span class="fl">2.5</span>,<span class="dv">5</span>))</span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> sns.violinplot(data<span class="op">=</span>adata_notch_control_3p.obs, y<span class="op">=</span><span class="st">"pct_counts_mt"</span>, x<span class="op">=</span><span class="st">"perturbation"</span>, palette<span class="op">=</span>palette)</span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a>plt.tick_params(axis<span class="op">=</span><span class="st">'x'</span>, which<span class="op">=</span><span class="st">'both'</span>, bottom<span class="op">=</span><span class="va">False</span>, top<span class="op">=</span><span class="va">False</span>, labelbottom<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a>plt.ylabel(<span class="st">"Percentage of UMI counts stemming </span><span class="ch">\n</span><span class="st"> from mitochondrial genes per droplet"</span>)</span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true" tabindex="-1"></a>ax.<span class="bu">set</span>(xlabel<span class="op">=</span><span class="va">None</span>)</span>
<span id="cb12-9"><a href="#cb12-9" aria-hidden="true" tabindex="-1"></a>plt.axhline(y<span class="op">=</span><span class="dv">30</span>, color<span class="op">=</span><span class="st">'red'</span>, linestyle<span class="op">=</span><span class="st">'--'</span>, linewidth<span class="op">=</span><span class="dv">1</span>, zorder<span class="op">=</span><span class="dv">10</span>)</span>
<span id="cb12-10"><a href="#cb12-10" aria-hidden="true" tabindex="-1"></a>sns.despine()</span>
<span id="cb12-11"><a href="#cb12-11" aria-hidden="true" tabindex="-1"></a>legend_elements <span class="op">=</span> [</span>
<span id="cb12-12"><a href="#cb12-12" aria-hidden="true" tabindex="-1"></a>    Line2D([<span class="dv">0</span>], [<span class="dv">0</span>], color<span class="op">=</span>palette[<span class="st">"Notch gRNA"</span>], lw<span class="op">=</span><span class="dv">4</span>, label<span class="op">=</span><span class="st">"$\it{Notch\ KO}$"</span>),</span>
<span id="cb12-13"><a href="#cb12-13" aria-hidden="true" tabindex="-1"></a>    Line2D([<span class="dv">0</span>], [<span class="dv">0</span>], color<span class="op">=</span>palette[<span class="st">"control"</span>], lw<span class="op">=</span><span class="dv">4</span>, label<span class="op">=</span><span class="st">"control"</span>)]</span>
<span id="cb12-14"><a href="#cb12-14" aria-hidden="true" tabindex="-1"></a>plt.legend(handles<span class="op">=</span>legend_elements, title<span class="op">=</span><span class="st">"Condition"</span>, loc<span class="op">=</span><span class="st">"upper left"</span>, bbox_to_anchor<span class="op">=</span>(<span class="dv">1</span>, <span class="dv">1</span>), frameon<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb12-15"><a href="#cb12-15" aria-hidden="true" tabindex="-1"></a>plt.show()</span>
<span id="cb12-16"><a href="#cb12-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-17"><a href="#cb12-17" aria-hidden="true" tabindex="-1"></a>plt.figure(figsize<span class="op">=</span>(<span class="fl">2.5</span>,<span class="dv">5</span>))</span>
<span id="cb12-18"><a href="#cb12-18" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> sns.violinplot(data<span class="op">=</span>adata_notch_control_3p.obs, y<span class="op">=</span><span class="st">"total_counts"</span>, x<span class="op">=</span><span class="st">"perturbation"</span>, palette<span class="op">=</span>palette)</span>
<span id="cb12-19"><a href="#cb12-19" aria-hidden="true" tabindex="-1"></a>plt.tick_params(axis<span class="op">=</span><span class="st">'x'</span>, which<span class="op">=</span><span class="st">'both'</span>, bottom<span class="op">=</span><span class="va">False</span>, top<span class="op">=</span><span class="va">False</span>, labelbottom<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb12-20"><a href="#cb12-20" aria-hidden="true" tabindex="-1"></a>plt.ylabel(<span class="st">"Number of total UMI counts per droplet"</span>)</span>
<span id="cb12-21"><a href="#cb12-21" aria-hidden="true" tabindex="-1"></a>ax.<span class="bu">set</span>(xlabel<span class="op">=</span><span class="va">None</span>)</span>
<span id="cb12-22"><a href="#cb12-22" aria-hidden="true" tabindex="-1"></a>plt.axhline(y<span class="op">=</span><span class="dv">500</span>, color<span class="op">=</span><span class="st">'red'</span>, linestyle<span class="op">=</span><span class="st">'--'</span>, linewidth<span class="op">=</span><span class="dv">1</span>, zorder<span class="op">=</span><span class="dv">10</span>)</span>
<span id="cb12-23"><a href="#cb12-23" aria-hidden="true" tabindex="-1"></a>sns.despine()</span>
<span id="cb12-24"><a href="#cb12-24" aria-hidden="true" tabindex="-1"></a>legend_elements <span class="op">=</span> [</span>
<span id="cb12-25"><a href="#cb12-25" aria-hidden="true" tabindex="-1"></a>    Line2D([<span class="dv">0</span>], [<span class="dv">0</span>], color<span class="op">=</span>palette[<span class="st">"Notch gRNA"</span>], lw<span class="op">=</span><span class="dv">4</span>, label<span class="op">=</span><span class="st">"$\it{Notch\ KO}$"</span>),</span>
<span id="cb12-26"><a href="#cb12-26" aria-hidden="true" tabindex="-1"></a>    Line2D([<span class="dv">0</span>], [<span class="dv">0</span>], color<span class="op">=</span>palette[<span class="st">"control"</span>], lw<span class="op">=</span><span class="dv">4</span>, label<span class="op">=</span><span class="st">"control"</span>)]</span>
<span id="cb12-27"><a href="#cb12-27" aria-hidden="true" tabindex="-1"></a>plt.legend(handles<span class="op">=</span>legend_elements, title<span class="op">=</span><span class="st">"Condition"</span>, loc<span class="op">=</span><span class="st">"upper left"</span>, bbox_to_anchor<span class="op">=</span>(<span class="dv">1</span>, <span class="dv">1</span>), frameon<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb12-28"><a href="#cb12-28" aria-hidden="true" tabindex="-1"></a>plt.show()</span>
<span id="cb12-29"><a href="#cb12-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-30"><a href="#cb12-30" aria-hidden="true" tabindex="-1"></a>plt.figure(figsize<span class="op">=</span>(<span class="fl">2.5</span>,<span class="dv">5</span>))</span>
<span id="cb12-31"><a href="#cb12-31" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> sns.violinplot(data<span class="op">=</span>adata_notch_control_3p.obs, y<span class="op">=</span><span class="st">"n_genes_by_counts"</span>, x<span class="op">=</span><span class="st">"perturbation"</span>, palette<span class="op">=</span>palette)</span>
<span id="cb12-32"><a href="#cb12-32" aria-hidden="true" tabindex="-1"></a>plt.tick_params(axis<span class="op">=</span><span class="st">'x'</span>, which<span class="op">=</span><span class="st">'both'</span>, bottom<span class="op">=</span><span class="va">False</span>, top<span class="op">=</span><span class="va">False</span>, labelbottom<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb12-33"><a href="#cb12-33" aria-hidden="true" tabindex="-1"></a>plt.ylabel(<span class="st">"Number of genes per droplet"</span>)</span>
<span id="cb12-34"><a href="#cb12-34" aria-hidden="true" tabindex="-1"></a>ax.<span class="bu">set</span>(xlabel<span class="op">=</span><span class="va">None</span>)</span>
<span id="cb12-35"><a href="#cb12-35" aria-hidden="true" tabindex="-1"></a>plt.axhline(y<span class="op">=</span><span class="dv">250</span>, color<span class="op">=</span><span class="st">'red'</span>, linestyle<span class="op">=</span><span class="st">'--'</span>, linewidth<span class="op">=</span><span class="dv">1</span>, zorder<span class="op">=</span><span class="dv">10</span>)</span>
<span id="cb12-36"><a href="#cb12-36" aria-hidden="true" tabindex="-1"></a>sns.despine()</span>
<span id="cb12-37"><a href="#cb12-37" aria-hidden="true" tabindex="-1"></a>legend_elements <span class="op">=</span> [</span>
<span id="cb12-38"><a href="#cb12-38" aria-hidden="true" tabindex="-1"></a>    Line2D([<span class="dv">0</span>], [<span class="dv">0</span>], color<span class="op">=</span>palette[<span class="st">"Notch gRNA"</span>], lw<span class="op">=</span><span class="dv">4</span>, label<span class="op">=</span><span class="st">"$\it{Notch\ KO}$"</span>),</span>
<span id="cb12-39"><a href="#cb12-39" aria-hidden="true" tabindex="-1"></a>    Line2D([<span class="dv">0</span>], [<span class="dv">0</span>], color<span class="op">=</span>palette[<span class="st">"control"</span>], lw<span class="op">=</span><span class="dv">4</span>, label<span class="op">=</span><span class="st">"control"</span>)]</span>
<span id="cb12-40"><a href="#cb12-40" aria-hidden="true" tabindex="-1"></a>plt.legend(handles<span class="op">=</span>legend_elements, title<span class="op">=</span><span class="st">"Condition"</span>, loc<span class="op">=</span><span class="st">"upper left"</span>, bbox_to_anchor<span class="op">=</span>(<span class="dv">1</span>, <span class="dv">1</span>), frameon<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb12-41"><a href="#cb12-41" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="main_analyses_files/figure-html/cell-9-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="main_analyses_files/figure-html/cell-9-output-2.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="main_analyses_files/figure-html/cell-9-output-3.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<div id="cell-11" class="cell" data-execution_count="8">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot scatter plot for pct_counts_mt versus n_genes_by_counts</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>fig, ax <span class="op">=</span> plt.subplots(figsize<span class="op">=</span>(<span class="dv">7</span>, <span class="dv">5</span>))  <span class="co"># Adjust figsize as needed</span></span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>sc.pl.scatter(adata_notch_control_3p, x<span class="op">=</span><span class="st">"total_counts"</span>, y<span class="op">=</span><span class="st">"n_genes_by_counts"</span>, color<span class="op">=</span><span class="st">"pct_counts_mt"</span>, ax<span class="op">=</span>ax, show<span class="op">=</span><span class="va">False</span>, frameon<span class="op">=</span><span class="va">False</span>, color_map<span class="op">=</span><span class="st">"RdBu_r"</span>)</span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Add title and customize plot appearance</span></span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a>ax.set_xlabel(<span class="st">"Total counts"</span>)</span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a>ax.set_ylabel(<span class="st">"Number of genes"</span>)</span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a>ax.spines[<span class="st">'right'</span>].set_visible(<span class="va">False</span>)  <span class="co"># Remove right spine</span></span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a>ax.spines[<span class="st">'top'</span>].set_visible(<span class="va">False</span>)    <span class="co"># Remove top spine</span></span>
<span id="cb13-9"><a href="#cb13-9" aria-hidden="true" tabindex="-1"></a>fig.tight_layout(rect<span class="op">=</span>[<span class="dv">0</span>, <span class="dv">0</span>, <span class="fl">0.85</span>, <span class="dv">1</span>])  <span class="co"># Adjust the rectangle to move the legend</span></span>
<span id="cb13-10"><a href="#cb13-10" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="main_analyses_files/figure-html/cell-10-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<div id="cell-12" class="cell" data-execution_count="9">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="co"># filtering the cells</span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>sc.pp.filter_cells(adata_notch_control_3p, min_counts<span class="op">=</span><span class="dv">500</span>)</span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>sc.pp.filter_cells(adata_notch_control_3p, min_genes<span class="op">=</span><span class="dv">250</span>)</span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a>adata_notch_control_3p <span class="op">=</span> adata_notch_control_3p[(adata_notch_control_3p.obs.pct_counts_mt<span class="op">&gt;</span><span class="fl">0.5</span>)<span class="op">&amp;</span>(adata_notch_control_3p.obs.pct_counts_mt<span class="op">&lt;</span><span class="dv">30</span>)].copy()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>filtered out 1954 cells that have less than 250 genes expressed</code></pre>
</div>
</div>
<div id="cell-13" class="cell" data-execution_count="10">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="co"># plot QC after filtering</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>plt.figure(figsize<span class="op">=</span>(<span class="fl">2.5</span>,<span class="dv">5</span>))</span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> sns.violinplot(data<span class="op">=</span>adata_notch_control_3p.obs, y<span class="op">=</span><span class="st">"pct_counts_mt"</span>, x<span class="op">=</span><span class="st">"perturbation"</span>, palette<span class="op">=</span>palette)</span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a>plt.tick_params(axis<span class="op">=</span><span class="st">'x'</span>, which<span class="op">=</span><span class="st">'both'</span>, bottom<span class="op">=</span><span class="va">False</span>, top<span class="op">=</span><span class="va">False</span>, labelbottom<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a>plt.ylabel(<span class="st">"Percentage of UMI counts stemming </span><span class="ch">\n</span><span class="st"> from mitochondrial genes per droplet"</span>)</span>
<span id="cb16-7"><a href="#cb16-7" aria-hidden="true" tabindex="-1"></a>ax.<span class="bu">set</span>(xlabel<span class="op">=</span><span class="va">None</span>)</span>
<span id="cb16-8"><a href="#cb16-8" aria-hidden="true" tabindex="-1"></a><span class="co"># plt.axhline(y=30, color='red', linestyle='--', linewidth=1, zorder=10)</span></span>
<span id="cb16-9"><a href="#cb16-9" aria-hidden="true" tabindex="-1"></a>sns.despine()</span>
<span id="cb16-10"><a href="#cb16-10" aria-hidden="true" tabindex="-1"></a>legend_elements <span class="op">=</span> [</span>
<span id="cb16-11"><a href="#cb16-11" aria-hidden="true" tabindex="-1"></a>    Line2D([<span class="dv">0</span>], [<span class="dv">0</span>], color<span class="op">=</span>palette[<span class="st">"Notch gRNA"</span>], lw<span class="op">=</span><span class="dv">4</span>, label<span class="op">=</span><span class="st">"$\it{Notch\ KO}$"</span>),</span>
<span id="cb16-12"><a href="#cb16-12" aria-hidden="true" tabindex="-1"></a>    Line2D([<span class="dv">0</span>], [<span class="dv">0</span>], color<span class="op">=</span>palette[<span class="st">"control"</span>], lw<span class="op">=</span><span class="dv">4</span>, label<span class="op">=</span><span class="st">"control"</span>)]</span>
<span id="cb16-13"><a href="#cb16-13" aria-hidden="true" tabindex="-1"></a>plt.legend(handles<span class="op">=</span>legend_elements, title<span class="op">=</span><span class="st">"Condition"</span>, loc<span class="op">=</span><span class="st">"upper left"</span>, bbox_to_anchor<span class="op">=</span>(<span class="dv">1</span>, <span class="dv">1</span>), frameon<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb16-14"><a href="#cb16-14" aria-hidden="true" tabindex="-1"></a>plt.show()</span>
<span id="cb16-15"><a href="#cb16-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-16"><a href="#cb16-16" aria-hidden="true" tabindex="-1"></a>plt.figure(figsize<span class="op">=</span>(<span class="fl">2.5</span>,<span class="dv">5</span>))</span>
<span id="cb16-17"><a href="#cb16-17" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> sns.violinplot(data<span class="op">=</span>adata_notch_control_3p.obs, y<span class="op">=</span><span class="st">"total_counts"</span>, x<span class="op">=</span><span class="st">"perturbation"</span>, palette<span class="op">=</span>palette)</span>
<span id="cb16-18"><a href="#cb16-18" aria-hidden="true" tabindex="-1"></a>plt.tick_params(axis<span class="op">=</span><span class="st">'x'</span>, which<span class="op">=</span><span class="st">'both'</span>, bottom<span class="op">=</span><span class="va">False</span>, top<span class="op">=</span><span class="va">False</span>, labelbottom<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb16-19"><a href="#cb16-19" aria-hidden="true" tabindex="-1"></a>plt.ylabel(<span class="st">"Number of total UMI counts per droplet"</span>)</span>
<span id="cb16-20"><a href="#cb16-20" aria-hidden="true" tabindex="-1"></a>ax.<span class="bu">set</span>(xlabel<span class="op">=</span><span class="va">None</span>)</span>
<span id="cb16-21"><a href="#cb16-21" aria-hidden="true" tabindex="-1"></a><span class="co"># plt.axhline(y=30, color='red', linestyle='--', linewidth=1, zorder=10)</span></span>
<span id="cb16-22"><a href="#cb16-22" aria-hidden="true" tabindex="-1"></a>sns.despine()</span>
<span id="cb16-23"><a href="#cb16-23" aria-hidden="true" tabindex="-1"></a>legend_elements <span class="op">=</span> [</span>
<span id="cb16-24"><a href="#cb16-24" aria-hidden="true" tabindex="-1"></a>    Line2D([<span class="dv">0</span>], [<span class="dv">0</span>], color<span class="op">=</span>palette[<span class="st">"Notch gRNA"</span>], lw<span class="op">=</span><span class="dv">4</span>, label<span class="op">=</span><span class="st">"$\it{Notch\ KO}$"</span>),</span>
<span id="cb16-25"><a href="#cb16-25" aria-hidden="true" tabindex="-1"></a>    Line2D([<span class="dv">0</span>], [<span class="dv">0</span>], color<span class="op">=</span>palette[<span class="st">"control"</span>], lw<span class="op">=</span><span class="dv">4</span>, label<span class="op">=</span><span class="st">"control"</span>)]</span>
<span id="cb16-26"><a href="#cb16-26" aria-hidden="true" tabindex="-1"></a>plt.legend(handles<span class="op">=</span>legend_elements, title<span class="op">=</span><span class="st">"Condition"</span>, loc<span class="op">=</span><span class="st">"upper left"</span>, bbox_to_anchor<span class="op">=</span>(<span class="dv">1</span>, <span class="dv">1</span>), frameon<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb16-27"><a href="#cb16-27" aria-hidden="true" tabindex="-1"></a>plt.show()</span>
<span id="cb16-28"><a href="#cb16-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-29"><a href="#cb16-29" aria-hidden="true" tabindex="-1"></a>plt.figure(figsize<span class="op">=</span>(<span class="fl">2.5</span>,<span class="dv">5</span>))</span>
<span id="cb16-30"><a href="#cb16-30" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> sns.violinplot(data<span class="op">=</span>adata_notch_control_3p.obs, y<span class="op">=</span><span class="st">"n_genes_by_counts"</span>, x<span class="op">=</span><span class="st">"perturbation"</span>, palette<span class="op">=</span>palette)</span>
<span id="cb16-31"><a href="#cb16-31" aria-hidden="true" tabindex="-1"></a>plt.tick_params(axis<span class="op">=</span><span class="st">'x'</span>, which<span class="op">=</span><span class="st">'both'</span>, bottom<span class="op">=</span><span class="va">False</span>, top<span class="op">=</span><span class="va">False</span>, labelbottom<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb16-32"><a href="#cb16-32" aria-hidden="true" tabindex="-1"></a>plt.ylabel(<span class="st">"Number of genes per droplet"</span>)</span>
<span id="cb16-33"><a href="#cb16-33" aria-hidden="true" tabindex="-1"></a>ax.<span class="bu">set</span>(xlabel<span class="op">=</span><span class="va">None</span>)</span>
<span id="cb16-34"><a href="#cb16-34" aria-hidden="true" tabindex="-1"></a><span class="co"># plt.axhline(y=30, color='red', linestyle='--', linewidth=1, zorder=10)</span></span>
<span id="cb16-35"><a href="#cb16-35" aria-hidden="true" tabindex="-1"></a>sns.despine()</span>
<span id="cb16-36"><a href="#cb16-36" aria-hidden="true" tabindex="-1"></a>legend_elements <span class="op">=</span> [</span>
<span id="cb16-37"><a href="#cb16-37" aria-hidden="true" tabindex="-1"></a>    Line2D([<span class="dv">0</span>], [<span class="dv">0</span>], color<span class="op">=</span>palette[<span class="st">"Notch gRNA"</span>], lw<span class="op">=</span><span class="dv">4</span>, label<span class="op">=</span><span class="st">"$\it{Notch\ KO}$"</span>),</span>
<span id="cb16-38"><a href="#cb16-38" aria-hidden="true" tabindex="-1"></a>    Line2D([<span class="dv">0</span>], [<span class="dv">0</span>], color<span class="op">=</span>palette[<span class="st">"control"</span>], lw<span class="op">=</span><span class="dv">4</span>, label<span class="op">=</span><span class="st">"control"</span>)]</span>
<span id="cb16-39"><a href="#cb16-39" aria-hidden="true" tabindex="-1"></a>plt.legend(handles<span class="op">=</span>legend_elements, title<span class="op">=</span><span class="st">"Condition"</span>, loc<span class="op">=</span><span class="st">"upper left"</span>, bbox_to_anchor<span class="op">=</span>(<span class="dv">1</span>, <span class="dv">1</span>), frameon<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb16-40"><a href="#cb16-40" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="main_analyses_files/figure-html/cell-12-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="main_analyses_files/figure-html/cell-12-output-2.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="main_analyses_files/figure-html/cell-12-output-3.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<div id="cell-14" class="cell" data-execution_count="11">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot scatter plot for pct_counts_mt versus n_genes_by_counts</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>fig, ax <span class="op">=</span> plt.subplots(figsize<span class="op">=</span>(<span class="dv">7</span>, <span class="dv">5</span>))  <span class="co"># Adjust figsize as needed</span></span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a>sc.pl.scatter(adata_notch_control_3p, x<span class="op">=</span><span class="st">"total_counts"</span>, y<span class="op">=</span><span class="st">"n_genes_by_counts"</span>, color<span class="op">=</span><span class="st">"pct_counts_mt"</span>, ax<span class="op">=</span>ax, show<span class="op">=</span><span class="va">False</span>, frameon<span class="op">=</span><span class="va">False</span>, color_map<span class="op">=</span><span class="st">"RdBu_r"</span>)</span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Add title and customize plot appearance</span></span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a>ax.set_xlabel(<span class="st">"Total counts"</span>)</span>
<span id="cb17-7"><a href="#cb17-7" aria-hidden="true" tabindex="-1"></a>ax.set_ylabel(<span class="st">"Number of genes"</span>)</span>
<span id="cb17-8"><a href="#cb17-8" aria-hidden="true" tabindex="-1"></a>ax.spines[<span class="st">'right'</span>].set_visible(<span class="va">False</span>)  <span class="co"># Remove right spine</span></span>
<span id="cb17-9"><a href="#cb17-9" aria-hidden="true" tabindex="-1"></a>ax.spines[<span class="st">'top'</span>].set_visible(<span class="va">False</span>)    <span class="co"># Remove top spine</span></span>
<span id="cb17-10"><a href="#cb17-10" aria-hidden="true" tabindex="-1"></a>fig.tight_layout(rect<span class="op">=</span>[<span class="dv">0</span>, <span class="dv">0</span>, <span class="fl">0.85</span>, <span class="dv">1</span>])  <span class="co"># Adjust the rectangle to move the legend</span></span>
<span id="cb17-11"><a href="#cb17-11" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="main_analyses_files/figure-html/cell-13-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<div id="cell-15" class="cell" data-execution_count="3">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="co">### plot total number of counts for spliced and unspliced layers</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a><span class="co"># env rvelo_milo</span></span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>adata <span class="op">=</span> sc.read_h5ad(<span class="st">"../data/a_v_datas_merged_DECODE_v1_local_machine.h5ad"</span>)</span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a><span class="co"># subset the data to the 3' Notch and 3' control samples together</span></span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a>adata_notch_control_3p <span class="op">=</span> adata[((adata.obs[<span class="st">"perturbation"</span>] <span class="op">==</span> <span class="st">"control"</span>) <span class="op">&amp;</span></span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a>                                ((adata.obs[<span class="st">"sample_name"</span>] <span class="op">==</span> <span class="st">"TX22_A"</span>) <span class="op">|</span></span>
<span id="cb18-7"><a href="#cb18-7" aria-hidden="true" tabindex="-1"></a>                                (adata.obs[<span class="st">"sample_name"</span>] <span class="op">==</span> <span class="st">"TX23_A"</span>))) <span class="op">|</span></span>
<span id="cb18-8"><a href="#cb18-8" aria-hidden="true" tabindex="-1"></a>                                ((adata.obs[<span class="st">"perturbation"</span>] <span class="op">==</span> <span class="st">"Notch gRNA"</span>) <span class="op">&amp;</span></span>
<span id="cb18-9"><a href="#cb18-9" aria-hidden="true" tabindex="-1"></a>                                ((adata.obs[<span class="st">"sample_name"</span>] <span class="op">==</span> <span class="st">"TX22_D"</span>) <span class="op">|</span></span>
<span id="cb18-10"><a href="#cb18-10" aria-hidden="true" tabindex="-1"></a>                                 (adata.obs[<span class="st">"sample_name"</span>] <span class="op">==</span> <span class="st">"TX23_D"</span>)))].copy()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="cell-16" class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>pd.DataFrame(adata_notch_control_3p.obsm[<span class="st">"X_diffmap"</span>], index<span class="op">=</span>adata_notch_control_3p.obs_names).to_csv(<span class="st">"/g/huber/users/kharuk/DECODE/adata_notch_control_3p_diffmap.csv"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="cell-17" class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>adata_notch_control_3p <span class="op">=</span> adata_notch_control_3p.copy()</span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>adata_notch_control_3p.X <span class="op">=</span> adata_notch_control_3p.layers[<span class="st">"spliced"</span>]</span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a><span class="co"># calculate QC</span></span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a>adata_notch_control_3p.var[<span class="st">"mt"</span>] <span class="op">=</span> adata_notch_control_3p.var_names.<span class="bu">str</span>.startswith(<span class="st">"mt:"</span>)</span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a>sc.pp.calculate_qc_metrics(adata_notch_control_3p, qc_vars<span class="op">=</span>[<span class="st">"mt"</span>], percent_top<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb20-6"><a href="#cb20-6" aria-hidden="true" tabindex="-1"></a>                            log1p<span class="op">=</span><span class="va">False</span>, inplace<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb20-7"><a href="#cb20-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-8"><a href="#cb20-8" aria-hidden="true" tabindex="-1"></a>fig, axs0 <span class="op">=</span> plt.subplots(<span class="dv">1</span>, <span class="dv">1</span>, figsize<span class="op">=</span>(<span class="dv">12</span>, <span class="dv">5</span>))  <span class="co"># Adjust figsize as needed</span></span>
<span id="cb20-9"><a href="#cb20-9" aria-hidden="true" tabindex="-1"></a>sc.pl.violin(adata_notch_control_3p, [<span class="st">"total_counts"</span>], jitter<span class="op">=</span><span class="fl">0.4</span>, show<span class="op">=</span><span class="va">False</span>, groupby <span class="op">=</span> <span class="st">"high_res_annotation"</span>, rotation<span class="op">=</span><span class="dv">90</span>, ax<span class="op">=</span>axs0) <span class="co"># do not have to specify the layer as it already calculated for specific layers</span></span>
<span id="cb20-10"><a href="#cb20-10" aria-hidden="true" tabindex="-1"></a>plt.suptitle(<span class="ss">f"Quality control metrics for spliced counts"</span>, y<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb20-11"><a href="#cb20-11" aria-hidden="true" tabindex="-1"></a>axs0.spines[<span class="st">'right'</span>].set_visible(<span class="va">False</span>)</span>
<span id="cb20-12"><a href="#cb20-12" aria-hidden="true" tabindex="-1"></a>axs0.spines[<span class="st">'top'</span>].set_visible(<span class="va">False</span>)</span>
<span id="cb20-13"><a href="#cb20-13" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="main_analyses_files/figure-html/cell-16-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<div id="cell-18" class="cell" data-execution_count="8">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>adata_notch_control_3p <span class="op">=</span> adata_notch_control_3p.copy()</span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>adata_notch_control_3p.X <span class="op">=</span> adata_notch_control_3p.layers[<span class="st">"unspliced"</span>]</span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a><span class="co"># calculate QC</span></span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a>adata_notch_control_3p.var[<span class="st">"mt"</span>] <span class="op">=</span> adata_notch_control_3p.var_names.<span class="bu">str</span>.startswith(<span class="st">"mt:"</span>)</span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a>sc.pp.calculate_qc_metrics(adata_notch_control_3p, qc_vars<span class="op">=</span>[<span class="st">"mt"</span>], percent_top<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb21-6"><a href="#cb21-6" aria-hidden="true" tabindex="-1"></a>                            log1p<span class="op">=</span><span class="va">False</span>, inplace<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb21-7"><a href="#cb21-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-8"><a href="#cb21-8" aria-hidden="true" tabindex="-1"></a>fig, axs0 <span class="op">=</span> plt.subplots(<span class="dv">1</span>, <span class="dv">1</span>, figsize<span class="op">=</span>(<span class="dv">12</span>, <span class="dv">5</span>))  <span class="co"># Adjust figsize as needed</span></span>
<span id="cb21-9"><a href="#cb21-9" aria-hidden="true" tabindex="-1"></a>sc.pl.violin(adata_notch_control_3p, [<span class="st">"total_counts"</span>], jitter<span class="op">=</span><span class="fl">0.4</span>, show<span class="op">=</span><span class="va">False</span>, groupby <span class="op">=</span> <span class="st">"high_res_annotation"</span>, rotation<span class="op">=</span><span class="dv">90</span>, ax<span class="op">=</span>axs0) <span class="co"># do not have to specify the layer as it already calculated for specific layers</span></span>
<span id="cb21-10"><a href="#cb21-10" aria-hidden="true" tabindex="-1"></a>plt.suptitle(<span class="ss">f"Quality control metrics for unspliced counts"</span>, y<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb21-11"><a href="#cb21-11" aria-hidden="true" tabindex="-1"></a>axs0.spines[<span class="st">'right'</span>].set_visible(<span class="va">False</span>)</span>
<span id="cb21-12"><a href="#cb21-12" aria-hidden="true" tabindex="-1"></a>axs0.spines[<span class="st">'top'</span>].set_visible(<span class="va">False</span>)</span>
<span id="cb21-13"><a href="#cb21-13" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="main_analyses_files/figure-html/cell-17-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<div id="cell-19" class="cell" data-execution_count="9">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="co"># calculate the spliced ratio</span></span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a><span class="co"># calculate the counts for spliced and unspliced layers per cell by summing the counts in the cell for genes</span></span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a>adata_notch_control_3p.obs[<span class="st">"ncounts_S"</span>] <span class="op">=</span> np.ravel(np.<span class="bu">sum</span>(adata_notch_control_3p.layers[<span class="st">"spliced"</span>], axis<span class="op">=</span><span class="dv">1</span>))</span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a>adata_notch_control_3p.obs[<span class="st">"ncounts_U"</span>] <span class="op">=</span> np.ravel(np.<span class="bu">sum</span>(adata_notch_control_3p.layers[<span class="st">"unspliced"</span>], axis<span class="op">=</span><span class="dv">1</span>))</span>
<span id="cb22-5"><a href="#cb22-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-6"><a href="#cb22-6" aria-hidden="true" tabindex="-1"></a><span class="co"># remove cells with 0 counts in spliced or unspliced layers</span></span>
<span id="cb22-7"><a href="#cb22-7" aria-hidden="true" tabindex="-1"></a>adata_notch_control_3p <span class="op">=</span> adata_notch_control_3p[adata_notch_control_3p.obs[<span class="st">"ncounts_S"</span>] <span class="op">&gt;</span> <span class="dv">0</span>].copy()</span>
<span id="cb22-8"><a href="#cb22-8" aria-hidden="true" tabindex="-1"></a>adata_notch_control_3p <span class="op">=</span> adata_notch_control_3p[adata_notch_control_3p.obs[<span class="st">"ncounts_U"</span>] <span class="op">&gt;</span> <span class="dv">0</span>].copy()</span>
<span id="cb22-9"><a href="#cb22-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-10"><a href="#cb22-10" aria-hidden="true" tabindex="-1"></a><span class="co"># calculate the ratio of spliced counts per cell</span></span>
<span id="cb22-11"><a href="#cb22-11" aria-hidden="true" tabindex="-1"></a>adata_notch_control_3p.obs[<span class="st">"spliced_ratio"</span>] <span class="op">=</span> adata_notch_control_3p.obs[<span class="st">"ncounts_S"</span>] <span class="op">/</span> adata_notch_control_3p.obs[<span class="st">"ncounts_U"</span>]</span>
<span id="cb22-12"><a href="#cb22-12" aria-hidden="true" tabindex="-1"></a>adata_notch_control_3p.obs[<span class="st">"spliced_ratio"</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="9">
<pre><code>cell_barcode
AGTACCATCCCGAAAT-TX22_D     28.712551
AATCGACGTACCCGCA-TX22_A     55.384615
ATCCTATGTTTCACTT-TX22_A    232.700000
CCGTGAGCAGGCCCTA-TX23_A      9.617347
CACGGGTTCGAGAACG-TX22_D     19.112676
                              ...    
GGTGAAGGTACTGCCG-TX22_D     15.305732
ACTTTGTGTATGACAA-TX22_D     16.090090
TTGACCCTCGGAGTGA-TX22_D     12.138728
CAAAGAATCCCTCTAG-TX22_A    198.666667
ATGCGATGTATCCTTT-TX22_D     13.487069
Name: spliced_ratio, Length: 35428, dtype: float64</code></pre>
</div>
</div>
<div id="cell-20" class="cell" data-execution_count="15">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>high_res_anotation_colors <span class="op">=</span> adata_notch_control_3p.uns[<span class="st">"high_res_annotation_colors"</span>]</span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a>sns.boxplot(data <span class="op">=</span> adata_notch_control_3p.obs, x<span class="op">=</span><span class="st">"spliced_ratio"</span>, y<span class="op">=</span><span class="st">"high_res_annotation"</span>, palette<span class="op">=</span>high_res_anotation_colors, fliersize<span class="op">=</span><span class="dv">4</span>)</span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a>sns.despine()</span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true" tabindex="-1"></a>plt.suptitle(<span class="st">"Spliced ratio per cell type"</span>)</span>
<span id="cb24-6"><a href="#cb24-6" aria-hidden="true" tabindex="-1"></a><span class="co"># make the x axis more</span></span>
<span id="cb24-7"><a href="#cb24-7" aria-hidden="true" tabindex="-1"></a>plt.xlim(<span class="dv">0</span>, <span class="dv">500</span>)</span>
<span id="cb24-8"><a href="#cb24-8" aria-hidden="true" tabindex="-1"></a>plt.ylabel(<span class="st">"Cell types"</span>)</span>
<span id="cb24-9"><a href="#cb24-9" aria-hidden="true" tabindex="-1"></a>plt.xlabel(<span class="st">"Spliced ratio"</span>)</span>
<span id="cb24-10"><a href="#cb24-10" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="main_analyses_files/figure-html/cell-19-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="exploring-data-structure" class="level1">
<h1>3.3 Exploring data structure</h1>
<p>UMAP and diffmap may look a bit differ from figures in thesis (mirrored or small differences in clustering)</p>
<div id="cell-23" class="cell" data-execution_count="4">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="co"># env rvelo_milo, local machine </span></span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a><span class="co"># upload data</span></span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a>adata_notch_control_3p <span class="op">=</span> sc.read_h5ad(<span class="st">"../data/a_v_data_merged_DECODE_v1_Notch_Control_diffmap.h5ad"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="cell-24" class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="co"># compute pca, knn, umap and diffmap</span></span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>scv.pp.pca(adata_notch_control_3p)</span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a>sc.pp.neighbors(adata_notch_control_3p)</span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a>sc.tl.umap(adata_notch_control_3p)</span>
<span id="cb26-5"><a href="#cb26-5" aria-hidden="true" tabindex="-1"></a>sc.tl.diffmap(adata_notch_control_3p)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="cell-25" class="cell" data-execution_count="4">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>sc.settings.set_figure_params(dpi<span class="op">=</span><span class="dv">300</span>, transparent<span class="op">=</span><span class="va">True</span>, frameon<span class="op">=</span><span class="va">False</span>) <span class="co"># set dpi and background color for figures</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="cell-26" class="cell" data-execution_count="5">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="co"># add batch column to obs</span></span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>adata_notch_control_3p.obs[<span class="st">"batch"</span>] <span class="op">=</span> adata_notch_control_3p.obs[<span class="st">"sample_name"</span>].<span class="bu">str</span>[:<span class="op">-</span><span class="dv">2</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="cell-27" class="cell" data-execution_count="6">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a>sc.pl.diffmap(adata_notch_control_3p, color<span class="op">=</span><span class="st">"batch"</span>, components<span class="op">=</span><span class="st">"2,3"</span>, frameon<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a>sc.pl.umap(adata_notch_control_3p, color<span class="op">=</span><span class="st">"batch"</span>, frameon<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="main_analyses_files/figure-html/cell-24-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="main_analyses_files/figure-html/cell-24-output-2.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<div id="cell-28" class="cell" data-execution_count="7">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="co"># run the Harmony correction for the nice diffmap</span></span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a>scv.pp.pca(adata_notch_control_3p) <span class="co"># n_comps 50 is default</span></span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a>sce.pp.harmony_integrate(adata_notch_control_3p, <span class="st">"batch"</span>, adjusted_basis<span class="op">=</span><span class="st">"X_pca_harmony_batch"</span>,  max_iter_harmony <span class="op">=</span> <span class="dv">50</span>)</span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a><span class="co"># compute the knn and UMAP for this harmony data</span></span>
<span id="cb30-5"><a href="#cb30-5" aria-hidden="true" tabindex="-1"></a>sc.pp.neighbors(adata_notch_control_3p, use_rep<span class="op">=</span><span class="st">"X_pca_harmony_batch"</span>, key_added<span class="op">=</span><span class="st">"neighbors_harmony_batch"</span>, n_neighbors<span class="op">=</span><span class="dv">15</span>)</span>
<span id="cb30-6"><a href="#cb30-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-7"><a href="#cb30-7" aria-hidden="true" tabindex="-1"></a><span class="co"># compute the diffmap</span></span>
<span id="cb30-8"><a href="#cb30-8" aria-hidden="true" tabindex="-1"></a>sc.tl.diffmap(adata_notch_control_3p, n_comps<span class="op">=</span><span class="dv">15</span>, neighbors_key<span class="op">=</span><span class="st">"neighbors_harmony_batch"</span>) <span class="co"># n_comps 15 is default</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>computing PCA
    on highly variable genes
    with n_comps=50
    finished (0:00:04)
computing neighbors
    finished: added to `.uns['neighbors_harmony_batch']`
    `.obsp['neighbors_harmony_batch_distances']`, distances for each pair of neighbors
    `.obsp['neighbors_harmony_batch_connectivities']`, weighted adjacency matrix (0:00:01)
computing Diffusion Maps using n_comps=15(=n_dcs)
computing transitions
    finished (0:00:00)
    eigenvalues of transition matrix
    [1.         0.9972027  0.99087805 0.981328   0.980607   0.98039234
     0.9755672  0.9724665  0.969779   0.96264523 0.95945174 0.9583174
     0.9571273  0.95503926 0.9529672 ]
    finished: added
    'X_diffmap', diffmap coordinates (adata.obsm)
    'diffmap_evals', eigenvalues of transition matrix (adata.uns) (0:00:00)</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>2024-11-30 12:56:36,045 - harmonypy - INFO - Computing initial centroids with sklearn.KMeans...
2024-11-30 12:56:41,385 - harmonypy - INFO - sklearn.KMeans initialization complete.
2024-11-30 12:56:41,457 - harmonypy - INFO - Iteration 1 of 50
2024-11-30 12:56:46,193 - harmonypy - INFO - Iteration 2 of 50
2024-11-30 12:56:50,882 - harmonypy - INFO - Iteration 3 of 50
2024-11-30 12:56:55,862 - harmonypy - INFO - Iteration 4 of 50
2024-11-30 12:57:01,006 - harmonypy - INFO - Iteration 5 of 50
2024-11-30 12:57:05,554 - harmonypy - INFO - Iteration 6 of 50
2024-11-30 12:57:10,745 - harmonypy - INFO - Iteration 7 of 50
2024-11-30 12:57:15,794 - harmonypy - INFO - Iteration 8 of 50
2024-11-30 12:57:19,403 - harmonypy - INFO - Iteration 9 of 50
2024-11-30 12:57:21,578 - harmonypy - INFO - Converged after 9 iterations</code></pre>
</div>
</div>
<div id="cell-29" class="cell" data-execution_count="8">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a>sc.pl.diffmap(adata_notch_control_3p, color<span class="op">=</span>[<span class="st">"batch"</span>], components<span class="op">=</span><span class="st">"2,3"</span>, frameon<span class="op">=</span><span class="va">True</span>, legend_loc<span class="op">=</span><span class="va">None</span>)</span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a>sc.tl.umap(adata_notch_control_3p, neighbors_key<span class="op">=</span><span class="st">"neighbors_harmony_batch"</span>)</span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a>sc.pl.umap(adata_notch_control_3p, color<span class="op">=</span>[<span class="st">"batch"</span>], legend_loc<span class="op">=</span><span class="st">"right margin"</span>, frameon<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="main_analyses_files/figure-html/cell-26-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>computing UMAP
    finished: added
    'X_umap', UMAP coordinates (adata.obsm) (0:00:15)</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="main_analyses_files/figure-html/cell-26-output-3.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<div id="cell-30" class="cell" data-execution_count="36">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="co"># assign colors for conditions</span></span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a>adata_notch_control_3p.uns[<span class="st">"perturbation_colors"</span>] <span class="op">=</span> [<span class="st">"#1f77b4"</span>, <span class="st">"#d62728"</span>]</span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a><span class="co"># plot umap and diffmap</span></span>
<span id="cb35-4"><a href="#cb35-4" aria-hidden="true" tabindex="-1"></a>sc.pl.diffmap(adata_notch_control_3p, color<span class="op">=</span>[<span class="st">"high_res_annotation"</span>], components<span class="op">=</span><span class="st">"2,3"</span>, frameon<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb35-5"><a href="#cb35-5" aria-hidden="true" tabindex="-1"></a>sc.pl.diffmap(adata_notch_control_3p, color<span class="op">=</span>[<span class="st">"perturbation"</span>], components<span class="op">=</span><span class="st">"2,3"</span>, frameon<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb35-6"><a href="#cb35-6" aria-hidden="true" tabindex="-1"></a>sc.pl.diffmap(adata_notch_control_3p, color<span class="op">=</span>[<span class="st">"sample_name"</span>], components<span class="op">=</span><span class="st">"2,3"</span>, frameon<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb35-7"><a href="#cb35-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-8"><a href="#cb35-8" aria-hidden="true" tabindex="-1"></a>sc.pl.umap(adata_notch_control_3p, color<span class="op">=</span>[<span class="st">"high_res_annotation"</span>], legend_loc<span class="op">=</span><span class="st">"right margin"</span>, frameon<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb35-9"><a href="#cb35-9" aria-hidden="true" tabindex="-1"></a>sc.pl.umap(adata_notch_control_3p, color<span class="op">=</span>[<span class="st">"sample_name"</span>], legend_loc<span class="op">=</span><span class="va">None</span>, frameon<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb35-10"><a href="#cb35-10" aria-hidden="true" tabindex="-1"></a>sc.pl.umap(adata_notch_control_3p, color<span class="op">=</span>[<span class="st">"perturbation"</span>], legend_loc<span class="op">=</span><span class="st">"right margin"</span>, frameon<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb35-11"><a href="#cb35-11" aria-hidden="true" tabindex="-1"></a></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="main_analyses_files/figure-html/cell-27-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="main_analyses_files/figure-html/cell-27-output-2.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="main_analyses_files/figure-html/cell-27-output-3.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="main_analyses_files/figure-html/cell-27-output-4.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="main_analyses_files/figure-html/cell-27-output-5.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="main_analyses_files/figure-html/cell-27-output-6.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="rna-velocity-quality-control" class="level1">
<h1>3.4 RNA velocity quality control</h1>
<div id="cell-32" class="cell" data-execution_count="3">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="co"># env rvelo_milo</span></span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a>adata <span class="op">=</span> sc.read_h5ad(<span class="st">"../data/a_v_datas_merged_DECODE_v1_local_machine.h5ad"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="cell-33" class="cell" data-execution_count="20">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a><span class="co"># subset the data to the 3' Notch and 3' control samples together</span></span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a>adata_notch_control_3p <span class="op">=</span> adata[((adata.obs[<span class="st">"perturbation"</span>] <span class="op">==</span> <span class="st">"control"</span>) <span class="op">&amp;</span></span>
<span id="cb37-3"><a href="#cb37-3" aria-hidden="true" tabindex="-1"></a>                                ((adata.obs[<span class="st">"sample_name"</span>] <span class="op">==</span> <span class="st">"TX22_A"</span>) <span class="op">|</span></span>
<span id="cb37-4"><a href="#cb37-4" aria-hidden="true" tabindex="-1"></a>                                (adata.obs[<span class="st">"sample_name"</span>] <span class="op">==</span> <span class="st">"TX23_A"</span>))) <span class="op">|</span></span>
<span id="cb37-5"><a href="#cb37-5" aria-hidden="true" tabindex="-1"></a>                                ((adata.obs[<span class="st">"perturbation"</span>] <span class="op">==</span> <span class="st">"Notch gRNA"</span>) <span class="op">&amp;</span></span>
<span id="cb37-6"><a href="#cb37-6" aria-hidden="true" tabindex="-1"></a>                                ((adata.obs[<span class="st">"sample_name"</span>] <span class="op">==</span> <span class="st">"TX22_D"</span>) <span class="op">|</span></span>
<span id="cb37-7"><a href="#cb37-7" aria-hidden="true" tabindex="-1"></a>                                 (adata.obs[<span class="st">"sample_name"</span>] <span class="op">==</span> <span class="st">"TX23_D"</span>)))].copy()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="cell-34" class="cell" data-execution_count="5">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a>adata_notch_control_3p.obs[<span class="st">"batch"</span>] <span class="op">=</span> adata_notch_control_3p.obs[<span class="st">"sample_name"</span>].<span class="bu">str</span>[:<span class="op">-</span><span class="dv">2</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="cell-35" class="cell" data-execution_count="7">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a>scv.pl.proportions(adata_notch_control_3p, groupby<span class="op">=</span><span class="st">"high_res_annotation"</span>, figsize<span class="op">=</span>(<span class="dv">20</span>, <span class="dv">6</span>), show<span class="op">=</span><span class="va">False</span>, dpi<span class="op">=</span><span class="dv">200</span>, fontsize<span class="op">=</span><span class="dv">17</span>)</span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a>plt.suptitle(<span class="st">"Proportions of cell types in 3' Notch + Control samples"</span>, y<span class="op">=</span><span class="fl">1.05</span>, x<span class="op">=</span><span class="fl">0.6</span>, fontsize<span class="op">=</span><span class="dv">20</span>)</span>
<span id="cb39-3"><a href="#cb39-3" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="main_analyses_files/figure-html/cell-31-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<div id="cell-36" class="cell" data-execution_count="9">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a>adata_control_notch_5p <span class="op">=</span> adata[(adata.obs[<span class="st">"perturbation"</span>] <span class="op">==</span> <span class="st">"control"</span>) <span class="op">|</span></span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a>                             (adata.obs[<span class="st">"perturbation"</span>] <span class="op">==</span> <span class="st">"Notch gRNA"</span>)].copy()</span>
<span id="cb40-3"><a href="#cb40-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-4"><a href="#cb40-4" aria-hidden="true" tabindex="-1"></a><span class="co">### there are 4 3' samples in the data 22 and 23 - TX22_A, TX23_A, TX22_D, TX23_D, </span></span>
<span id="cb40-5"><a href="#cb40-5" aria-hidden="true" tabindex="-1"></a><span class="co"># I have to remove 3' Notch and control</span></span>
<span id="cb40-6"><a href="#cb40-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-7"><a href="#cb40-7" aria-hidden="true" tabindex="-1"></a><span class="co"># create a boolean for the samples</span></span>
<span id="cb40-8"><a href="#cb40-8" aria-hidden="true" tabindex="-1"></a>prime3_samples_control <span class="op">=</span> (adata_control_notch_5p.obs[<span class="st">"sample_name"</span>] <span class="op">==</span> <span class="st">"TX22_A"</span>) <span class="op">|</span> (adata_control_notch_5p.obs[<span class="st">"sample_name"</span>] <span class="op">==</span> <span class="st">"TX23_A"</span>)</span>
<span id="cb40-9"><a href="#cb40-9" aria-hidden="true" tabindex="-1"></a><span class="co"># remove the samples if the bool == True</span></span>
<span id="cb40-10"><a href="#cb40-10" aria-hidden="true" tabindex="-1"></a>adata_control_notch_5p <span class="op">=</span> adata_control_notch_5p[<span class="op">~</span>prime3_samples_control]</span>
<span id="cb40-11"><a href="#cb40-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-12"><a href="#cb40-12" aria-hidden="true" tabindex="-1"></a>prime3_samples_notch <span class="op">=</span> (adata_control_notch_5p.obs[<span class="st">"sample_name"</span>] <span class="op">==</span> <span class="st">"TX22_D"</span>) <span class="op">|</span> (adata_control_notch_5p.obs[<span class="st">"sample_name"</span>] <span class="op">==</span> <span class="st">"TX23_D"</span>)</span>
<span id="cb40-13"><a href="#cb40-13" aria-hidden="true" tabindex="-1"></a>adata_control_notch_5p <span class="op">=</span> adata_control_notch_5p[<span class="op">~</span>prime3_samples_notch]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="cell-37" class="cell" data-execution_count="19">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a>scv.pl.proportions(adata_control_notch_5p, groupby<span class="op">=</span><span class="st">"high_res_annotation"</span>, figsize<span class="op">=</span>(<span class="dv">20</span>, <span class="dv">6</span>), show<span class="op">=</span><span class="va">False</span>, dpi<span class="op">=</span><span class="dv">200</span>, fontsize<span class="op">=</span><span class="dv">17</span>)</span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a>plt.suptitle(<span class="st">"Proportions of cell types in 5' Notch + Control samples"</span>, y<span class="op">=</span><span class="fl">1.05</span>, x<span class="op">=</span><span class="fl">0.6</span>, fontsize<span class="op">=</span><span class="dv">20</span>)</span>
<span id="cb41-3"><a href="#cb41-3" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="main_analyses_files/figure-html/cell-33-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<div id="cell-38" class="cell">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a><span class="co"># RNA velocity phase portraits for 3' Notch + control</span></span>
<span id="cb42-2"><a href="#cb42-2" aria-hidden="true" tabindex="-1"></a>adata_notch_control_3p, top_ranked_genes_notch_contol_3p <span class="op">=</span> velocity_inferring_wo_plot(adata_notch_control_3p, </span>
<span id="cb42-3"><a href="#cb42-3" aria-hidden="true" tabindex="-1"></a>                                                                     min_shared_counts<span class="op">=</span><span class="dv">20</span>, n_top_genes<span class="op">=</span><span class="dv">500</span>, mode <span class="op">=</span> <span class="st">"stochastic"</span>, </span>
<span id="cb42-4"><a href="#cb42-4" aria-hidden="true" tabindex="-1"></a>                                                                     annotation <span class="op">=</span> <span class="st">"high_res_annotation"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
 Applying min_shared_counts filter 

Filtered out 8918 genes that are detected 20 counts (shared).
WARNING: Did not normalize X as it looks processed already. To enforce normalization, set `enforce=True`.
Normalized count data: spliced, unspliced.
If you pass `n_top_genes`, all cutoffs are ignored.
extracting highly variable genes
    finished (0:00:01)
--&gt; added
    'highly_variable', boolean vector (adata.var)
    'means', float vector (adata.var)
    'dispersions', float vector (adata.var)
    'dispersions_norm', float vector (adata.var)
WARNING: adata.X seems to be already log-transformed.
'
 Computing PCA, knn and UMAP 
 
computing PCA
    on highly variable genes
    with n_comps=50
    finished (0:00:01)
computing neighbors
    finished: added to `.uns['neighbors']`
    `.obsp['distances']`, distances for each pair of neighbors
    `.obsp['connectivities']`, weighted adjacency matrix (0:00:01)
computing UMAP
    finished: added
    'X_umap', UMAP coordinates (adata.obsm) (0:00:15)

 Computing RNA velocity 
 
computing neighbors
    finished (0:00:03) --&gt; added 
    'distances' and 'connectivities', weighted adjacency matrices (adata.obsp)
computing moments based on connectivities
    finished (0:00:04) --&gt; added 
    'Ms' and 'Mu', moments of un/spliced abundances (adata.layers)

 Applying stochastic mode 

computing velocities
    finished (0:00:07) --&gt; added 
    'velocity', velocity vectors for each individual cell (adata.layers)
computing velocity graph (using 1/8 cores)
    finished (0:00:48) --&gt; added 
    'velocity_graph', sparse matrix with cosine correlations (adata.uns)
ranking velocity genes
    finished (0:00:23) --&gt; added 
    'rank_velocity_genes', sorted scores by group ids (adata.uns) 
    'spearmans_score', spearmans correlation scores (adata.var)</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>/opt/homebrew/Caskroom/miniconda/base/envs/rvelo_milo/lib/python3.10/site-packages/scvelo/tools/optimization.py:184: DeprecationWarning: Conversion of an array with ndim &gt; 0 to a scalar is deprecated, and will error in future. Ensure you extract a single element from your array before performing this operation. (Deprecated NumPy 1.25.)
  gamma[i] = np.linalg.pinv(A.T.dot(A)).dot(A.T.dot(y[:, i]))
python(55559) MallocStackLogging: can't turn off malloc stack logging because it was not enabled.
/opt/homebrew/Caskroom/miniconda/base/envs/rvelo_milo/lib/python3.10/site-packages/scvelo/tools/utils.py:463: DeprecationWarning: Please use `rankdata` from the `scipy.stats` namespace, the `scipy.stats.stats` namespace is deprecated.
  from scipy.stats.stats import rankdata</code></pre>
</div>
</div>
<div id="cell-39" class="cell" data-execution_count="20">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a>kwargs <span class="op">=</span> <span class="bu">dict</span>(frameon<span class="op">=</span><span class="va">False</span>, size<span class="op">=</span><span class="dv">15</span>, linewidth<span class="op">=</span><span class="fl">1.5</span>)</span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-3"><a href="#cb45-3" aria-hidden="true" tabindex="-1"></a><span class="co"># plot the some marker genes for 3' samples</span></span>
<span id="cb45-4"><a href="#cb45-4" aria-hidden="true" tabindex="-1"></a>scv.pl.scatter(adata_notch_control_3p, [<span class="st">"N"</span>, <span class="st">"klu"</span>, <span class="st">"AstC"</span>, <span class="st">"pros"</span>, <span class="st">"Tk"</span>], <span class="op">**</span>kwargs, color<span class="op">=</span><span class="st">"high_res_annotation"</span>, show<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb45-5"><a href="#cb45-5" aria-hidden="true" tabindex="-1"></a>plt.suptitle(<span class="st">"Phase portraits for 3' Notch+Control samples"</span>, y<span class="op">=</span><span class="fl">1.25</span>)</span>
<span id="cb45-6"><a href="#cb45-6" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="main_analyses_files/figure-html/cell-35-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<div id="cell-40" class="cell" data-execution_count="21">
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a><span class="co"># RNA velocity phase portraits for 5' Notch + control</span></span>
<span id="cb46-2"><a href="#cb46-2" aria-hidden="true" tabindex="-1"></a>adata_notch_control_5p, top_ranked_genes_notch_contol_5p <span class="op">=</span> velocity_inferring_wo_plot(adata_control_notch_5p, </span>
<span id="cb46-3"><a href="#cb46-3" aria-hidden="true" tabindex="-1"></a>                                                                     min_shared_counts<span class="op">=</span><span class="dv">20</span>, n_top_genes<span class="op">=</span><span class="dv">500</span>, mode <span class="op">=</span> <span class="st">"stochastic"</span>, </span>
<span id="cb46-4"><a href="#cb46-4" aria-hidden="true" tabindex="-1"></a>                                                                     annotation <span class="op">=</span> <span class="st">"high_res_annotation"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
 Applying min_shared_counts filter 

Filtered out 10822 genes that are detected 20 counts (shared).
WARNING: Did not normalize X as it looks processed already. To enforce normalization, set `enforce=True`.
Normalized count data: spliced, unspliced.
If you pass `n_top_genes`, all cutoffs are ignored.
extracting highly variable genes
    finished (0:00:00)
--&gt; added
    'highly_variable', boolean vector (adata.var)
    'means', float vector (adata.var)
    'dispersions', float vector (adata.var)
    'dispersions_norm', float vector (adata.var)
WARNING: adata.X seems to be already log-transformed.
'
 Computing PCA, knn and UMAP 
 
computing PCA
    on highly variable genes
    with n_comps=50
    finished (0:00:02)
computing neighbors
    finished: added to `.uns['neighbors']`
    `.obsp['distances']`, distances for each pair of neighbors
    `.obsp['connectivities']`, weighted adjacency matrix (0:00:02)
computing UMAP
    finished: added
    'X_umap', UMAP coordinates (adata.obsm) (0:00:18)

 Computing RNA velocity 
 
computing neighbors
    finished (0:00:03) --&gt; added 
    'distances' and 'connectivities', weighted adjacency matrices (adata.obsp)
computing moments based on connectivities
    finished (0:00:01) --&gt; added 
    'Ms' and 'Mu', moments of un/spliced abundances (adata.layers)

 Applying stochastic mode 

computing velocities
    finished (0:00:02) --&gt; added 
    'velocity', velocity vectors for each individual cell (adata.layers)
computing velocity graph (using 1/8 cores)
    finished (0:00:17) --&gt; added 
    'velocity_graph', sparse matrix with cosine correlations (adata.uns)
ranking velocity genes
    finished (0:00:09) --&gt; added 
    'rank_velocity_genes', sorted scores by group ids (adata.uns) 
    'spearmans_score', spearmans correlation scores (adata.var)</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>/opt/homebrew/Caskroom/miniconda/base/envs/rvelo_milo/lib/python3.10/site-packages/scvelo/tools/optimization.py:184: DeprecationWarning: Conversion of an array with ndim &gt; 0 to a scalar is deprecated, and will error in future. Ensure you extract a single element from your array before performing this operation. (Deprecated NumPy 1.25.)
  gamma[i] = np.linalg.pinv(A.T.dot(A)).dot(A.T.dot(y[:, i]))
python(55586) MallocStackLogging: can't turn off malloc stack logging because it was not enabled.
/opt/homebrew/Caskroom/miniconda/base/envs/rvelo_milo/lib/python3.10/site-packages/scvelo/tools/utils.py:463: DeprecationWarning: Please use `rankdata` from the `scipy.stats` namespace, the `scipy.stats.stats` namespace is deprecated.
  from scipy.stats.stats import rankdata</code></pre>
</div>
</div>
<div id="cell-41" class="cell" data-execution_count="23">
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a>adata_notch_control_5p.uns[<span class="st">"high_res_annotation_colors"</span>] <span class="op">=</span> [<span class="st">'#1f77b4'</span>, <span class="st">'#ff7f0e'</span>, <span class="st">'#279e68'</span>, </span>
<span id="cb49-2"><a href="#cb49-2" aria-hidden="true" tabindex="-1"></a>                                                               <span class="st">'#d62728'</span>, <span class="st">'#aa40fc'</span>, <span class="st">'#8c564b'</span>, <span class="st">'#e377c2'</span>,</span>
<span id="cb49-3"><a href="#cb49-3" aria-hidden="true" tabindex="-1"></a>                                                               <span class="st">'#b5bd61'</span>, <span class="st">'#ffbb78'</span>, <span class="st">'#98df8a'</span>,</span>
<span id="cb49-4"><a href="#cb49-4" aria-hidden="true" tabindex="-1"></a>                                                                <span class="st">'#ff9896'</span>, <span class="st">'#c5b0d5'</span>, <span class="st">'#c49c94'</span>, <span class="st">'#f7b6d2'</span>]</span>
<span id="cb49-5"><a href="#cb49-5" aria-hidden="true" tabindex="-1"></a><span class="co"># rename the the cell types to the short names</span></span>
<span id="cb49-6"><a href="#cb49-6" aria-hidden="true" tabindex="-1"></a>adata_notch_control_5p.obs[<span class="st">"high_res_annotation"</span>] <span class="op">=</span> adata_notch_control_5p.obs[<span class="st">"high_res_annotation"</span>].replace({<span class="st">"Enteroblasts"</span>: <span class="st">"EB"</span>,</span>
<span id="cb49-7"><a href="#cb49-7" aria-hidden="true" tabindex="-1"></a>                                                                                                               <span class="st">"Large flat cells"</span>: <span class="st">"LFC"</span>,</span>
<span id="cb49-8"><a href="#cb49-8" aria-hidden="true" tabindex="-1"></a>                                                                                                                <span class="st">"middle Enterocytes"</span>: <span class="st">"mEC"</span>,</span>
<span id="cb49-9"><a href="#cb49-9" aria-hidden="true" tabindex="-1"></a>                                                                                                                <span class="st">"anterior Enterocytes"</span>: <span class="st">"aEC"</span>,</span>
<span id="cb49-10"><a href="#cb49-10" aria-hidden="true" tabindex="-1"></a>                                                                                                                <span class="st">"Intestinal stem cells"</span>: <span class="st">"ISC"</span>,</span>
<span id="cb49-11"><a href="#cb49-11" aria-hidden="true" tabindex="-1"></a>                                                                                                                <span class="st">"differentiating Enterocytes"</span>: <span class="st">"dEC"</span>,</span>
<span id="cb49-12"><a href="#cb49-12" aria-hidden="true" tabindex="-1"></a>                                                                                                                <span class="st">"Copper cells"</span>: <span class="st">"Copper"</span>,</span>
<span id="cb49-13"><a href="#cb49-13" aria-hidden="true" tabindex="-1"></a>                                                                                                                <span class="st">"polyploid Enterocytes"</span>: <span class="st">"pEC"</span>,</span>
<span id="cb49-14"><a href="#cb49-14" aria-hidden="true" tabindex="-1"></a>                                                                                                                <span class="st">"Malpighian tubule"</span>: <span class="st">"MT"</span>,</span>
<span id="cb49-15"><a href="#cb49-15" aria-hidden="true" tabindex="-1"></a>                                                                                                                <span class="st">"differentiating anterior Enterocytes"</span>: <span class="st">"daEC"</span>,</span>
<span id="cb49-16"><a href="#cb49-16" aria-hidden="true" tabindex="-1"></a>                                                                                                                <span class="st">"AstC Enteroendocrine cells"</span>: <span class="st">"AstC EEc"</span>,</span>
<span id="cb49-17"><a href="#cb49-17" aria-hidden="true" tabindex="-1"></a>                                                                                                                <span class="st">"Enteroendocrine progenitor cell"</span>: <span class="st">"EEpc"</span>,</span>
<span id="cb49-18"><a href="#cb49-18" aria-hidden="true" tabindex="-1"></a>                                                                                                                <span class="st">"Tk Enteroendocrine cells"</span>: <span class="st">"Tk EEc"</span>,</span>
<span id="cb49-19"><a href="#cb49-19" aria-hidden="true" tabindex="-1"></a>                                                                                                                <span class="st">"classIII Enteroendocrine cells"</span>: <span class="st">"classIII EEc"</span>})</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="cell-42" class="cell" data-execution_count="24">
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a><span class="co"># plot PPs for 5' samples</span></span>
<span id="cb50-2"><a href="#cb50-2" aria-hidden="true" tabindex="-1"></a>scv.pl.scatter(adata_notch_control_5p, [<span class="st">"N"</span>, <span class="st">"klu"</span>, <span class="st">"AstC"</span>, <span class="st">"pros"</span>, <span class="st">"Tk"</span>], <span class="op">**</span>kwargs, color<span class="op">=</span><span class="st">"high_res_annotation"</span>, show<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb50-3"><a href="#cb50-3" aria-hidden="true" tabindex="-1"></a>plt.suptitle(<span class="st">"Phase portraits for 5' Notch+Control samples"</span>, y<span class="op">=</span><span class="fl">1.25</span>)</span>
<span id="cb50-4"><a href="#cb50-4" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="main_analyses_files/figure-html/cell-38-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<div id="cell-43" class="cell" data-execution_count="5">
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> corr_adata(adata, genes<span class="op">=</span><span class="va">None</span>, method<span class="op">=</span><span class="st">'spearman'</span>):</span>
<span id="cb51-2"><a href="#cb51-2" aria-hidden="true" tabindex="-1"></a>    <span class="co">"""Linearly correlates features (genes/obs_keys) of adata with a given array.</span></span>
<span id="cb51-3"><a href="#cb51-3" aria-hidden="true" tabindex="-1"></a><span class="co">    Computes pearson linear correlation (r and p value) for each selected feature</span></span>
<span id="cb51-4"><a href="#cb51-4" aria-hidden="true" tabindex="-1"></a><span class="co">    with the given values in x.</span></span>
<span id="cb51-5"><a href="#cb51-5" aria-hidden="true" tabindex="-1"></a><span class="co">    ----------</span></span>
<span id="cb51-6"><a href="#cb51-6" aria-hidden="true" tabindex="-1"></a><span class="co">    adata: An adata object.</span></span>
<span id="cb51-7"><a href="#cb51-7" aria-hidden="true" tabindex="-1"></a><span class="co">    genes: list, optional</span></span>
<span id="cb51-8"><a href="#cb51-8" aria-hidden="true" tabindex="-1"></a><span class="co">    List of genes to analyze. If None, uses all genes</span></span>
<span id="cb51-9"><a href="#cb51-9" aria-hidden="true" tabindex="-1"></a><span class="co">    method: str</span></span>
<span id="cb51-10"><a href="#cb51-10" aria-hidden="true" tabindex="-1"></a><span class="co">    method: Either 'spearman' or 'pearson'.</span></span>
<span id="cb51-11"><a href="#cb51-11" aria-hidden="true" tabindex="-1"></a><span class="co">    Returns</span></span>
<span id="cb51-12"><a href="#cb51-12" aria-hidden="true" tabindex="-1"></a><span class="co">    -------</span></span>
<span id="cb51-13"><a href="#cb51-13" aria-hidden="true" tabindex="-1"></a><span class="co">    df: A pandas DataFrame</span></span>
<span id="cb51-14"><a href="#cb51-14" aria-hidden="true" tabindex="-1"></a><span class="co">        The dataframe has genes as index and columns pearson_r and pearson_p. It</span></span>
<span id="cb51-15"><a href="#cb51-15" aria-hidden="true" tabindex="-1"></a><span class="co">        is sorted by correlation coefficient (pearson_r).</span></span>
<span id="cb51-16"><a href="#cb51-16" aria-hidden="true" tabindex="-1"></a><span class="co">    """</span></span>
<span id="cb51-17"><a href="#cb51-17" aria-hidden="true" tabindex="-1"></a>    <span class="im">from</span> scipy.stats <span class="im">import</span> spearmanr, pearsonr</span>
<span id="cb51-18"><a href="#cb51-18" aria-hidden="true" tabindex="-1"></a>    <span class="im">from</span> scipy.sparse <span class="im">import</span> issparse</span>
<span id="cb51-19"><a href="#cb51-19" aria-hidden="true" tabindex="-1"></a>    <span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb51-20"><a href="#cb51-20" aria-hidden="true" tabindex="-1"></a>    <span class="im">import</span> numpy <span class="im">as</span> np</span>
<span id="cb51-21"><a href="#cb51-21" aria-hidden="true" tabindex="-1"></a>    <span class="im">from</span> tqdm <span class="im">import</span> tqdm</span>
<span id="cb51-22"><a href="#cb51-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb51-23"><a href="#cb51-23" aria-hidden="true" tabindex="-1"></a>    <span class="co"># select genes</span></span>
<span id="cb51-24"><a href="#cb51-24" aria-hidden="true" tabindex="-1"></a>    genes <span class="op">=</span> genes <span class="cf">if</span> genes <span class="kw">is</span> <span class="kw">not</span> <span class="va">None</span> <span class="cf">else</span> adata.var_names.tolist()</span>
<span id="cb51-25"><a href="#cb51-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb51-26"><a href="#cb51-26" aria-hidden="true" tabindex="-1"></a>    <span class="co"># select correlation method</span></span>
<span id="cb51-27"><a href="#cb51-27" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> method <span class="op">==</span> <span class="st">'spearman'</span>:</span>
<span id="cb51-28"><a href="#cb51-28" aria-hidden="true" tabindex="-1"></a>        correlate <span class="op">=</span> spearmanr</span>
<span id="cb51-29"><a href="#cb51-29" aria-hidden="true" tabindex="-1"></a>    <span class="cf">elif</span> method <span class="op">==</span> <span class="st">'pearson'</span>:</span>
<span id="cb51-30"><a href="#cb51-30" aria-hidden="true" tabindex="-1"></a>        correlate <span class="op">=</span> pearsonr</span>
<span id="cb51-31"><a href="#cb51-31" aria-hidden="true" tabindex="-1"></a>    <span class="cf">else</span>:</span>
<span id="cb51-32"><a href="#cb51-32" aria-hidden="true" tabindex="-1"></a>        <span class="cf">raise</span> <span class="pp">ValueError</span>(<span class="ss">f'Method </span><span class="sc">{</span>method<span class="sc">}</span><span class="ss"> not valid (pearson or spearman only).'</span>)</span>
<span id="cb51-33"><a href="#cb51-33" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb51-34"><a href="#cb51-34" aria-hidden="true" tabindex="-1"></a>    <span class="co"># get spliced and unspliced lahers</span></span>
<span id="cb51-35"><a href="#cb51-35" aria-hidden="true" tabindex="-1"></a>    spliced <span class="op">=</span> adata.layers[<span class="st">'spliced'</span>]</span>
<span id="cb51-36"><a href="#cb51-36" aria-hidden="true" tabindex="-1"></a>    unspliced <span class="op">=</span> adata.layers[<span class="st">'unspliced'</span>]</span>
<span id="cb51-37"><a href="#cb51-37" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb51-38"><a href="#cb51-38" aria-hidden="true" tabindex="-1"></a>    <span class="co"># feature set</span></span>
<span id="cb51-39"><a href="#cb51-39" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> issparse (spliced):</span>
<span id="cb51-40"><a href="#cb51-40" aria-hidden="true" tabindex="-1"></a>        spliced <span class="op">=</span> spliced.A</span>
<span id="cb51-41"><a href="#cb51-41" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> issparse (unspliced):</span>
<span id="cb51-42"><a href="#cb51-42" aria-hidden="true" tabindex="-1"></a>        unspliced <span class="op">=</span> unspliced.A</span>
<span id="cb51-43"><a href="#cb51-43" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb51-44"><a href="#cb51-44" aria-hidden="true" tabindex="-1"></a>    lincors <span class="op">=</span> []</span>
<span id="cb51-45"><a href="#cb51-45" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> genes_id, gene <span class="kw">in</span> <span class="bu">enumerate</span>(tqdm(genes)):</span>
<span id="cb51-46"><a href="#cb51-46" aria-hidden="true" tabindex="-1"></a>        s <span class="op">=</span> spliced[:, genes_id]</span>
<span id="cb51-47"><a href="#cb51-47" aria-hidden="true" tabindex="-1"></a>        u <span class="op">=</span> unspliced[:, genes_id]</span>
<span id="cb51-48"><a href="#cb51-48" aria-hidden="true" tabindex="-1"></a>        r, p <span class="op">=</span> correlate(s, u)</span>
<span id="cb51-49"><a href="#cb51-49" aria-hidden="true" tabindex="-1"></a>        lincors.append([gene, r, p])</span>
<span id="cb51-50"><a href="#cb51-50" aria-hidden="true" tabindex="-1"></a>    <span class="co"># format result as pandas.DataFrame</span></span>
<span id="cb51-51"><a href="#cb51-51" aria-hidden="true" tabindex="-1"></a>    df <span class="op">=</span> pd.DataFrame(lincors, columns<span class="op">=</span>[<span class="st">"gene"</span>, <span class="ss">f'</span><span class="sc">{</span>method<span class="sc">}</span><span class="ss">_r'</span>, <span class="ss">f'</span><span class="sc">{</span>method<span class="sc">}</span><span class="ss">_p'</span>])</span>
<span id="cb51-52"><a href="#cb51-52" aria-hidden="true" tabindex="-1"></a>    df <span class="op">=</span> df.set_index(<span class="st">"gene"</span>)</span>
<span id="cb51-53"><a href="#cb51-53" aria-hidden="true" tabindex="-1"></a>    df <span class="op">=</span> df.sort_values(<span class="ss">f'</span><span class="sc">{</span>method<span class="sc">}</span><span class="ss">_r'</span>, ascending<span class="op">=</span><span class="va">False</span>)  <span class="co"># sort by correlation</span></span>
<span id="cb51-54"><a href="#cb51-54" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> df</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="cell-44" class="cell" data-execution_count="6">
<div class="sourceCode cell-code" id="cb52"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a>scv.pp.filter_genes(adata_notch_control_3p, min_shared_counts<span class="op">=</span><span class="dv">20</span>)</span>
<span id="cb52-2"><a href="#cb52-2" aria-hidden="true" tabindex="-1"></a>scv.pp.filter_genes_dispersion(adata_notch_control_3p, n_top_genes<span class="op">=</span><span class="dv">1000</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Filtered out 8918 genes that are detected 20 counts (shared).
Extracted 1000 highly variable genes.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>/opt/homebrew/Caskroom/miniconda/base/envs/rvelo/lib/python3.10/site-packages/pandas/core/dtypes/cast.py:1641: DeprecationWarning: np.find_common_type is deprecated.  Please use `np.result_type` or `np.promote_types`.
See https://numpy.org/devdocs/release/1.25.0-notes.html and the docs for more information.  (Deprecated NumPy 1.25)
  return np.find_common_type(types, [])
/opt/homebrew/Caskroom/miniconda/base/envs/rvelo/lib/python3.10/site-packages/pandas/core/algorithms.py:522: DeprecationWarning: np.find_common_type is deprecated.  Please use `np.result_type` or `np.promote_types`.
See https://numpy.org/devdocs/release/1.25.0-notes.html and the docs for more information.  (Deprecated NumPy 1.25)
  common = np.find_common_type([values.dtype, comps_array.dtype], [])
/opt/homebrew/Caskroom/miniconda/base/envs/rvelo/lib/python3.10/site-packages/pandas/core/algorithms.py:522: DeprecationWarning: np.find_common_type is deprecated.  Please use `np.result_type` or `np.promote_types`.
See https://numpy.org/devdocs/release/1.25.0-notes.html and the docs for more information.  (Deprecated NumPy 1.25)
  common = np.find_common_type([values.dtype, comps_array.dtype], [])</code></pre>
</div>
</div>
<div id="cell-45" class="cell">
<div class="sourceCode cell-code" id="cb55"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a><span class="co">"""Correlation coef for 3' unspliced and spliced counts layers"""</span></span>
<span id="cb55-2"><a href="#cb55-2" aria-hidden="true" tabindex="-1"></a>df_corr_3p_spearman <span class="op">=</span> corr_adata(adata_notch_control_3p, method<span class="op">=</span><span class="st">"spearman"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="cell-46" class="cell" data-execution_count="12">
<div class="sourceCode cell-code" id="cb56"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a><span class="co"># correlation for 5' samples</span></span>
<span id="cb56-2"><a href="#cb56-2" aria-hidden="true" tabindex="-1"></a>scv.pp.filter_genes(adata_control_notch_5p, min_shared_counts<span class="op">=</span><span class="dv">20</span>)</span>
<span id="cb56-3"><a href="#cb56-3" aria-hidden="true" tabindex="-1"></a>scv.pp.filter_genes_dispersion(adata_control_notch_5p, n_top_genes<span class="op">=</span><span class="dv">1000</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Filtered out 10822 genes that are detected 20 counts (shared).
Extracted 1000 highly variable genes.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>/opt/homebrew/Caskroom/miniconda/base/envs/rvelo/lib/python3.10/site-packages/pandas/core/dtypes/cast.py:1641: DeprecationWarning: np.find_common_type is deprecated.  Please use `np.result_type` or `np.promote_types`.
See https://numpy.org/devdocs/release/1.25.0-notes.html and the docs for more information.  (Deprecated NumPy 1.25)
  return np.find_common_type(types, [])
/opt/homebrew/Caskroom/miniconda/base/envs/rvelo/lib/python3.10/site-packages/pandas/core/algorithms.py:522: DeprecationWarning: np.find_common_type is deprecated.  Please use `np.result_type` or `np.promote_types`.
See https://numpy.org/devdocs/release/1.25.0-notes.html and the docs for more information.  (Deprecated NumPy 1.25)
  common = np.find_common_type([values.dtype, comps_array.dtype], [])
/opt/homebrew/Caskroom/miniconda/base/envs/rvelo/lib/python3.10/site-packages/pandas/core/algorithms.py:522: DeprecationWarning: np.find_common_type is deprecated.  Please use `np.result_type` or `np.promote_types`.
See https://numpy.org/devdocs/release/1.25.0-notes.html and the docs for more information.  (Deprecated NumPy 1.25)
  common = np.find_common_type([values.dtype, comps_array.dtype], [])</code></pre>
</div>
</div>
<div id="cell-47" class="cell">
<div class="sourceCode cell-code" id="cb59"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb59-1"><a href="#cb59-1" aria-hidden="true" tabindex="-1"></a>df_corr_5p_spearman <span class="op">=</span> corr_adata(adata_control_notch_5p, method<span class="op">=</span><span class="st">"spearman"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="compare-correlation-for-3-and-5-samples" class="level4">
<h4 class="anchored" data-anchor-id="compare-correlation-for-3-and-5-samples">Compare correlation for 3 and 5 samples</h4>
<div id="cell-49" class="cell">
<div class="sourceCode cell-code" id="cb60"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a><span class="co">### Spearman correlation</span></span>
<span id="cb60-2"><a href="#cb60-2" aria-hidden="true" tabindex="-1"></a>df_corr_3p_spearman.rename(columns<span class="op">=</span>{<span class="st">"spearman_r"</span>: <span class="st">"spearman_r_3p_u_s"</span>}, inplace<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb60-3"><a href="#cb60-3" aria-hidden="true" tabindex="-1"></a>df_corr_5p_spearman.rename(columns<span class="op">=</span>{<span class="st">"spearman_r"</span>: <span class="st">"spearman_r_5p_u_s"</span>}, inplace<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb60-4"><a href="#cb60-4" aria-hidden="true" tabindex="-1"></a><span class="co"># change the gene index names to numbers</span></span>
<span id="cb60-5"><a href="#cb60-5" aria-hidden="true" tabindex="-1"></a>df_corr_3p_spearman.index <span class="op">=</span> <span class="bu">range</span>(<span class="dv">1</span>, <span class="dv">1001</span>)</span>
<span id="cb60-6"><a href="#cb60-6" aria-hidden="true" tabindex="-1"></a>df_corr_5p_spearman.index <span class="op">=</span> <span class="bu">range</span>(<span class="dv">1</span>, <span class="dv">1001</span>)</span>
<span id="cb60-7"><a href="#cb60-7" aria-hidden="true" tabindex="-1"></a><span class="co"># concatenate the dataframes in one for plotting</span></span>
<span id="cb60-8"><a href="#cb60-8" aria-hidden="true" tabindex="-1"></a>df_corr_3p_5p_spearman <span class="op">=</span> pd.concat([df_corr_3p_spearman, df_corr_5p_spearman], axis<span class="op">=</span><span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="cell-50" class="cell">
<div class="sourceCode cell-code" id="cb61"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb61-1"><a href="#cb61-1" aria-hidden="true" tabindex="-1"></a><span class="co">"""Filter outliers for Spearman correlation"""</span></span>
<span id="cb61-2"><a href="#cb61-2" aria-hidden="true" tabindex="-1"></a><span class="co"># filter outliers from the dataframe as in the boxplot</span></span>
<span id="cb61-3"><a href="#cb61-3" aria-hidden="true" tabindex="-1"></a>Q1 <span class="op">=</span> df_corr_3p_5p_spearman[<span class="st">'spearman_r_3p_u_s'</span>].quantile(<span class="fl">0.25</span>)</span>
<span id="cb61-4"><a href="#cb61-4" aria-hidden="true" tabindex="-1"></a>Q3 <span class="op">=</span> df_corr_3p_5p_spearman[<span class="st">'spearman_r_3p_u_s'</span>].quantile(<span class="fl">0.75</span>)</span>
<span id="cb61-5"><a href="#cb61-5" aria-hidden="true" tabindex="-1"></a>IQR <span class="op">=</span> Q3 <span class="op">-</span> Q1    <span class="co">#IQR is interquartile range. </span></span>
<span id="cb61-6"><a href="#cb61-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-7"><a href="#cb61-7" aria-hidden="true" tabindex="-1"></a><span class="bu">filter</span> <span class="op">=</span> (df_corr_3p_5p_spearman[<span class="st">'spearman_r_3p_u_s'</span>] <span class="op">&gt;=</span> Q1 <span class="op">-</span> <span class="fl">1.5</span> <span class="op">*</span> IQR) <span class="op">&amp;</span> (df_corr_3p_5p_spearman[<span class="st">'spearman_r_3p_u_s'</span>] <span class="op">&lt;=</span> Q3 <span class="op">+</span> <span class="fl">1.5</span> <span class="op">*</span>IQR)</span>
<span id="cb61-8"><a href="#cb61-8" aria-hidden="true" tabindex="-1"></a>df_corr_3p_5p_spearman[<span class="st">'spearman_r_3p_u_s'</span>].loc[<span class="bu">filter</span>]  </span>
<span id="cb61-9"><a href="#cb61-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-10"><a href="#cb61-10" aria-hidden="true" tabindex="-1"></a><span class="co"># make filtered pandas df</span></span>
<span id="cb61-11"><a href="#cb61-11" aria-hidden="true" tabindex="-1"></a>df_corr_3p_5p_spearman_filtered <span class="op">=</span> df_corr_3p_5p_spearman.loc[<span class="bu">filter</span>]</span>
<span id="cb61-12"><a href="#cb61-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-13"><a href="#cb61-13" aria-hidden="true" tabindex="-1"></a><span class="co"># filter outliers from the dataframe as in the boxplot</span></span>
<span id="cb61-14"><a href="#cb61-14" aria-hidden="true" tabindex="-1"></a>Q1 <span class="op">=</span> df_corr_3p_5p_spearman[<span class="st">'spearman_r_5p_u_s'</span>].quantile(<span class="fl">0.25</span>)</span>
<span id="cb61-15"><a href="#cb61-15" aria-hidden="true" tabindex="-1"></a>Q3 <span class="op">=</span> df_corr_3p_5p_spearman[<span class="st">'spearman_r_5p_u_s'</span>].quantile(<span class="fl">0.75</span>)</span>
<span id="cb61-16"><a href="#cb61-16" aria-hidden="true" tabindex="-1"></a>IQR <span class="op">=</span> Q3 <span class="op">-</span> Q1    <span class="co">#IQR is interquartile range. </span></span>
<span id="cb61-17"><a href="#cb61-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-18"><a href="#cb61-18" aria-hidden="true" tabindex="-1"></a><span class="bu">filter</span> <span class="op">=</span> (df_corr_3p_5p_spearman[<span class="st">'spearman_r_5p_u_s'</span>] <span class="op">&gt;=</span> Q1 <span class="op">-</span> <span class="fl">1.5</span> <span class="op">*</span> IQR) <span class="op">&amp;</span> (df_corr_3p_5p_spearman[<span class="st">'spearman_r_5p_u_s'</span>] <span class="op">&lt;=</span> Q3 <span class="op">+</span> <span class="fl">1.5</span> <span class="op">*</span>IQR)</span>
<span id="cb61-19"><a href="#cb61-19" aria-hidden="true" tabindex="-1"></a>df_corr_3p_5p_spearman[<span class="st">'spearman_r_5p_u_s'</span>].loc[<span class="bu">filter</span>]</span>
<span id="cb61-20"><a href="#cb61-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-21"><a href="#cb61-21" aria-hidden="true" tabindex="-1"></a><span class="co"># make filtered pandas df</span></span>
<span id="cb61-22"><a href="#cb61-22" aria-hidden="true" tabindex="-1"></a>df_corr_3p_5p_spearman_filtered_1 <span class="op">=</span> df_corr_3p_5p_spearman.loc[<span class="bu">filter</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="cell-51" class="cell" data-execution_count="17">
<div class="sourceCode cell-code" id="cb62"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb62-1"><a href="#cb62-1" aria-hidden="true" tabindex="-1"></a><span class="co"># make one with two columns from the filtered dataframes for Speamrman correlation</span></span>
<span id="cb62-2"><a href="#cb62-2" aria-hidden="true" tabindex="-1"></a><span class="co"># from df_corr_3p_5p_spearman_filtered take only the spearman_r_3p_u_s column</span></span>
<span id="cb62-3"><a href="#cb62-3" aria-hidden="true" tabindex="-1"></a><span class="co"># from df_corr_3p_5p_spearman_filtered_1 take only the spearman_r_5p_u_s column</span></span>
<span id="cb62-4"><a href="#cb62-4" aria-hidden="true" tabindex="-1"></a>df_corr_3p_5p_spearman_filtered_2 <span class="op">=</span> pd.concat([df_corr_3p_5p_spearman_filtered[<span class="st">"spearman_r_3p_u_s"</span>],</span>
<span id="cb62-5"><a href="#cb62-5" aria-hidden="true" tabindex="-1"></a>                                                  df_corr_3p_5p_spearman_filtered_1[<span class="st">"spearman_r_5p_u_s"</span>]], axis<span class="op">=</span><span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="cell-52" class="cell" data-execution_count="19">
<div class="sourceCode cell-code" id="cb63"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a><span class="co"># plot the filtered data with t-test Welch's</span></span>
<span id="cb63-2"><a href="#cb63-2" aria-hidden="true" tabindex="-1"></a><span class="co"># ! after filtering a few outliers still remain</span></span>
<span id="cb63-3"><a href="#cb63-3" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> plt.figure(figsize<span class="op">=</span>(<span class="fl">1.5</span>, <span class="dv">3</span>))</span>
<span id="cb63-4"><a href="#cb63-4" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> sns.boxplot(data<span class="op">=</span>df_corr_3p_5p_spearman_filtered_2[[<span class="st">"spearman_r_3p_u_s"</span>, <span class="st">"spearman_r_5p_u_s"</span>]], width<span class="op">=</span><span class="fl">0.8</span>,fliersize<span class="op">=</span><span class="dv">0</span>, dodge<span class="op">=</span><span class="va">False</span>) <span class="co"># fliersize - remove the outliers from the boxplot</span></span>
<span id="cb63-5"><a href="#cb63-5" aria-hidden="true" tabindex="-1"></a>annot <span class="op">=</span> Annotator(ax, [(<span class="st">"spearman_r_3p_u_s"</span>, <span class="st">"spearman_r_5p_u_s"</span>)], data<span class="op">=</span>df_corr_3p_5p_spearman_filtered_2[[<span class="st">"spearman_r_3p_u_s"</span>, <span class="st">"spearman_r_5p_u_s"</span>]])  </span>
<span id="cb63-6"><a href="#cb63-6" aria-hidden="true" tabindex="-1"></a>annot.configure(test<span class="op">=</span><span class="st">'t-test_welch'</span>, text_format<span class="op">=</span><span class="st">'star'</span>, loc<span class="op">=</span><span class="st">'inside'</span>, verbose<span class="op">=</span><span class="dv">2</span>)</span>
<span id="cb63-7"><a href="#cb63-7" aria-hidden="true" tabindex="-1"></a>annot.apply_test()</span>
<span id="cb63-8"><a href="#cb63-8" aria-hidden="true" tabindex="-1"></a>ax, test_results <span class="op">=</span> annot.annotate()</span>
<span id="cb63-9"><a href="#cb63-9" aria-hidden="true" tabindex="-1"></a>sns.despine()</span>
<span id="cb63-10"><a href="#cb63-10" aria-hidden="true" tabindex="-1"></a>plt.ylim(<span class="op">-</span><span class="fl">0.05</span>, <span class="fl">0.35</span>)</span>
<span id="cb63-11"><a href="#cb63-11" aria-hidden="true" tabindex="-1"></a>plt.yticks(np.arange(<span class="op">-</span><span class="fl">0.05</span>, <span class="fl">0.40</span>, <span class="fl">0.1</span>))</span>
<span id="cb63-12"><a href="#cb63-12" aria-hidden="true" tabindex="-1"></a>plt.ylabel(<span class="st">"Spearman correlation coefficient"</span>)</span>
<span id="cb63-13"><a href="#cb63-13" aria-hidden="true" tabindex="-1"></a>plt.xticks([<span class="dv">0</span>, <span class="dv">1</span>], [<span class="st">'3</span><span class="ch">\'</span><span class="st">'</span>, <span class="st">'5</span><span class="ch">\'</span><span class="st">'</span>])</span>
<span id="cb63-14"><a href="#cb63-14" aria-hidden="true" tabindex="-1"></a>plt.xlabel(<span class="st">"Samples"</span>)</span>
<span id="cb63-15"><a href="#cb63-15" aria-hidden="true" tabindex="-1"></a>plt.suptitle(<span class="st">"Spearman correlation for unspliced / spliced layers </span><span class="ch">\n</span><span class="st"> between 3' and 5' samples"</span>, fontsize<span class="op">=</span><span class="st">"medium"</span>, y<span class="op">=</span><span class="fl">1.15</span>, x<span class="op">=</span><span class="fl">0.6</span>)</span>
<span id="cb63-16"><a href="#cb63-16" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>p-value annotation legend:
      ns: 5.00e-02 &lt; p &lt;= 1.00e+00
       *: 1.00e-02 &lt; p &lt;= 5.00e-02
      **: 1.00e-03 &lt; p &lt;= 1.00e-02
     ***: 1.00e-04 &lt; p &lt;= 1.00e-03
    ****: p &lt;= 1.00e-04

spearman_r_3p_u_s vs. spearman_r_5p_u_s: Welch's t-test independent samples, P_val:1.209e-94 t=2.262e+01</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="main_analyses_files/figure-html/cell-47-output-2.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="rna-velocity-estimation-approaches-affect-phase-portraits" class="level1">
<h1>3.6 RNA velocity estimation approaches affect phase portraits</h1>
<div id="cell-54" class="cell" data-execution_count="6">
<div class="sourceCode cell-code" id="cb65"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb65-1"><a href="#cb65-1" aria-hidden="true" tabindex="-1"></a><span class="co"># env rvelo milo</span></span>
<span id="cb65-2"><a href="#cb65-2" aria-hidden="true" tabindex="-1"></a><span class="co"># subset the data to the 3' 22 Samples</span></span>
<span id="cb65-3"><a href="#cb65-3" aria-hidden="true" tabindex="-1"></a>adata_3p_22 <span class="op">=</span> adata[((adata.obs[<span class="st">"sample_name"</span>] <span class="op">==</span> <span class="st">"TX22_A"</span>) <span class="op">|</span></span>
<span id="cb65-4"><a href="#cb65-4" aria-hidden="true" tabindex="-1"></a>                    (adata.obs[<span class="st">"sample_name"</span>] <span class="op">==</span> <span class="st">"TX22_D"</span>))].copy()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="cell-55" class="cell" data-execution_count="8">
<div class="sourceCode cell-code" id="cb66"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb66-1"><a href="#cb66-1" aria-hidden="true" tabindex="-1"></a>adata_notch_control_3p_22, top_ranked_genes_n_c_3p_22 <span class="op">=</span> velocity_inferring_wo_plot(adata_3p_22, min_shared_counts<span class="op">=</span><span class="dv">20</span>, n_top_genes<span class="op">=</span><span class="dv">500</span>, mode <span class="op">=</span> <span class="st">"stochastic"</span>, annotation <span class="op">=</span> <span class="st">"high_res_annotation"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
 Applying min_shared_counts filter 

Filtered out 9139 genes that are detected 20 counts (shared).
WARNING: Did not normalize X as it looks processed already. To enforce normalization, set `enforce=True`.
Normalized count data: spliced, unspliced.
If you pass `n_top_genes`, all cutoffs are ignored.
extracting highly variable genes
    finished (0:00:01)
--&gt; added
    'highly_variable', boolean vector (adata.var)
    'means', float vector (adata.var)
    'dispersions', float vector (adata.var)
    'dispersions_norm', float vector (adata.var)
WARNING: adata.X seems to be already log-transformed.
'
 Computing PCA, knn and UMAP 
 
computing PCA
    on highly variable genes
    with n_comps=50
    finished (0:00:01)
computing neighbors
    finished: added to `.uns['neighbors']`
    `.obsp['distances']`, distances for each pair of neighbors
    `.obsp['connectivities']`, weighted adjacency matrix (0:00:09)
computing UMAP
    finished: added
    'X_umap', UMAP coordinates (adata.obsm) (0:00:10)

 Computing RNA velocity 
 
computing neighbors
    finished (0:00:02) --&gt; added 
    'distances' and 'connectivities', weighted adjacency matrices (adata.obsp)
computing moments based on connectivities
    finished (0:00:03) --&gt; added 
    'Ms' and 'Mu', moments of un/spliced abundances (adata.layers)

 Applying stochastic mode 

computing velocities
    finished (0:00:03) --&gt; added 
    'velocity', velocity vectors for each individual cell (adata.layers)
computing velocity graph (using 1/8 cores)
WARNING: Unable to create progress bar. Consider installing `tqdm` as `pip install tqdm` and `ipywidgets` as `pip install ipywidgets`,
or disable the progress bar using `show_progress_bar=False`.
    finished (0:00:28) --&gt; added 
    'velocity_graph', sparse matrix with cosine correlations (adata.uns)
ranking velocity genes
    finished (0:00:12) --&gt; added 
    'rank_velocity_genes', sorted scores by group ids (adata.uns) 
    'spearmans_score', spearmans correlation scores (adata.var)</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>/opt/homebrew/Caskroom/miniconda/base/envs/rvelo_milo/lib/python3.10/site-packages/scvelo/tools/optimization.py:184: DeprecationWarning: Conversion of an array with ndim &gt; 0 to a scalar is deprecated, and will error in future. Ensure you extract a single element from your array before performing this operation. (Deprecated NumPy 1.25.)
  gamma[i] = np.linalg.pinv(A.T.dot(A)).dot(A.T.dot(y[:, i]))
/opt/homebrew/Caskroom/miniconda/base/envs/rvelo_milo/lib/python3.10/site-packages/scvelo/tools/utils.py:463: DeprecationWarning: Please use `rankdata` from the `scipy.stats` namespace, the `scipy.stats.stats` namespace is deprecated.
  from scipy.stats.stats import rankdata</code></pre>
</div>
</div>
<div id="cell-56" class="cell" data-execution_count="22">
<div class="sourceCode cell-code" id="cb69"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb69-1"><a href="#cb69-1" aria-hidden="true" tabindex="-1"></a>adata_notch_control_3p_22.uns[<span class="st">"high_res_annotation_colors"</span>] </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="22">
<pre><code>['#023fa5',
 '#7d87b9',
 '#bec1d4',
 '#d6bcc0',
 '#bb7784',
 '#8e063b',
 '#4a6fe3',
 '#8595e1',
 '#b5bbe3',
 '#e6afb9',
 '#e07b91',
 '#d33f6a',
 '#11c638',
 '#8dd593',
 '#c6dec7',
 '#ead3c6']</code></pre>
</div>
</div>
<div id="cell-57" class="cell" data-execution_count="24">
<div class="sourceCode cell-code" id="cb71"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb71-1"><a href="#cb71-1" aria-hidden="true" tabindex="-1"></a>adata_notch_control_3p_22.uns[<span class="st">"high_res_annotation_colors"</span>] <span class="op">=</span> [<span class="st">'#1f77b4'</span>, </span>
<span id="cb71-2"><a href="#cb71-2" aria-hidden="true" tabindex="-1"></a>                                                               <span class="st">'#ff7f0e'</span>, </span>
<span id="cb71-3"><a href="#cb71-3" aria-hidden="true" tabindex="-1"></a>                                                               <span class="st">'#279e68'</span>, </span>
<span id="cb71-4"><a href="#cb71-4" aria-hidden="true" tabindex="-1"></a>                                                               <span class="st">'#d62728'</span>, </span>
<span id="cb71-5"><a href="#cb71-5" aria-hidden="true" tabindex="-1"></a>                                                               <span class="st">'#aa40fc'</span>, </span>
<span id="cb71-6"><a href="#cb71-6" aria-hidden="true" tabindex="-1"></a>                                                               <span class="st">'#8c564b'</span>, </span>
<span id="cb71-7"><a href="#cb71-7" aria-hidden="true" tabindex="-1"></a>                                                               <span class="st">'#e377c2'</span>,</span>
<span id="cb71-8"><a href="#cb71-8" aria-hidden="true" tabindex="-1"></a>                                                               <span class="st">'#b5bd61'</span>, </span>
<span id="cb71-9"><a href="#cb71-9" aria-hidden="true" tabindex="-1"></a>                                                               <span class="st">'#17becf'</span>, </span>
<span id="cb71-10"><a href="#cb71-10" aria-hidden="true" tabindex="-1"></a>                                                               <span class="st">'#ffbb78'</span>, </span>
<span id="cb71-11"><a href="#cb71-11" aria-hidden="true" tabindex="-1"></a>                                                               <span class="st">'#98df8a'</span>,</span>
<span id="cb71-12"><a href="#cb71-12" aria-hidden="true" tabindex="-1"></a>                                                                <span class="st">'#ff9896'</span>, </span>
<span id="cb71-13"><a href="#cb71-13" aria-hidden="true" tabindex="-1"></a>                                                                <span class="st">'#c5b0d5'</span>, </span>
<span id="cb71-14"><a href="#cb71-14" aria-hidden="true" tabindex="-1"></a>                                                                <span class="st">'#c49c94'</span>, </span>
<span id="cb71-15"><a href="#cb71-15" aria-hidden="true" tabindex="-1"></a>                                                                <span class="st">'#f7b6d2'</span>] </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="cell-58" class="cell">
<div class="sourceCode cell-code" id="cb72"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb72-1"><a href="#cb72-1" aria-hidden="true" tabindex="-1"></a>scv.pl.scatter(adata_notch_control_3p_22, [<span class="st">"msi"</span>, <span class="st">"cpo"</span>, <span class="st">"trol"</span>, <span class="st">'ed'</span>, <span class="st">"mbl"</span>, <span class="st">"Rbfox1"</span>], color<span class="op">=</span><span class="st">'high_res_annotation'</span>, frameon<span class="op">=</span><span class="va">False</span>, size<span class="op">=</span><span class="dv">15</span>, linewidth<span class="op">=</span><span class="fl">1.5</span>, show<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb72-2"><a href="#cb72-2" aria-hidden="true" tabindex="-1"></a>plt.suptitle(<span class="st">"RNA velocity, per batch (22) workflow, 3' Notch + Control"</span>, y<span class="op">=</span><span class="fl">1.25</span>)</span>
<span id="cb72-3"><a href="#cb72-3" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="main_analyses_files/figure-html/cell-52-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<div id="cell-59" class="cell" data-execution_count="16">
<div class="sourceCode cell-code" id="cb73"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb73-1"><a href="#cb73-1" aria-hidden="true" tabindex="-1"></a><span class="co"># subset the data to the 3' 22 Samples</span></span>
<span id="cb73-2"><a href="#cb73-2" aria-hidden="true" tabindex="-1"></a>adata_3p_23 <span class="op">=</span> adata[((adata.obs[<span class="st">"sample_name"</span>] <span class="op">==</span> <span class="st">"TX23_A"</span>) <span class="op">|</span></span>
<span id="cb73-3"><a href="#cb73-3" aria-hidden="true" tabindex="-1"></a>                    (adata.obs[<span class="st">"sample_name"</span>] <span class="op">==</span> <span class="st">"TX23_D"</span>))].copy()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="cell-60" class="cell" data-execution_count="17">
<div class="sourceCode cell-code" id="cb74"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb74-1"><a href="#cb74-1" aria-hidden="true" tabindex="-1"></a>adata_notch_control_3p_23, top_ranked_genes_n_c_3p_23 <span class="op">=</span> velocity_inferring_wo_plot(adata_3p_23, min_shared_counts<span class="op">=</span><span class="dv">20</span>, n_top_genes<span class="op">=</span><span class="dv">500</span>, mode <span class="op">=</span> <span class="st">"stochastic"</span>, annotation <span class="op">=</span> <span class="st">"high_res_annotation"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
 Applying min_shared_counts filter 

Filtered out 10065 genes that are detected 20 counts (shared).
WARNING: Did not normalize X as it looks processed already. To enforce normalization, set `enforce=True`.
Normalized count data: spliced, unspliced.
If you pass `n_top_genes`, all cutoffs are ignored.
extracting highly variable genes
    finished (0:00:00)
--&gt; added
    'highly_variable', boolean vector (adata.var)
    'means', float vector (adata.var)
    'dispersions', float vector (adata.var)
    'dispersions_norm', float vector (adata.var)
WARNING: adata.X seems to be already log-transformed.
'
 Computing PCA, knn and UMAP 
 
computing PCA
    on highly variable genes
    with n_comps=50
    finished (0:00:00)
computing neighbors
    finished: added to `.uns['neighbors']`
    `.obsp['distances']`, distances for each pair of neighbors
    `.obsp['connectivities']`, weighted adjacency matrix (0:00:09)
computing UMAP
    finished: added
    'X_umap', UMAP coordinates (adata.obsm) (0:00:06)

 Computing RNA velocity 
 
computing neighbors
    finished (0:00:01) --&gt; added 
    'distances' and 'connectivities', weighted adjacency matrices (adata.obsp)
computing moments based on connectivities
    finished (0:00:01) --&gt; added 
    'Ms' and 'Mu', moments of un/spliced abundances (adata.layers)

 Applying stochastic mode 

computing velocities
    finished (0:00:01) --&gt; added 
    'velocity', velocity vectors for each individual cell (adata.layers)
computing velocity graph (using 1/8 cores)
WARNING: Unable to create progress bar. Consider installing `tqdm` as `pip install tqdm` and `ipywidgets` as `pip install ipywidgets`,
or disable the progress bar using `show_progress_bar=False`.
    finished (0:00:14) --&gt; added 
    'velocity_graph', sparse matrix with cosine correlations (adata.uns)
ranking velocity genes
    finished (0:00:04) --&gt; added 
    'rank_velocity_genes', sorted scores by group ids (adata.uns) 
    'spearmans_score', spearmans correlation scores (adata.var)</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>/opt/homebrew/Caskroom/miniconda/base/envs/rvelo_milo/lib/python3.10/site-packages/scvelo/tools/optimization.py:184: DeprecationWarning: Conversion of an array with ndim &gt; 0 to a scalar is deprecated, and will error in future. Ensure you extract a single element from your array before performing this operation. (Deprecated NumPy 1.25.)
  gamma[i] = np.linalg.pinv(A.T.dot(A)).dot(A.T.dot(y[:, i]))
/opt/homebrew/Caskroom/miniconda/base/envs/rvelo_milo/lib/python3.10/site-packages/scvelo/tools/utils.py:463: DeprecationWarning: Please use `rankdata` from the `scipy.stats` namespace, the `scipy.stats.stats` namespace is deprecated.
  from scipy.stats.stats import rankdata</code></pre>
</div>
</div>
<div id="cell-61" class="cell" data-execution_count="18">
<div class="sourceCode cell-code" id="cb77"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb77-1"><a href="#cb77-1" aria-hidden="true" tabindex="-1"></a>adata_notch_control_3p_23.uns[<span class="st">"high_res_annotation_colors"</span>] <span class="op">=</span> [<span class="st">'#1f77b4'</span>, </span>
<span id="cb77-2"><a href="#cb77-2" aria-hidden="true" tabindex="-1"></a>                                                               <span class="st">'#ff7f0e'</span>, </span>
<span id="cb77-3"><a href="#cb77-3" aria-hidden="true" tabindex="-1"></a>                                                               <span class="st">'#279e68'</span>, </span>
<span id="cb77-4"><a href="#cb77-4" aria-hidden="true" tabindex="-1"></a>                                                               <span class="st">'#d62728'</span>, </span>
<span id="cb77-5"><a href="#cb77-5" aria-hidden="true" tabindex="-1"></a>                                                               <span class="st">'#aa40fc'</span>, </span>
<span id="cb77-6"><a href="#cb77-6" aria-hidden="true" tabindex="-1"></a>                                                               <span class="st">'#8c564b'</span>, </span>
<span id="cb77-7"><a href="#cb77-7" aria-hidden="true" tabindex="-1"></a>                                                               <span class="st">'#e377c2'</span>,</span>
<span id="cb77-8"><a href="#cb77-8" aria-hidden="true" tabindex="-1"></a>                                                               <span class="st">'#b5bd61'</span>, </span>
<span id="cb77-9"><a href="#cb77-9" aria-hidden="true" tabindex="-1"></a>                                                               <span class="st">'#17becf'</span>, </span>
<span id="cb77-10"><a href="#cb77-10" aria-hidden="true" tabindex="-1"></a>                                                               <span class="st">'#ffbb78'</span>, </span>
<span id="cb77-11"><a href="#cb77-11" aria-hidden="true" tabindex="-1"></a>                                                               <span class="st">'#98df8a'</span>,</span>
<span id="cb77-12"><a href="#cb77-12" aria-hidden="true" tabindex="-1"></a>                                                                <span class="st">'#ff9896'</span>, </span>
<span id="cb77-13"><a href="#cb77-13" aria-hidden="true" tabindex="-1"></a>                                                                <span class="st">'#c5b0d5'</span>, </span>
<span id="cb77-14"><a href="#cb77-14" aria-hidden="true" tabindex="-1"></a>                                                                <span class="st">'#c49c94'</span>, </span>
<span id="cb77-15"><a href="#cb77-15" aria-hidden="true" tabindex="-1"></a>                                                                <span class="st">'#f7b6d2'</span>] </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="cell-62" class="cell" data-execution_count="19">
<div class="sourceCode cell-code" id="cb78"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb78-1"><a href="#cb78-1" aria-hidden="true" tabindex="-1"></a>scv.pl.scatter(adata_notch_control_3p_23, [<span class="st">"msi"</span>, <span class="st">"cpo"</span>, <span class="st">"trol"</span>, <span class="st">'ed'</span>, <span class="st">"mbl"</span>, <span class="st">"Rbfox1"</span>], color<span class="op">=</span><span class="st">'high_res_annotation'</span>, frameon<span class="op">=</span><span class="va">False</span>, size<span class="op">=</span><span class="dv">15</span>, linewidth<span class="op">=</span><span class="fl">1.5</span>, show<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb78-2"><a href="#cb78-2" aria-hidden="true" tabindex="-1"></a>plt.suptitle(<span class="st">"RNA velocity, per batch (23) workflow, 3' Notch + Control"</span>, y<span class="op">=</span><span class="fl">1.25</span>)</span>
<span id="cb78-3"><a href="#cb78-3" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="main_analyses_files/figure-html/cell-56-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<div id="cell-63" class="cell">
<div class="sourceCode cell-code" id="cb79"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb79-1"><a href="#cb79-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Standard rna velocity workflow</span></span>
<span id="cb79-2"><a href="#cb79-2" aria-hidden="true" tabindex="-1"></a>scv.pp.filter_genes(adata_notch_control_3p, min_shared_counts<span class="op">=</span><span class="dv">20</span>)        </span>
<span id="cb79-3"><a href="#cb79-3" aria-hidden="true" tabindex="-1"></a>scv.pp.normalize_per_cell(adata_notch_control_3p)</span>
<span id="cb79-4"><a href="#cb79-4" aria-hidden="true" tabindex="-1"></a>sc.pp.highly_variable_genes(adata_notch_control_3p, n_top_genes<span class="op">=</span><span class="dv">500</span>)</span>
<span id="cb79-5"><a href="#cb79-5" aria-hidden="true" tabindex="-1"></a>sc.pp.log1p(adata_notch_control_3p)</span>
<span id="cb79-6"><a href="#cb79-6" aria-hidden="true" tabindex="-1"></a>scv.pp.pca(adata_notch_control_3p, n_comps<span class="op">=</span><span class="dv">50</span>) <span class="co"># 50 is the default value</span></span>
<span id="cb79-7"><a href="#cb79-7" aria-hidden="true" tabindex="-1"></a>sc.pp.neighbors(adata_notch_control_3p, use_rep<span class="op">=</span><span class="st">"X_pca"</span>, n_neighbors<span class="op">=</span><span class="dv">15</span>) <span class="co"># knn 15 is the default value </span></span>
<span id="cb79-8"><a href="#cb79-8" aria-hidden="true" tabindex="-1"></a>scv.pp.moments(adata_notch_control_3p, n_pcs<span class="op">=</span><span class="dv">30</span>, n_neighbors<span class="op">=</span><span class="dv">30</span>)</span>
<span id="cb79-9"><a href="#cb79-9" aria-hidden="true" tabindex="-1"></a>scv.tl.velocity(adata_notch_control_3p, mode<span class="op">=</span><span class="st">"stochastic"</span>)</span>
<span id="cb79-10"><a href="#cb79-10" aria-hidden="true" tabindex="-1"></a>scv.tl.velocity_graph(adata_notch_control_3p)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Filtered out 8918 genes that are detected 20 counts (shared).
WARNING: Did not normalize X as it looks processed already. To enforce normalization, set `enforce=True`.
Normalized count data: spliced, unspliced.
If you pass `n_top_genes`, all cutoffs are ignored.
extracting highly variable genes
    finished (0:00:01)
--&gt; added
    'highly_variable', boolean vector (adata.var)
    'means', float vector (adata.var)
    'dispersions', float vector (adata.var)
    'dispersions_norm', float vector (adata.var)
WARNING: adata.X seems to be already log-transformed.
computing PCA
    on highly variable genes
    with n_comps=50
    finished (0:00:01)
computing neighbors
    finished: added to `.uns['neighbors']`
    `.obsp['distances']`, distances for each pair of neighbors
    `.obsp['connectivities']`, weighted adjacency matrix (0:00:02)
computing neighbors
    finished (0:00:03) --&gt; added 
    'distances' and 'connectivities', weighted adjacency matrices (adata.obsp)
computing moments based on connectivities
    finished (0:00:04) --&gt; added 
    'Ms' and 'Mu', moments of un/spliced abundances (adata.layers)
computing velocities
    finished (0:00:07) --&gt; added 
    'velocity', velocity vectors for each individual cell (adata.layers)
computing velocity graph (using 1/8 cores)
    finished (0:00:46) --&gt; added 
    'velocity_graph', sparse matrix with cosine correlations (adata.uns)</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>/opt/homebrew/Caskroom/miniconda/base/envs/rvelo_milo/lib/python3.10/site-packages/scvelo/tools/optimization.py:184: DeprecationWarning: Conversion of an array with ndim &gt; 0 to a scalar is deprecated, and will error in future. Ensure you extract a single element from your array before performing this operation. (Deprecated NumPy 1.25.)
  gamma[i] = np.linalg.pinv(A.T.dot(A)).dot(A.T.dot(y[:, i]))</code></pre>
</div>
</div>
<div id="cell-64" class="cell" data-execution_count="28">
<div class="sourceCode cell-code" id="cb82"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb82-1"><a href="#cb82-1" aria-hidden="true" tabindex="-1"></a>adata_notch_control_3p.uns[<span class="st">"high_res_annotation_colors"</span>] <span class="op">=</span> [<span class="st">'#1f77b4'</span>, </span>
<span id="cb82-2"><a href="#cb82-2" aria-hidden="true" tabindex="-1"></a>                                                               <span class="st">'#ff7f0e'</span>, </span>
<span id="cb82-3"><a href="#cb82-3" aria-hidden="true" tabindex="-1"></a>                                                               <span class="st">'#279e68'</span>, </span>
<span id="cb82-4"><a href="#cb82-4" aria-hidden="true" tabindex="-1"></a>                                                               <span class="st">'#d62728'</span>, </span>
<span id="cb82-5"><a href="#cb82-5" aria-hidden="true" tabindex="-1"></a>                                                               <span class="st">'#aa40fc'</span>, </span>
<span id="cb82-6"><a href="#cb82-6" aria-hidden="true" tabindex="-1"></a>                                                               <span class="st">'#8c564b'</span>, </span>
<span id="cb82-7"><a href="#cb82-7" aria-hidden="true" tabindex="-1"></a>                                                               <span class="st">'#e377c2'</span>,</span>
<span id="cb82-8"><a href="#cb82-8" aria-hidden="true" tabindex="-1"></a>                                                               <span class="st">'#b5bd61'</span>, </span>
<span id="cb82-9"><a href="#cb82-9" aria-hidden="true" tabindex="-1"></a>                                                               <span class="st">'#17becf'</span>, </span>
<span id="cb82-10"><a href="#cb82-10" aria-hidden="true" tabindex="-1"></a>                                                               <span class="st">'#ffbb78'</span>, </span>
<span id="cb82-11"><a href="#cb82-11" aria-hidden="true" tabindex="-1"></a>                                                               <span class="st">'#98df8a'</span>,</span>
<span id="cb82-12"><a href="#cb82-12" aria-hidden="true" tabindex="-1"></a>                                                                <span class="st">'#ff9896'</span>, </span>
<span id="cb82-13"><a href="#cb82-13" aria-hidden="true" tabindex="-1"></a>                                                                <span class="st">'#c5b0d5'</span>, </span>
<span id="cb82-14"><a href="#cb82-14" aria-hidden="true" tabindex="-1"></a>                                                                <span class="st">'#c49c94'</span>, </span>
<span id="cb82-15"><a href="#cb82-15" aria-hidden="true" tabindex="-1"></a>                                                                <span class="st">'#f7b6d2'</span>] </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="cell-65" class="cell" data-execution_count="30">
<div class="sourceCode cell-code" id="cb83"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb83-1"><a href="#cb83-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> warnings</span>
<span id="cb83-2"><a href="#cb83-2" aria-hidden="true" tabindex="-1"></a>warnings.filterwarnings(<span class="st">"ignore"</span>, category<span class="op">=</span><span class="pp">DeprecationWarning</span>)</span>
<span id="cb83-3"><a href="#cb83-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb83-4"><a href="#cb83-4" aria-hidden="true" tabindex="-1"></a><span class="co"># try to get rid of the warnings</span></span>
<span id="cb83-5"><a href="#cb83-5" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> warnings</span>
<span id="cb83-6"><a href="#cb83-6" aria-hidden="true" tabindex="-1"></a>warnings.filterwarnings(<span class="st">"ignore"</span>, category<span class="op">=</span><span class="pp">DeprecationWarning</span>, module<span class="op">=</span><span class="st">"scvelo.plotting.utils"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="cell-66" class="cell" data-execution_count="32">
<div class="sourceCode cell-code" id="cb84"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb84-1"><a href="#cb84-1" aria-hidden="true" tabindex="-1"></a>scv.pl.scatter(adata_notch_control_3p, [<span class="st">"msi"</span>, <span class="st">"cpo"</span>, <span class="st">"trol"</span>, <span class="st">'ed'</span>, <span class="st">"mbl"</span>, <span class="st">"Rbfox1"</span>], color<span class="op">=</span><span class="st">'high_res_annotation'</span>, frameon<span class="op">=</span><span class="va">False</span>, size<span class="op">=</span><span class="dv">15</span>, linewidth<span class="op">=</span><span class="fl">1.5</span>, show<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb84-2"><a href="#cb84-2" aria-hidden="true" tabindex="-1"></a>plt.suptitle(<span class="st">"RNA velocity, standard approach, 3' Notch + Control"</span>, y<span class="op">=</span><span class="fl">1.25</span>)</span>
<span id="cb84-3"><a href="#cb84-3" aria-hidden="true" tabindex="-1"></a>plt.show()</span>
<span id="cb84-4"><a href="#cb84-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-5"><a href="#cb84-5" aria-hidden="true" tabindex="-1"></a>adata_notch_control_3p.uns[<span class="st">"perturbation_colors"</span>] <span class="op">=</span> [<span class="st">"#1f77b4"</span>, <span class="st">"#d62728"</span>]</span>
<span id="cb84-6"><a href="#cb84-6" aria-hidden="true" tabindex="-1"></a>scv.pl.scatter(adata_notch_control_3p, [<span class="st">"msi"</span>, <span class="st">"cpo"</span>, <span class="st">"trol"</span>, <span class="st">'ed'</span>, <span class="st">"mbl"</span>, <span class="st">"Rbfox1"</span>], color<span class="op">=</span><span class="st">'perturbation'</span>, frameon<span class="op">=</span><span class="va">False</span>, size<span class="op">=</span><span class="dv">15</span>, linewidth<span class="op">=</span><span class="fl">1.5</span>, show<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb84-7"><a href="#cb84-7" aria-hidden="true" tabindex="-1"></a>plt.suptitle(<span class="st">"RNA velocity, standard approach, 3' Notch + Control"</span>, y<span class="op">=</span><span class="fl">1.25</span>)</span>
<span id="cb84-8"><a href="#cb84-8" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="main_analyses_files/figure-html/cell-60-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="main_analyses_files/figure-html/cell-60-output-2.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<div id="cell-67" class="cell" data-execution_count="17">
<div class="sourceCode cell-code" id="cb85"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb85-1"><a href="#cb85-1" aria-hidden="true" tabindex="-1"></a><span class="co">##### RNA velocity "split moments" approach</span></span>
<span id="cb85-2"><a href="#cb85-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb85-3"><a href="#cb85-3" aria-hidden="true" tabindex="-1"></a>adata_control_3p <span class="op">=</span> adata[(adata.obs[<span class="st">"perturbation"</span>] <span class="op">==</span> <span class="st">"control"</span>) <span class="op">&amp;</span> </span>
<span id="cb85-4"><a href="#cb85-4" aria-hidden="true" tabindex="-1"></a>                         ((adata.obs[<span class="st">"sample_name"</span>] <span class="op">==</span> <span class="st">"TX22_A"</span>) <span class="op">|</span> </span>
<span id="cb85-5"><a href="#cb85-5" aria-hidden="true" tabindex="-1"></a>                          (adata.obs[<span class="st">"sample_name"</span>] <span class="op">==</span> <span class="st">"TX23_A"</span>))].copy()</span>
<span id="cb85-6"><a href="#cb85-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb85-7"><a href="#cb85-7" aria-hidden="true" tabindex="-1"></a>adata_notch_3p <span class="op">=</span> adata[(adata.obs[<span class="st">"perturbation"</span>] <span class="op">==</span> <span class="st">"Notch gRNA"</span>) <span class="op">&amp;</span></span>
<span id="cb85-8"><a href="#cb85-8" aria-hidden="true" tabindex="-1"></a>                            ((adata.obs[<span class="st">"sample_name"</span>] <span class="op">==</span> <span class="st">"TX22_D"</span>) <span class="op">|</span></span>
<span id="cb85-9"><a href="#cb85-9" aria-hidden="true" tabindex="-1"></a>                            (adata.obs[<span class="st">"sample_name"</span>] <span class="op">==</span> <span class="st">"TX23_D"</span>))].copy()</span>
<span id="cb85-10"><a href="#cb85-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb85-11"><a href="#cb85-11" aria-hidden="true" tabindex="-1"></a>adata_notch_control_3p_moments <span class="op">=</span> velocity_inference_moments(adata_control_3p, adata_notch_3p)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>WARNING: Did not normalize X as it looks processed already. To enforce normalization, set `enforce=True`.
Normalized count data: spliced, unspliced.
computing PCA
    on highly variable genes
    with n_comps=50
    finished (0:00:01)
computing neighbors
    finished (0:00:01) --&gt; added 
    'distances' and 'connectivities', weighted adjacency matrices (adata.obsp)
computing moments based on connectivities
    finished (0:00:03) --&gt; added 
    'Ms' and 'Mu', moments of un/spliced abundances (adata.layers)
WARNING: Did not normalize X as it looks processed already. To enforce normalization, set `enforce=True`.
Normalized count data: spliced, unspliced.
computing PCA
    on highly variable genes
    with n_comps=50
    finished (0:00:01)
computing neighbors
    finished (0:00:01) --&gt; added 
    'distances' and 'connectivities', weighted adjacency matrices (adata.obsp)
computing moments based on connectivities
    finished (0:00:04) --&gt; added 
    'Ms' and 'Mu', moments of un/spliced abundances (adata.layers)
Filtered out 9458 genes that are detected 20 counts (shared).
WARNING: Did not normalize X as it looks processed already. To enforce normalization, set `enforce=True`.
WARNING: Did not normalize spliced as it looks processed already. To enforce normalization, set `enforce=True`.
WARNING: Did not normalize unspliced as it looks processed already. To enforce normalization, set `enforce=True`.
Extracted 2000 highly variable genes.
Logarithmized X.
computing neighbors
    using 'X_pca' with n_pcs = 50
    finished: added to `.uns['neighbors']`
    `.obsp['distances']`, distances for each pair of neighbors
    `.obsp['connectivities']`, weighted adjacency matrix (0:00:01)
computing velocities
    finished (0:00:06) --&gt; added 
    'velocity', velocity vectors for each individual cell (adata.layers)
computing velocity graph (using 1/8 cores)
    finished (0:00:37) --&gt; added 
    'velocity_graph', sparse matrix with cosine correlations (adata.uns)</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>/opt/homebrew/Caskroom/miniconda/base/envs/rvelo_milo/lib/python3.10/site-packages/scvelo/preprocessing/utils.py:705: DeprecationWarning: `log1p` is deprecated since scVelo v0.3.0 and will be removed in a future version. Please use `log1p` from `scanpy.pp` instead.
  log1p(adata)
/opt/homebrew/Caskroom/miniconda/base/envs/rvelo_milo/lib/python3.10/site-packages/scvelo/tools/optimization.py:184: DeprecationWarning: Conversion of an array with ndim &gt; 0 to a scalar is deprecated, and will error in future. Ensure you extract a single element from your array before performing this operation. (Deprecated NumPy 1.25.)
  gamma[i] = np.linalg.pinv(A.T.dot(A)).dot(A.T.dot(y[:, i]))</code></pre>
</div>
</div>
<div id="cell-68" class="cell" data-execution_count="18">
<div class="sourceCode cell-code" id="cb88"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb88-1"><a href="#cb88-1" aria-hidden="true" tabindex="-1"></a>adata_notch_control_3p_moments.uns[<span class="st">"high_res_annotation_colors"</span>] <span class="op">=</span> [<span class="st">'#1f77b4'</span>, </span>
<span id="cb88-2"><a href="#cb88-2" aria-hidden="true" tabindex="-1"></a>                                                               <span class="st">'#ff7f0e'</span>, </span>
<span id="cb88-3"><a href="#cb88-3" aria-hidden="true" tabindex="-1"></a>                                                               <span class="st">'#279e68'</span>, </span>
<span id="cb88-4"><a href="#cb88-4" aria-hidden="true" tabindex="-1"></a>                                                               <span class="st">'#d62728'</span>, </span>
<span id="cb88-5"><a href="#cb88-5" aria-hidden="true" tabindex="-1"></a>                                                               <span class="st">'#aa40fc'</span>, </span>
<span id="cb88-6"><a href="#cb88-6" aria-hidden="true" tabindex="-1"></a>                                                               <span class="st">'#8c564b'</span>, </span>
<span id="cb88-7"><a href="#cb88-7" aria-hidden="true" tabindex="-1"></a>                                                               <span class="st">'#e377c2'</span>,</span>
<span id="cb88-8"><a href="#cb88-8" aria-hidden="true" tabindex="-1"></a>                                                               <span class="st">'#b5bd61'</span>, </span>
<span id="cb88-9"><a href="#cb88-9" aria-hidden="true" tabindex="-1"></a>                                                               <span class="st">'#17becf'</span>, </span>
<span id="cb88-10"><a href="#cb88-10" aria-hidden="true" tabindex="-1"></a>                                                               <span class="st">'#ffbb78'</span>, </span>
<span id="cb88-11"><a href="#cb88-11" aria-hidden="true" tabindex="-1"></a>                                                               <span class="st">'#98df8a'</span>,</span>
<span id="cb88-12"><a href="#cb88-12" aria-hidden="true" tabindex="-1"></a>                                                                <span class="st">'#ff9896'</span>, </span>
<span id="cb88-13"><a href="#cb88-13" aria-hidden="true" tabindex="-1"></a>                                                                <span class="st">'#c5b0d5'</span>, </span>
<span id="cb88-14"><a href="#cb88-14" aria-hidden="true" tabindex="-1"></a>                                                                <span class="st">'#c49c94'</span>, </span>
<span id="cb88-15"><a href="#cb88-15" aria-hidden="true" tabindex="-1"></a>                                                                <span class="st">'#f7b6d2'</span>] </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="cell-69" class="cell" data-execution_count="22">
<div class="sourceCode cell-code" id="cb89"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb89-1"><a href="#cb89-1" aria-hidden="true" tabindex="-1"></a>scv.pl.scatter(adata_notch_control_3p_moments, [<span class="st">"msi"</span>, <span class="st">"cpo"</span>, <span class="st">"trol"</span>, <span class="st">'ed'</span>, <span class="st">"mbl"</span>, <span class="st">"Rbfox1"</span>], color<span class="op">=</span><span class="st">'high_res_annotation'</span>, frameon<span class="op">=</span><span class="va">False</span>, size<span class="op">=</span><span class="dv">15</span>, linewidth<span class="op">=</span><span class="fl">1.5</span>, show<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb89-2"><a href="#cb89-2" aria-hidden="true" tabindex="-1"></a>plt.suptitle(<span class="st">"RNA velocity, split (moments approach), 3' Notch + Control"</span>, y<span class="op">=</span><span class="fl">1.25</span>)</span>
<span id="cb89-3"><a href="#cb89-3" aria-hidden="true" tabindex="-1"></a>plt.show()</span>
<span id="cb89-4"><a href="#cb89-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb89-5"><a href="#cb89-5" aria-hidden="true" tabindex="-1"></a>adata_notch_control_3p_moments.uns[<span class="st">"perturbation_colors"</span>] <span class="op">=</span> [<span class="st">"#1f77b4"</span>, <span class="st">"#d62728"</span>]</span>
<span id="cb89-6"><a href="#cb89-6" aria-hidden="true" tabindex="-1"></a>scv.pl.scatter(adata_notch_control_3p_moments, [<span class="st">"msi"</span>, <span class="st">"cpo"</span>, <span class="st">"trol"</span>, <span class="st">'ed'</span>, <span class="st">"mbl"</span>, <span class="st">"Rbfox1"</span>], color<span class="op">=</span><span class="st">'perturbation'</span>, frameon<span class="op">=</span><span class="va">False</span>, size<span class="op">=</span><span class="dv">15</span>, linewidth<span class="op">=</span><span class="fl">1.5</span>, show<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb89-7"><a href="#cb89-7" aria-hidden="true" tabindex="-1"></a>plt.suptitle(<span class="st">"RNA velocity, split (moments approach), 3' Notch + Control"</span>, y<span class="op">=</span><span class="fl">1.25</span>)</span>
<span id="cb89-8"><a href="#cb89-8" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="main_analyses_files/figure-html/cell-63-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="main_analyses_files/figure-html/cell-63-output-2.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<div id="cell-70" class="cell" data-execution_count="9">
<div class="sourceCode cell-code" id="cb90"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb90-1"><a href="#cb90-1" aria-hidden="true" tabindex="-1"></a><span class="co"># correct the RNA velocity data </span></span>
<span id="cb90-2"><a href="#cb90-2" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> to_hansen(X1, X2):</span>
<span id="cb90-3"><a href="#cb90-3" aria-hidden="true" tabindex="-1"></a>    <span class="co">"""Computes matrices for performing batch effect correction of spliced and unspliced counts.</span></span>
<span id="cb90-4"><a href="#cb90-4" aria-hidden="true" tabindex="-1"></a><span class="co">       idea for batch effect correcting spliced and unspliced counts credited to Hansen lab from thread: https://www.hansenlab.org/velocity_batch</span></span>
<span id="cb90-5"><a href="#cb90-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-6"><a href="#cb90-6" aria-hidden="true" tabindex="-1"></a><span class="co">       M = S + U</span></span>
<span id="cb90-7"><a href="#cb90-7" aria-hidden="true" tabindex="-1"></a><span class="co">       R = S / (S + U)</span></span>
<span id="cb90-8"><a href="#cb90-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-9"><a href="#cb90-9" aria-hidden="true" tabindex="-1"></a><span class="co">       Sb = Mb*R</span></span>
<span id="cb90-10"><a href="#cb90-10" aria-hidden="true" tabindex="-1"></a><span class="co">       Ub = Mb*(1-R)</span></span>
<span id="cb90-11"><a href="#cb90-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-12"><a href="#cb90-12" aria-hidden="true" tabindex="-1"></a><span class="co">    X: list</span></span>
<span id="cb90-13"><a href="#cb90-13" aria-hidden="true" tabindex="-1"></a><span class="co">        list of matrices corresponding to data. Dimensions are 2*n x p</span></span>
<span id="cb90-14"><a href="#cb90-14" aria-hidden="true" tabindex="-1"></a><span class="co">    ----------</span></span>
<span id="cb90-15"><a href="#cb90-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-16"><a href="#cb90-16" aria-hidden="true" tabindex="-1"></a><span class="co">    Returns</span></span>
<span id="cb90-17"><a href="#cb90-17" aria-hidden="true" tabindex="-1"></a><span class="co">    M: ndarray</span></span>
<span id="cb90-18"><a href="#cb90-18" aria-hidden="true" tabindex="-1"></a><span class="co">        matrix of summed spliced and unspliced counts. This is log transformed</span></span>
<span id="cb90-19"><a href="#cb90-19" aria-hidden="true" tabindex="-1"></a><span class="co">    R: ndarray </span></span>
<span id="cb90-20"><a href="#cb90-20" aria-hidden="true" tabindex="-1"></a><span class="co">        matrix to transform data back to original </span></span>
<span id="cb90-21"><a href="#cb90-21" aria-hidden="true" tabindex="-1"></a><span class="co">    ----------</span></span>
<span id="cb90-22"><a href="#cb90-22" aria-hidden="true" tabindex="-1"></a><span class="co">    """</span></span>
<span id="cb90-23"><a href="#cb90-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-24"><a href="#cb90-24" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> scipy.sparse.issparse(X1) <span class="op">==</span> <span class="va">False</span>:</span>
<span id="cb90-25"><a href="#cb90-25" aria-hidden="true" tabindex="-1"></a>            X1 <span class="op">=</span> scipy.sparse.csr_matrix(X1)</span>
<span id="cb90-26"><a href="#cb90-26" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> scipy.sparse.issparse(X2) <span class="op">==</span> <span class="va">False</span>:</span>
<span id="cb90-27"><a href="#cb90-27" aria-hidden="true" tabindex="-1"></a>            X2 <span class="op">=</span> scipy.sparse.csr_matrix(X2)</span>
<span id="cb90-28"><a href="#cb90-28" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb90-29"><a href="#cb90-29" aria-hidden="true" tabindex="-1"></a>    M <span class="op">=</span> X1 <span class="op">+</span> X2</span>
<span id="cb90-30"><a href="#cb90-30" aria-hidden="true" tabindex="-1"></a>    R <span class="op">=</span> X1 <span class="op">/</span> (X1 <span class="op">+</span> X2)</span>
<span id="cb90-31"><a href="#cb90-31" aria-hidden="true" tabindex="-1"></a>    M <span class="op">=</span> np.log1p(M)</span>
<span id="cb90-32"><a href="#cb90-32" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb90-33"><a href="#cb90-33" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> M, R</span>
<span id="cb90-34"><a href="#cb90-34" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb90-35"><a href="#cb90-35" aria-hidden="true" tabindex="-1"></a><span class="co"># make the function to correct the data</span></span>
<span id="cb90-36"><a href="#cb90-36" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> velocity_correct_combat(adata, key<span class="op">=</span><span class="va">None</span>):</span>
<span id="cb90-37"><a href="#cb90-37" aria-hidden="true" tabindex="-1"></a>    <span class="co">"""Correct the RNA velocity data using Combat algorithm from EVi function</span></span>
<span id="cb90-38"><a href="#cb90-38" aria-hidden="true" tabindex="-1"></a><span class="co">    ----------</span></span>
<span id="cb90-39"><a href="#cb90-39" aria-hidden="true" tabindex="-1"></a><span class="co">    adata: AnnData</span></span>
<span id="cb90-40"><a href="#cb90-40" aria-hidden="true" tabindex="-1"></a><span class="co">        AnnData object</span></span>
<span id="cb90-41"><a href="#cb90-41" aria-hidden="true" tabindex="-1"></a><span class="co">        key: str</span></span>
<span id="cb90-42"><a href="#cb90-42" aria-hidden="true" tabindex="-1"></a><span class="co">        key is the batch</span></span>
<span id="cb90-43"><a href="#cb90-43" aria-hidden="true" tabindex="-1"></a><span class="co">        """</span></span>
<span id="cb90-44"><a href="#cb90-44" aria-hidden="true" tabindex="-1"></a>    <span class="co"># first step is to sum the spliced and unspliced counts</span></span>
<span id="cb90-45"><a href="#cb90-45" aria-hidden="true" tabindex="-1"></a>    M, R <span class="op">=</span> to_hansen(adata.layers[<span class="st">"spliced"</span>], adata.layers[<span class="st">"unspliced"</span>])</span>
<span id="cb90-46"><a href="#cb90-46" aria-hidden="true" tabindex="-1"></a>    adata.layers[<span class="st">"M"</span>] <span class="op">=</span> M</span>
<span id="cb90-47"><a href="#cb90-47" aria-hidden="true" tabindex="-1"></a>    adata <span class="op">=</span> adata.copy()</span>
<span id="cb90-48"><a href="#cb90-48" aria-hidden="true" tabindex="-1"></a>    adata.X <span class="op">=</span> adata.layers[<span class="st">"M"</span>]</span>
<span id="cb90-49"><a href="#cb90-49" aria-hidden="true" tabindex="-1"></a>    <span class="co"># correct using Combat</span></span>
<span id="cb90-50"><a href="#cb90-50" aria-hidden="true" tabindex="-1"></a>    sc.pp.combat(adata, key<span class="op">=</span><span class="st">"batch"</span>)</span>
<span id="cb90-51"><a href="#cb90-51" aria-hidden="true" tabindex="-1"></a>    <span class="co"># X1_corr, X2_corr = from_hansen(adata.X, M, R)</span></span>
<span id="cb90-52"><a href="#cb90-52" aria-hidden="true" tabindex="-1"></a>    <span class="co"># adata.X = scipy.sparse.csr_matrix(X1_corr) #reset layers, this is corrected data which is in log space</span></span>
<span id="cb90-53"><a href="#cb90-53" aria-hidden="true" tabindex="-1"></a>    <span class="co"># adata.layers['spliced'] = scipy.sparse.csr_matrix(np.expm1(X1_corr)) #spliced and unspliced matrices are scaled but not log transformed so invert</span></span>
<span id="cb90-54"><a href="#cb90-54" aria-hidden="true" tabindex="-1"></a>    <span class="co"># adata.layers['unspliced'] = scipy.sparse.csr_matrix(np.expm1(X2_corr))</span></span>
<span id="cb90-55"><a href="#cb90-55" aria-hidden="true" tabindex="-1"></a>    adata.layers[<span class="st">"spliced"</span>], adata.layers[<span class="st">"unspliced"</span>] <span class="op">=</span> from_hansen(adata.X, M, R)</span>
<span id="cb90-56"><a href="#cb90-56" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-57"><a href="#cb90-57" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> adata</span>
<span id="cb90-58"><a href="#cb90-58" aria-hidden="true" tabindex="-1"></a>   </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="cell-71" class="cell">
<div class="sourceCode cell-code" id="cb91"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb91-1"><a href="#cb91-1" aria-hidden="true" tabindex="-1"></a><span class="co">### RNA velocity batch effect correction with approach from Hansen Lab</span></span>
<span id="cb91-2"><a href="#cb91-2" aria-hidden="true" tabindex="-1"></a>sc.pp.pca(adata_notch_control_3p)</span>
<span id="cb91-3"><a href="#cb91-3" aria-hidden="true" tabindex="-1"></a>sc.pp.neighbors(adata_notch_control_3p, n_neighbors<span class="op">=</span><span class="dv">15</span>)</span>
<span id="cb91-4"><a href="#cb91-4" aria-hidden="true" tabindex="-1"></a>adata_combat_notch_control_3p <span class="op">=</span> velocity_correct_combat(adata_notch_control_3p, key<span class="op">=</span><span class="st">"batch"</span>)</span>
<span id="cb91-5"><a href="#cb91-5" aria-hidden="true" tabindex="-1"></a>scv.pp.moments(adata_combat_notch_control_3p, n_pcs<span class="op">=</span><span class="dv">30</span>, n_neighbors<span class="op">=</span><span class="dv">30</span>)</span>
<span id="cb91-6"><a href="#cb91-6" aria-hidden="true" tabindex="-1"></a>scv.tl.velocity(adata_combat_notch_control_3p, mode<span class="op">=</span><span class="st">"stochastic"</span>)</span>
<span id="cb91-7"><a href="#cb91-7" aria-hidden="true" tabindex="-1"></a>scv.tl.velocity_graph(adata_combat_notch_control_3p)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="cell-72" class="cell">
<div class="sourceCode cell-code" id="cb92"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb92-1"><a href="#cb92-1" aria-hidden="true" tabindex="-1"></a>scv.pl.scatter(adata_combat_notch_control_3p, [<span class="st">"msi"</span>, <span class="st">"cpo"</span>, <span class="st">"trol"</span>, <span class="st">'ed'</span>, <span class="st">"mbl"</span>, <span class="st">"Rbfox1"</span>], color<span class="op">=</span><span class="st">'high_res_annotation'</span>, frameon<span class="op">=</span><span class="va">False</span>, size<span class="op">=</span><span class="dv">15</span>, linewidth<span class="op">=</span><span class="fl">1.5</span>, show<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb92-2"><a href="#cb92-2" aria-hidden="true" tabindex="-1"></a>plt.suptitle(<span class="st">"RNA velocity, batch correction Combat Hansen approach, 3' Notch + Control"</span>, y<span class="op">=</span><span class="fl">1.25</span>)</span>
<span id="cb92-3"><a href="#cb92-3" aria-hidden="true" tabindex="-1"></a>plt.show()</span>
<span id="cb92-4"><a href="#cb92-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb92-5"><a href="#cb92-5" aria-hidden="true" tabindex="-1"></a>scv.pl.scatter(adata_combat_notch_control_3p, [<span class="st">"msi"</span>, <span class="st">"cpo"</span>, <span class="st">"trol"</span>, <span class="st">'ed'</span>, <span class="st">"mbl"</span>, <span class="st">"Rbfox1"</span>], color<span class="op">=</span><span class="st">'perturbation'</span>, frameon<span class="op">=</span><span class="va">False</span>, size<span class="op">=</span><span class="dv">15</span>, linewidth<span class="op">=</span><span class="fl">1.5</span>, show<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb92-6"><a href="#cb92-6" aria-hidden="true" tabindex="-1"></a>plt.suptitle(<span class="st">"RNA velocity, batch correction Combat Hansen approach, 3' Notch + Control"</span>, y<span class="op">=</span><span class="fl">1.25</span>)</span>
<span id="cb92-7"><a href="#cb92-7" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="main_analyses_files/figure-html/cell-66-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="main_analyses_files/figure-html/cell-66-output-2.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="rna-velocity-recapitulates-trajectories-with-certain-cell-types" class="level1">
<h1>3.7 RNA velocity recapitulates trajectories with certain cell types</h1>
<div id="cell-75" class="cell" data-execution_count="33">
<div class="sourceCode cell-code" id="cb93"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb93-1"><a href="#cb93-1" aria-hidden="true" tabindex="-1"></a><span class="co"># upload the data</span></span>
<span id="cb93-2"><a href="#cb93-2" aria-hidden="true" tabindex="-1"></a><span class="co"># adata = sc.read_h5ad('/g/huber/users/kharuk/DECODE/a_v_data_merged_DECODE_v1.h5ad')</span></span>
<span id="cb93-3"><a href="#cb93-3" aria-hidden="true" tabindex="-1"></a><span class="co"># uplad the files from the local machine</span></span>
<span id="cb93-4"><a href="#cb93-4" aria-hidden="true" tabindex="-1"></a>adata <span class="op">=</span> sc.read_h5ad(<span class="st">"/Users/sviatoslavkharuk/Documents/StudyBioinformatics/Bioinformatics_EMBL/Bioinformatics_project_EMBL/scRNA-seq-RNA-velocity-CRISPR-DECODE/Quarto-notebook-DECODE/data/a_v_datas_merged_DECODE_v1_local_machine.h5ad"</span>)</span>
<span id="cb93-5"><a href="#cb93-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb93-6"><a href="#cb93-6" aria-hidden="true" tabindex="-1"></a><span class="co"># subset the data to the 3' control samples together</span></span>
<span id="cb93-7"><a href="#cb93-7" aria-hidden="true" tabindex="-1"></a>adata_control_3p <span class="op">=</span> adata[((adata.obs[<span class="st">"perturbation"</span>] <span class="op">==</span> <span class="st">"control"</span>) <span class="op">&amp;</span></span>
<span id="cb93-8"><a href="#cb93-8" aria-hidden="true" tabindex="-1"></a>                                ((adata.obs[<span class="st">"sample_name"</span>] <span class="op">==</span> <span class="st">"TX22_A"</span>) <span class="op">|</span></span>
<span id="cb93-9"><a href="#cb93-9" aria-hidden="true" tabindex="-1"></a>                                (adata.obs[<span class="st">"sample_name"</span>] <span class="op">==</span> <span class="st">"TX23_A"</span>)))].copy()</span>
<span id="cb93-10"><a href="#cb93-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb93-11"><a href="#cb93-11" aria-hidden="true" tabindex="-1"></a>adata_control_3p <span class="op">=</span> adata_control_3p[<span class="op">~</span>adata_control_3p.obs[<span class="st">"high_res_annotation"</span>].isin([<span class="st">"Malpighian tubule"</span>])].copy()</span>
<span id="cb93-12"><a href="#cb93-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb93-13"><a href="#cb93-13" aria-hidden="true" tabindex="-1"></a><span class="co">""""fit with ISC instead of EB"""</span></span>
<span id="cb93-14"><a href="#cb93-14" aria-hidden="true" tabindex="-1"></a><span class="co"># it looks better!!! </span></span>
<span id="cb93-15"><a href="#cb93-15" aria-hidden="true" tabindex="-1"></a><span class="co"># fit control samples subsetted data</span></span>
<span id="cb93-16"><a href="#cb93-16" aria-hidden="true" tabindex="-1"></a>adata_control_3p <span class="op">=</span> adata[((adata.obs[<span class="st">"perturbation"</span>] <span class="op">==</span> <span class="st">"control"</span>) <span class="op">&amp;</span></span>
<span id="cb93-17"><a href="#cb93-17" aria-hidden="true" tabindex="-1"></a>                                ((adata.obs[<span class="st">"sample_name"</span>] <span class="op">==</span> <span class="st">"TX22_A"</span>) <span class="op">|</span></span>
<span id="cb93-18"><a href="#cb93-18" aria-hidden="true" tabindex="-1"></a>                                (adata.obs[<span class="st">"sample_name"</span>] <span class="op">==</span> <span class="st">"TX23_A"</span>)))].copy()</span>
<span id="cb93-19"><a href="#cb93-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb93-20"><a href="#cb93-20" aria-hidden="true" tabindex="-1"></a>adata_control_3p <span class="op">=</span> adata_control_3p[<span class="op">~</span>adata_control_3p.obs[<span class="st">"high_res_annotation"</span>].isin([<span class="st">"Malpighian tubule"</span>])].copy()</span>
<span id="cb93-21"><a href="#cb93-21" aria-hidden="true" tabindex="-1"></a>adata_control_3p_filtered <span class="op">=</span> adata_control_3p[adata_control_3p.obs[<span class="st">"high_res_annotation"</span>].isin([<span class="st">"Intestinal stem cells"</span>, <span class="st">"AstC Enteroendocrine cells"</span>, <span class="st">"Enteroendocrine progenitor cell"</span>, <span class="st">"differentiating Enterocytes"</span>, <span class="st">"polyploid Enterocytes"</span>])].copy()</span>
<span id="cb93-22"><a href="#cb93-22" aria-hidden="true" tabindex="-1"></a><span class="co"># velocity_inferring(adata_control_3p_filtered, mode="dynamical", min_shared_counts=20, n_top_genes=500, annotation="high_res_annotation", groups_for_fit=None, condition="control")</span></span>
<span id="cb93-23"><a href="#cb93-23" aria-hidden="true" tabindex="-1"></a>velocity_inferring_local(adata_control_3p_filtered, mode<span class="op">=</span><span class="st">"dynamical"</span>, min_shared_counts<span class="op">=</span><span class="dv">20</span>, n_top_genes<span class="op">=</span><span class="dv">500</span>, annotation<span class="op">=</span><span class="st">"high_res_annotation"</span>, groups_for_fit<span class="op">=</span><span class="va">None</span>, condition<span class="op">=</span><span class="st">"control"</span>)</span>
<span id="cb93-24"><a href="#cb93-24" aria-hidden="true" tabindex="-1"></a><span class="co"># Set plotting style</span></span>
<span id="cb93-25"><a href="#cb93-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb93-26"><a href="#cb93-26" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> matplotlib <span class="im">import</span> rcParams</span>
<span id="cb93-27"><a href="#cb93-27" aria-hidden="true" tabindex="-1"></a><span class="co"># Type 2/TrueType fonts</span></span>
<span id="cb93-28"><a href="#cb93-28" aria-hidden="true" tabindex="-1"></a>rcParams[<span class="st">'pdf.fonttype'</span>] <span class="op">=</span> <span class="dv">42</span></span>
<span id="cb93-29"><a href="#cb93-29" aria-hidden="true" tabindex="-1"></a>rcParams[<span class="st">'ps.fonttype'</span>] <span class="op">=</span> <span class="dv">42</span></span>
<span id="cb93-30"><a href="#cb93-30" aria-hidden="true" tabindex="-1"></a><span class="co"># Arial font</span></span>
<span id="cb93-31"><a href="#cb93-31" aria-hidden="true" tabindex="-1"></a>rcParams[<span class="st">'font.sans-serif'</span>] <span class="op">=</span> <span class="st">"Arial"</span></span>
<span id="cb93-32"><a href="#cb93-32" aria-hidden="true" tabindex="-1"></a>rcParams[<span class="st">'font.family'</span>] <span class="op">=</span> <span class="st">"sans-serif"</span></span>
<span id="cb93-33"><a href="#cb93-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb93-34"><a href="#cb93-34" aria-hidden="true" tabindex="-1"></a><span class="co"># assign the colors to the adata_control_3p</span></span>
<span id="cb93-35"><a href="#cb93-35" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb93-36"><a href="#cb93-36" aria-hidden="true" tabindex="-1"></a><span class="co"># add diffmap to adata object</span></span>
<span id="cb93-37"><a href="#cb93-37" aria-hidden="true" tabindex="-1"></a><span class="co"># diffmap_csv = pd.read_csv("/g/huber/users/kharuk/DECODE/adata_notch_control_3p_diffmap.csv", index_col=0)</span></span>
<span id="cb93-38"><a href="#cb93-38" aria-hidden="true" tabindex="-1"></a><span class="co"># diffmap from the local machine</span></span>
<span id="cb93-39"><a href="#cb93-39" aria-hidden="true" tabindex="-1"></a>diffmap_csv <span class="op">=</span> pd.read_csv(<span class="st">"/Users/sviatoslavkharuk/Documents/StudyBioinformatics/Bioinformatics_EMBL/Bioinformatics_project_EMBL/scRNA-seq-RNA-velocity-CRISPR-DECODE/Quarto-notebook-DECODE/data/adata_notch_control_3p_diffmap.csv"</span>, index_col<span class="op">=</span><span class="dv">0</span>)</span>
<span id="cb93-40"><a href="#cb93-40" aria-hidden="true" tabindex="-1"></a>adata_control_3p.obsm[<span class="st">"X_diffmap"</span>] <span class="op">=</span> diffmap_csv.loc[adata_control_3p.obs_names].values</span>
<span id="cb93-41"><a href="#cb93-41" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb93-42"><a href="#cb93-42" aria-hidden="true" tabindex="-1"></a><span class="co"># plot the streams of subsetted data on entire</span></span>
<span id="cb93-43"><a href="#cb93-43" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> scv.pl.scatter(adata_control_3p, basis<span class="op">=</span><span class="st">"diffmap"</span>, color<span class="op">=</span><span class="st">"high_res_annotation"</span>, legend_loc<span class="op">=</span><span class="st">"right_margin"</span>, legend_fontsize<span class="op">=</span><span class="dv">6</span>, components<span class="op">=</span><span class="st">"1,2"</span>, dpi<span class="op">=</span><span class="dv">300</span>, alpha<span class="op">=</span><span class="fl">0.2</span>, size<span class="op">=</span><span class="dv">35</span>, show<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb93-44"><a href="#cb93-44" aria-hidden="true" tabindex="-1"></a>scv.pl.velocity_embedding_grid(adata_control_3p_filtered, color<span class="op">=</span><span class="st">"high_res_annotation"</span>, basis<span class="op">=</span><span class="st">"diffmap"</span>,components<span class="op">=</span><span class="st">"1,2"</span>, alpha<span class="op">=</span><span class="dv">0</span>, ax<span class="op">=</span>ax, dpi<span class="op">=</span><span class="dv">300</span>, arrow_size<span class="op">=</span><span class="fl">1.25</span>, min_mass<span class="op">=</span><span class="dv">10</span>, title<span class="op">=</span><span class="st">"emb_2_new, alpha=0.2, point_size=35, min_mass=10, arrow_size=1.25 </span><span class="ch">\n</span><span class="st"> control, moments default"</span>)</span>
<span id="cb93-45"><a href="#cb93-45" aria-hidden="true" tabindex="-1"></a>adata_control_3p.uns[<span class="st">"high_res_annotation_colors"</span>]</span>
<span id="cb93-46"><a href="#cb93-46" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb93-47"><a href="#cb93-47" aria-hidden="true" tabindex="-1"></a><span class="co"># add diffmap to adata object</span></span>
<span id="cb93-48"><a href="#cb93-48" aria-hidden="true" tabindex="-1"></a><span class="co"># diffmap_csv = pd.read_csv("/g/huber/users/kharuk/DECODE/adata_notch_control_3p_diffmap.csv", index_col=0)</span></span>
<span id="cb93-49"><a href="#cb93-49" aria-hidden="true" tabindex="-1"></a><span class="co"># diffmap from the local machine</span></span>
<span id="cb93-50"><a href="#cb93-50" aria-hidden="true" tabindex="-1"></a>fig, ax <span class="op">=</span> plt.subplots(figsize<span class="op">=</span>(<span class="dv">8</span>, <span class="dv">6</span>))</span>
<span id="cb93-51"><a href="#cb93-51" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb93-52"><a href="#cb93-52" aria-hidden="true" tabindex="-1"></a>diffmap_csv <span class="op">=</span> pd.read_csv(<span class="st">"/Users/sviatoslavkharuk/Documents/StudyBioinformatics/Bioinformatics_EMBL/Bioinformatics_project_EMBL/scRNA-seq-RNA-velocity-CRISPR-DECODE/Quarto-notebook-DECODE/data/adata_notch_control_3p_diffmap.csv"</span>, index_col<span class="op">=</span><span class="dv">0</span>)</span>
<span id="cb93-53"><a href="#cb93-53" aria-hidden="true" tabindex="-1"></a>adata_control_3p.obsm[<span class="st">"X_diffmap"</span>] <span class="op">=</span> diffmap_csv.loc[adata_control_3p.obs_names].values</span>
<span id="cb93-54"><a href="#cb93-54" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb93-55"><a href="#cb93-55" aria-hidden="true" tabindex="-1"></a><span class="co"># assign the colors to the adata_control_3p</span></span>
<span id="cb93-56"><a href="#cb93-56" aria-hidden="true" tabindex="-1"></a>adata_control_3p.uns[<span class="st">"high_res_annotation_colors"</span>] <span class="op">=</span> [<span class="st">'#1f77b4'</span>, <span class="st">'#ff7f0e'</span>, <span class="st">'#279e68'</span>, <span class="st">'#d62728'</span>, <span class="st">'#aa40fc'</span>, <span class="st">'#8c564b'</span>,</span>
<span id="cb93-57"><a href="#cb93-57" aria-hidden="true" tabindex="-1"></a>                                                      <span class="st">'#b5bd61'</span>, <span class="st">'#17becf'</span>, <span class="st">'#ffbb78'</span>, <span class="st">'#98df8a'</span>, <span class="st">'#ff9896'</span>, <span class="st">'#c5b0d5'</span>,</span>
<span id="cb93-58"><a href="#cb93-58" aria-hidden="true" tabindex="-1"></a>                                                      <span class="st">'#c49c94'</span>, <span class="st">'#f7b6d2'</span>]</span>
<span id="cb93-59"><a href="#cb93-59" aria-hidden="true" tabindex="-1"></a><span class="co"># plot the streams of subsetted data on entire</span></span>
<span id="cb93-60"><a href="#cb93-60" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> scv.pl.scatter(adata_control_3p, basis<span class="op">=</span><span class="st">"diffmap"</span>, color<span class="op">=</span><span class="st">"high_res_annotation"</span>, legend_loc<span class="op">=</span><span class="st">"right_margin"</span>, legend_fontsize<span class="op">=</span><span class="dv">6</span>, components<span class="op">=</span><span class="st">"1,2"</span>, dpi<span class="op">=</span><span class="dv">300</span>, alpha<span class="op">=</span><span class="fl">0.2</span>, size<span class="op">=</span><span class="dv">35</span>, show<span class="op">=</span><span class="va">False</span>, ax<span class="op">=</span>ax)</span>
<span id="cb93-61"><a href="#cb93-61" aria-hidden="true" tabindex="-1"></a>scv.pl.velocity_embedding_grid(adata_control_3p_filtered, color<span class="op">=</span><span class="st">"high_res_annotation"</span>, basis<span class="op">=</span><span class="st">"diffmap"</span>,components<span class="op">=</span><span class="st">"1,2"</span>, alpha<span class="op">=</span><span class="dv">0</span>, ax<span class="op">=</span>ax, dpi<span class="op">=</span><span class="dv">300</span>, arrow_size<span class="op">=</span><span class="fl">1.25</span>, min_mass<span class="op">=</span><span class="dv">10</span>, title<span class="op">=</span><span class="st">"Control"</span>)</span>
<span id="cb93-62"><a href="#cb93-62" aria-hidden="true" tabindex="-1"></a>fig.savefig(<span class="st">"/Users/sviatoslavkharuk/Documents/StudyBioinformatics/Bioinformatics_EMBL/Bioinformatics_project_EMBL/scRNA-seq-RNA-velocity-CRISPR-DECODE/Quarto-notebook-DECODE/data/DECODE_streamlines_control.pdf"</span>, dpi<span class="op">=</span><span class="dv">300</span>, bbox_inches<span class="op">=</span><span class="st">'tight'</span>)</span>
<span id="cb93-63"><a href="#cb93-63" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
 Applying min_shared_counts filter 

Filtered out 10891 genes that are detected 20 counts (shared).
WARNING: Did not normalize X as it looks processed already. To enforce normalization, set `enforce=True`.
Normalized count data: spliced, unspliced.
If you pass `n_top_genes`, all cutoffs are ignored.
extracting highly variable genes
    finished (0:00:00)
--&gt; added
    'highly_variable', boolean vector (adata.var)
    'means', float vector (adata.var)
    'dispersions', float vector (adata.var)
    'dispersions_norm', float vector (adata.var)
WARNING: adata.X seems to be already log-transformed.
'
 Computing PCA, knn and UMAP 
 
computing PCA
    on highly variable genes
    with n_comps=50
    finished (0:00:01)
computing neighbors
    finished: added to `.uns['neighbors']`
    `.obsp['distances']`, distances for each pair of neighbors
    `.obsp['connectivities']`, weighted adjacency matrix (0:00:01)
computing UMAP
    finished: added
    'X_umap', UMAP coordinates (adata.obsm) (0:00:02)

 Computing RNA velocity 
 
computing neighbors
    finished (0:00:00) --&gt; added 
    'distances' and 'connectivities', weighted adjacency matrices (adata.obsp)
computing moments based on connectivities
    finished (0:00:00) --&gt; added 
    'Ms' and 'Mu', moments of un/spliced abundances (adata.layers)

 Applying dynamical mode 

recovering dynamics (using 1/8 cores)
    finished (0:01:01) --&gt; added 
    'fit_pars', fitted parameters for splicing dynamics (adata.var)
computing velocities
    finished (0:00:00) --&gt; added 
    'velocity', velocity vectors for each individual cell (adata.layers)
computing velocity graph (using 1/8 cores)
    finished (0:00:00) --&gt; added 
    'velocity_graph', sparse matrix with cosine correlations (adata.uns)
computing velocity embedding
    finished (0:00:00) --&gt; added
    'velocity_diffmap', embedded velocity vectors (adata.obsm)</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="main_analyses_files/figure-html/cell-68-output-2.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="main_analyses_files/figure-html/cell-68-output-3.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="main_analyses_files/figure-html/cell-68-output-4.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="main_analyses_files/figure-html/cell-68-output-5.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="main_analyses_files/figure-html/cell-68-output-6.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<div id="cell-76" class="cell" data-execution_count="34">
<div class="sourceCode cell-code" id="cb95"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb95-1"><a href="#cb95-1" aria-hidden="true" tabindex="-1"></a><span class="co"># changing the size of the arrows</span></span>
<span id="cb95-2"><a href="#cb95-2" aria-hidden="true" tabindex="-1"></a>fig, ax <span class="op">=</span> plt.subplots(figsize<span class="op">=</span>(<span class="dv">8</span>, <span class="dv">6</span>))</span>
<span id="cb95-3"><a href="#cb95-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb95-4"><a href="#cb95-4" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> scv.pl.scatter(adata_control_3p, basis<span class="op">=</span><span class="st">"diffmap"</span>, color<span class="op">=</span><span class="st">"high_res_annotation"</span>, legend_loc<span class="op">=</span><span class="st">"right_margin"</span>, legend_fontsize<span class="op">=</span><span class="dv">6</span>, components<span class="op">=</span><span class="st">"1,2"</span>, dpi<span class="op">=</span><span class="dv">300</span>, alpha<span class="op">=</span><span class="fl">0.2</span>, size<span class="op">=</span><span class="dv">35</span>, show<span class="op">=</span><span class="va">False</span>, ax<span class="op">=</span>ax)</span>
<span id="cb95-5"><a href="#cb95-5" aria-hidden="true" tabindex="-1"></a>scv.pl.velocity_embedding_grid(adata_control_3p_filtered, color<span class="op">=</span><span class="st">"high_res_annotation"</span>, basis<span class="op">=</span><span class="st">"diffmap"</span>,components<span class="op">=</span><span class="st">"1,2"</span>, alpha<span class="op">=</span><span class="dv">0</span>, ax<span class="op">=</span>ax, dpi<span class="op">=</span><span class="dv">300</span>, arrow_size<span class="op">=</span><span class="fl">2.5</span>, min_mass<span class="op">=</span><span class="dv">10</span>,title<span class="op">=</span><span class="st">"Control"</span>,)</span>
<span id="cb95-6"><a href="#cb95-6" aria-hidden="true" tabindex="-1"></a>fig.savefig(<span class="st">"/Users/sviatoslavkharuk/Documents/StudyBioinformatics/Bioinformatics_EMBL/Bioinformatics_project_EMBL/scRNA-seq-RNA-velocity-CRISPR-DECODE/Quarto-notebook-DECODE/figures/DECODE_streamlines_control_correct.pdf"</span>, dpi<span class="op">=</span><span class="dv">300</span>, bbox_inches<span class="op">=</span><span class="st">'tight'</span>)</span>
<span id="cb95-7"><a href="#cb95-7" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="main_analyses_files/figure-html/cell-69-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<div id="cell-78" class="cell" data-execution_count="35">
<div class="sourceCode cell-code" id="cb96"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb96-1"><a href="#cb96-1" aria-hidden="true" tabindex="-1"></a><span class="co">"""Notch KO"""</span></span>
<span id="cb96-2"><a href="#cb96-2" aria-hidden="true" tabindex="-1"></a><span class="co">"""Fit with ISC instead of EB"""</span></span>
<span id="cb96-3"><a href="#cb96-3" aria-hidden="true" tabindex="-1"></a><span class="co"># now subset the cells for the fitting and for Notch condition</span></span>
<span id="cb96-4"><a href="#cb96-4" aria-hidden="true" tabindex="-1"></a>adata_notch_3p <span class="op">=</span> adata[((adata.obs[<span class="st">"perturbation"</span>] <span class="op">==</span> <span class="st">"Notch gRNA"</span>) <span class="op">&amp;</span></span>
<span id="cb96-5"><a href="#cb96-5" aria-hidden="true" tabindex="-1"></a>                                ((adata.obs[<span class="st">"sample_name"</span>] <span class="op">==</span> <span class="st">"TX22_D"</span>) <span class="op">|</span></span>
<span id="cb96-6"><a href="#cb96-6" aria-hidden="true" tabindex="-1"></a>                                (adata.obs[<span class="st">"sample_name"</span>] <span class="op">==</span> <span class="st">"TX23_D"</span>)))].copy()</span>
<span id="cb96-7"><a href="#cb96-7" aria-hidden="true" tabindex="-1"></a><span class="co"># add diffmap to adata object</span></span>
<span id="cb96-8"><a href="#cb96-8" aria-hidden="true" tabindex="-1"></a><span class="co"># diffmap_csv = pd.read_csv("/g/huber/users/kharuk/DECODE/adata_notch_control_3p_diffmap.csv", index_col=0)</span></span>
<span id="cb96-9"><a href="#cb96-9" aria-hidden="true" tabindex="-1"></a>diffmap_csv <span class="op">=</span> pd.read_csv(<span class="st">"/Users/sviatoslavkharuk/Documents/StudyBioinformatics/Bioinformatics_EMBL/Bioinformatics_project_EMBL/scRNA-seq-RNA-velocity-CRISPR-DECODE/Quarto-notebook-DECODE/data/adata_notch_control_3p_diffmap.csv"</span>, index_col<span class="op">=</span><span class="dv">0</span>)</span>
<span id="cb96-10"><a href="#cb96-10" aria-hidden="true" tabindex="-1"></a>adata_notch_3p.obsm[<span class="st">"X_diffmap"</span>] <span class="op">=</span> diffmap_csv.loc[adata_notch_3p.obs_names].values</span>
<span id="cb96-11"><a href="#cb96-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb96-12"><a href="#cb96-12" aria-hidden="true" tabindex="-1"></a><span class="co"># keep these cell types for inferring</span></span>
<span id="cb96-13"><a href="#cb96-13" aria-hidden="true" tabindex="-1"></a>adata_notch_3p_filtered <span class="op">=</span> adata_notch_3p[adata_notch_3p.obs[<span class="st">"high_res_annotation"</span>].isin([<span class="st">"Intestinal stem cells"</span>, <span class="st">"AstC Enteroendocrine cells"</span>, <span class="st">"Enteroendocrine progenitor cell"</span>, <span class="st">"differentiating Enterocytes"</span>, <span class="st">"polyploid Enterocytes"</span>])].copy()</span>
<span id="cb96-14"><a href="#cb96-14" aria-hidden="true" tabindex="-1"></a><span class="co"># velocity_inferring(adata_notch_3p_filtered, mode="dynamical", min_shared_counts=20, n_top_genes=500, annotation="high_res_annotation", groups_for_fit=None, condition="Notch")</span></span>
<span id="cb96-15"><a href="#cb96-15" aria-hidden="true" tabindex="-1"></a>velocity_inferring_local(adata_notch_3p_filtered, mode<span class="op">=</span><span class="st">"dynamical"</span>, min_shared_counts<span class="op">=</span><span class="dv">20</span>, n_top_genes<span class="op">=</span><span class="dv">500</span>, annotation<span class="op">=</span><span class="st">"high_res_annotation"</span>, groups_for_fit<span class="op">=</span><span class="va">None</span>, condition<span class="op">=</span><span class="st">"Notch"</span>)</span>
<span id="cb96-16"><a href="#cb96-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb96-17"><a href="#cb96-17" aria-hidden="true" tabindex="-1"></a><span class="co"># plot streams on the entire embedding</span></span>
<span id="cb96-18"><a href="#cb96-18" aria-hidden="true" tabindex="-1"></a>adata_notch_3p <span class="op">=</span> adata_notch_3p[<span class="op">~</span>adata_notch_3p.obs[<span class="st">"high_res_annotation"</span>].isin([<span class="st">"Malpighian tubule"</span>])].copy()</span>
<span id="cb96-19"><a href="#cb96-19" aria-hidden="true" tabindex="-1"></a>adata_notch_3p.uns[<span class="st">"high_res_annotation_colors"</span>] <span class="op">=</span> [<span class="st">'#1f77b4'</span>, <span class="st">'#ff7f0e'</span>, <span class="st">'#279e68'</span>, <span class="st">'#d62728'</span>, <span class="st">'#aa40fc'</span>, <span class="st">'#8c564b'</span>, <span class="st">'#b5bd61'</span>, <span class="st">'#17becf'</span>, <span class="st">'#ffbb78'</span>, <span class="st">'#98df8a'</span>, <span class="st">'#c5b0d5'</span>, <span class="st">'#c49c94'</span>, <span class="st">'#f7b6d2'</span>]</span>
<span id="cb96-20"><a href="#cb96-20" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> scv.pl.scatter(adata_notch_3p, basis<span class="op">=</span><span class="st">"diffmap"</span>, color<span class="op">=</span><span class="st">"high_res_annotation"</span>, legend_loc<span class="op">=</span><span class="st">"right_margin"</span>, legend_fontsize<span class="op">=</span><span class="dv">6</span>, components<span class="op">=</span><span class="st">"1,2"</span>, dpi<span class="op">=</span><span class="dv">300</span>, alpha<span class="op">=</span><span class="fl">0.2</span>, size<span class="op">=</span><span class="dv">35</span>, show<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb96-21"><a href="#cb96-21" aria-hidden="true" tabindex="-1"></a>scv.pl.velocity_embedding_grid(adata_notch_3p_filtered, color<span class="op">=</span><span class="st">"high_res_annotation"</span>, basis<span class="op">=</span><span class="st">"diffmap"</span>,components<span class="op">=</span><span class="st">"1,2"</span>, alpha<span class="op">=</span><span class="dv">0</span>, ax<span class="op">=</span>ax, dpi<span class="op">=</span><span class="dv">300</span>, arrow_size<span class="op">=</span><span class="fl">1.25</span>, min_mass<span class="op">=</span><span class="dv">10</span>, title<span class="op">=</span><span class="st">"emb_2_new, alpha=0.2, point_size=35, min_mass=10, arrow_size=1.25 </span><span class="ch">\n</span><span class="st"> Notch KO, moments default"</span>)</span>
<span id="cb96-22"><a href="#cb96-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb96-23"><a href="#cb96-23" aria-hidden="true" tabindex="-1"></a>fig, ax <span class="op">=</span> plt.subplots(figsize<span class="op">=</span>(<span class="dv">8</span>, <span class="dv">6</span>))</span>
<span id="cb96-24"><a href="#cb96-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb96-25"><a href="#cb96-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb96-26"><a href="#cb96-26" aria-hidden="true" tabindex="-1"></a><span class="co"># assign the colors to the adata_notch_3p</span></span>
<span id="cb96-27"><a href="#cb96-27" aria-hidden="true" tabindex="-1"></a>adata_notch_3p.uns[<span class="st">"high_res_annotation_colors"</span>] <span class="op">=</span> [<span class="st">'#1f77b4'</span>, <span class="st">'#ff7f0e'</span>, <span class="st">'#279e68'</span>, <span class="st">'#d62728'</span>, <span class="st">'#aa40fc'</span>, <span class="st">'#8c564b'</span>, <span class="st">'#b5bd61'</span>, <span class="st">'#17becf'</span>, <span class="st">'#ffbb78'</span>, <span class="st">'#98df8a'</span>, <span class="st">'#c5b0d5'</span>, <span class="st">'#c49c94'</span>, <span class="st">'#f7b6d2'</span>]</span>
<span id="cb96-28"><a href="#cb96-28" aria-hidden="true" tabindex="-1"></a>adata_notch_3p <span class="op">=</span> adata_notch_3p[<span class="op">~</span>adata_notch_3p.obs[<span class="st">"high_res_annotation"</span>].isin([<span class="st">"Malpighian tubule"</span>])].copy()</span>
<span id="cb96-29"><a href="#cb96-29" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> scv.pl.scatter(adata_notch_3p, basis<span class="op">=</span><span class="st">"diffmap"</span>, color<span class="op">=</span><span class="st">"high_res_annotation"</span>, legend_loc<span class="op">=</span><span class="st">"right_margin"</span>, legend_fontsize<span class="op">=</span><span class="dv">6</span>, components<span class="op">=</span><span class="st">"1,2"</span>, dpi<span class="op">=</span><span class="dv">300</span>, alpha<span class="op">=</span><span class="fl">0.2</span>, size<span class="op">=</span><span class="dv">35</span>, show<span class="op">=</span><span class="va">False</span>, ax<span class="op">=</span>ax)</span>
<span id="cb96-30"><a href="#cb96-30" aria-hidden="true" tabindex="-1"></a>scv.pl.velocity_embedding_grid(adata_notch_3p_filtered, color<span class="op">=</span><span class="st">"high_res_annotation"</span>, basis<span class="op">=</span><span class="st">"diffmap"</span>,components<span class="op">=</span><span class="st">"1,2"</span>, alpha<span class="op">=</span><span class="dv">0</span>, ax<span class="op">=</span>ax, dpi<span class="op">=</span><span class="dv">300</span>, arrow_size<span class="op">=</span><span class="fl">1.25</span>, min_mass<span class="op">=</span><span class="dv">10</span>, title<span class="op">=</span><span class="st">"Notch KO"</span>)</span>
<span id="cb96-31"><a href="#cb96-31" aria-hidden="true" tabindex="-1"></a>fig.savefig(<span class="st">"/Users/sviatoslavkharuk/Documents/StudyBioinformatics/Bioinformatics_EMBL/Bioinformatics_project_EMBL/scRNA-seq-RNA-velocity-CRISPR-DECODE/Quarto-notebook-DECODE/data/DECODE_streamlines_notch.pdf"</span>, dpi<span class="op">=</span><span class="dv">300</span>, bbox_inches<span class="op">=</span><span class="st">'tight'</span>)</span>
<span id="cb96-32"><a href="#cb96-32" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
 Applying min_shared_counts filter 

Filtered out 9788 genes that are detected 20 counts (shared).
WARNING: Did not normalize X as it looks processed already. To enforce normalization, set `enforce=True`.
Normalized count data: spliced, unspliced.
If you pass `n_top_genes`, all cutoffs are ignored.
extracting highly variable genes
    finished (0:00:00)
--&gt; added
    'highly_variable', boolean vector (adata.var)
    'means', float vector (adata.var)
    'dispersions', float vector (adata.var)
    'dispersions_norm', float vector (adata.var)
WARNING: adata.X seems to be already log-transformed.
'
 Computing PCA, knn and UMAP 
 
computing PCA
    on highly variable genes
    with n_comps=50
    finished (0:00:01)
computing neighbors
    finished: added to `.uns['neighbors']`
    `.obsp['distances']`, distances for each pair of neighbors
    `.obsp['connectivities']`, weighted adjacency matrix (0:00:00)
computing UMAP
    finished: added
    'X_umap', UMAP coordinates (adata.obsm) (0:00:06)

 Computing RNA velocity 
 
computing neighbors
    finished (0:00:01) --&gt; added 
    'distances' and 'connectivities', weighted adjacency matrices (adata.obsp)
computing moments based on connectivities
    finished (0:00:01) --&gt; added 
    'Ms' and 'Mu', moments of un/spliced abundances (adata.layers)

 Applying dynamical mode 

recovering dynamics (using 1/8 cores)
    finished (0:04:24) --&gt; added 
    'fit_pars', fitted parameters for splicing dynamics (adata.var)
computing velocities
    finished (0:00:01) --&gt; added 
    'velocity', velocity vectors for each individual cell (adata.layers)
computing velocity graph (using 1/8 cores)
    finished (0:00:03) --&gt; added 
    'velocity_graph', sparse matrix with cosine correlations (adata.uns)
computing velocity embedding
    finished (0:00:01) --&gt; added
    'velocity_diffmap', embedded velocity vectors (adata.obsm)</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="main_analyses_files/figure-html/cell-71-output-2.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="main_analyses_files/figure-html/cell-71-output-3.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="main_analyses_files/figure-html/cell-71-output-4.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="main_analyses_files/figure-html/cell-71-output-5.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="main_analyses_files/figure-html/cell-71-output-6.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<div id="cell-79" class="cell" data-execution_count="36">
<div class="sourceCode cell-code" id="cb98"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb98-1"><a href="#cb98-1" aria-hidden="true" tabindex="-1"></a>fig, ax <span class="op">=</span> plt.subplots(figsize<span class="op">=</span>(<span class="dv">8</span>, <span class="dv">6</span>))</span>
<span id="cb98-2"><a href="#cb98-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb98-3"><a href="#cb98-3" aria-hidden="true" tabindex="-1"></a><span class="co"># assign the colors to the adata_notch_3p</span></span>
<span id="cb98-4"><a href="#cb98-4" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> scv.pl.scatter(adata_notch_3p, basis<span class="op">=</span><span class="st">"diffmap"</span>, color<span class="op">=</span><span class="st">"high_res_annotation"</span>, legend_loc<span class="op">=</span><span class="st">"right_margin"</span>, legend_fontsize<span class="op">=</span><span class="dv">6</span>, components<span class="op">=</span><span class="st">"1,2"</span>, dpi<span class="op">=</span><span class="dv">300</span>, alpha<span class="op">=</span><span class="fl">0.2</span>, size<span class="op">=</span><span class="dv">35</span>, show<span class="op">=</span><span class="va">False</span>, ax<span class="op">=</span>ax)</span>
<span id="cb98-5"><a href="#cb98-5" aria-hidden="true" tabindex="-1"></a>scv.pl.velocity_embedding_grid(adata_notch_3p_filtered, color<span class="op">=</span><span class="st">"high_res_annotation"</span>, basis<span class="op">=</span><span class="st">"diffmap"</span>,components<span class="op">=</span><span class="st">"1,2"</span>, alpha<span class="op">=</span><span class="dv">0</span>, ax<span class="op">=</span>ax, dpi<span class="op">=</span><span class="dv">300</span>, arrow_size<span class="op">=</span><span class="fl">2.5</span>, min_mass<span class="op">=</span><span class="dv">10</span>, title<span class="op">=</span><span class="st">"Notch KO"</span>)</span>
<span id="cb98-6"><a href="#cb98-6" aria-hidden="true" tabindex="-1"></a>fig.savefig(<span class="st">"/Users/sviatoslavkharuk/Documents/StudyBioinformatics/Bioinformatics_EMBL/Bioinformatics_project_EMBL/scRNA-seq-RNA-velocity-CRISPR-DECODE/Quarto-notebook-DECODE/figures/DECODE_streamlines_notch_correct.pdf"</span>, dpi<span class="op">=</span><span class="dv">300</span>, bbox_inches<span class="op">=</span><span class="st">'tight'</span>)</span>
<span id="cb98-7"><a href="#cb98-7" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="main_analyses_files/figure-html/cell-72-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<section id="rna-velocity-does-not-recapitulate-trajectories-with-all-cell-types" class="level4">
<h4 class="anchored" data-anchor-id="rna-velocity-does-not-recapitulate-trajectories-with-all-cell-types">RNA velocity does not recapitulate trajectories with all cell types</h4>
<div id="cell-81" class="cell" data-execution_count="37">
<div class="sourceCode cell-code" id="cb99"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb99-1"><a href="#cb99-1" aria-hidden="true" tabindex="-1"></a><span class="co"># upload the data</span></span>
<span id="cb99-2"><a href="#cb99-2" aria-hidden="true" tabindex="-1"></a><span class="co"># adata = sc.read_h5ad('/g/huber/users/kharuk/DECODE/a_v_data_merged_DECODE_v1.h5ad')</span></span>
<span id="cb99-3"><a href="#cb99-3" aria-hidden="true" tabindex="-1"></a><span class="co"># uplad the files from the local machine</span></span>
<span id="cb99-4"><a href="#cb99-4" aria-hidden="true" tabindex="-1"></a>adata <span class="op">=</span> sc.read_h5ad(<span class="st">"/Users/sviatoslavkharuk/Documents/StudyBioinformatics/Bioinformatics_EMBL/Bioinformatics_project_EMBL/scRNA-seq-RNA-velocity-CRISPR-DECODE/Quarto-notebook-DECODE/data/a_v_datas_merged_DECODE_v1_local_machine.h5ad"</span>)</span>
<span id="cb99-5"><a href="#cb99-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb99-6"><a href="#cb99-6" aria-hidden="true" tabindex="-1"></a><span class="co"># subset the data to the 3' control samples together</span></span>
<span id="cb99-7"><a href="#cb99-7" aria-hidden="true" tabindex="-1"></a>adata_control_3p <span class="op">=</span> adata[((adata.obs[<span class="st">"perturbation"</span>] <span class="op">==</span> <span class="st">"control"</span>) <span class="op">&amp;</span></span>
<span id="cb99-8"><a href="#cb99-8" aria-hidden="true" tabindex="-1"></a>                                ((adata.obs[<span class="st">"sample_name"</span>] <span class="op">==</span> <span class="st">"TX22_A"</span>) <span class="op">|</span></span>
<span id="cb99-9"><a href="#cb99-9" aria-hidden="true" tabindex="-1"></a>                                (adata.obs[<span class="st">"sample_name"</span>] <span class="op">==</span> <span class="st">"TX23_A"</span>)))].copy()</span>
<span id="cb99-10"><a href="#cb99-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb99-11"><a href="#cb99-11" aria-hidden="true" tabindex="-1"></a>adata_control_3p <span class="op">=</span> adata_control_3p[<span class="op">~</span>adata_control_3p.obs[<span class="st">"high_res_annotation"</span>].isin([<span class="st">"Malpighian tubule"</span>])].copy()</span>
<span id="cb99-12"><a href="#cb99-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb99-13"><a href="#cb99-13" aria-hidden="true" tabindex="-1"></a><span class="co">""""fit with ISC instead of EB"""</span></span>
<span id="cb99-14"><a href="#cb99-14" aria-hidden="true" tabindex="-1"></a><span class="co"># it looks better!!! </span></span>
<span id="cb99-15"><a href="#cb99-15" aria-hidden="true" tabindex="-1"></a><span class="co"># fit control samples subsetted data</span></span>
<span id="cb99-16"><a href="#cb99-16" aria-hidden="true" tabindex="-1"></a>adata_control_3p <span class="op">=</span> adata[((adata.obs[<span class="st">"perturbation"</span>] <span class="op">==</span> <span class="st">"control"</span>) <span class="op">&amp;</span></span>
<span id="cb99-17"><a href="#cb99-17" aria-hidden="true" tabindex="-1"></a>                                ((adata.obs[<span class="st">"sample_name"</span>] <span class="op">==</span> <span class="st">"TX22_A"</span>) <span class="op">|</span></span>
<span id="cb99-18"><a href="#cb99-18" aria-hidden="true" tabindex="-1"></a>                                (adata.obs[<span class="st">"sample_name"</span>] <span class="op">==</span> <span class="st">"TX23_A"</span>)))].copy()</span>
<span id="cb99-19"><a href="#cb99-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb99-20"><a href="#cb99-20" aria-hidden="true" tabindex="-1"></a>adata_control_3p <span class="op">=</span> adata_control_3p[<span class="op">~</span>adata_control_3p.obs[<span class="st">"high_res_annotation"</span>].isin([<span class="st">"Malpighian tubule"</span>])].copy()</span>
<span id="cb99-21"><a href="#cb99-21" aria-hidden="true" tabindex="-1"></a><span class="co"># velocity_inferring(adata_control_3p_filtered, mode="dynamical", min_shared_counts=20, n_top_genes=500, annotation="high_res_annotation", groups_for_fit=None, condition="control")</span></span>
<span id="cb99-22"><a href="#cb99-22" aria-hidden="true" tabindex="-1"></a>velocity_inferring_local(adata_control_3p, mode<span class="op">=</span><span class="st">"dynamical"</span>, min_shared_counts<span class="op">=</span><span class="dv">20</span>, n_top_genes<span class="op">=</span><span class="dv">500</span>, annotation<span class="op">=</span><span class="st">"high_res_annotation"</span>, groups_for_fit<span class="op">=</span><span class="va">None</span>, condition<span class="op">=</span><span class="st">"control"</span>)</span>
<span id="cb99-23"><a href="#cb99-23" aria-hidden="true" tabindex="-1"></a><span class="co"># Set plotting style</span></span>
<span id="cb99-24"><a href="#cb99-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb99-25"><a href="#cb99-25" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> matplotlib <span class="im">import</span> rcParams</span>
<span id="cb99-26"><a href="#cb99-26" aria-hidden="true" tabindex="-1"></a><span class="co"># Type 2/TrueType fonts</span></span>
<span id="cb99-27"><a href="#cb99-27" aria-hidden="true" tabindex="-1"></a>rcParams[<span class="st">'pdf.fonttype'</span>] <span class="op">=</span> <span class="dv">42</span></span>
<span id="cb99-28"><a href="#cb99-28" aria-hidden="true" tabindex="-1"></a>rcParams[<span class="st">'ps.fonttype'</span>] <span class="op">=</span> <span class="dv">42</span></span>
<span id="cb99-29"><a href="#cb99-29" aria-hidden="true" tabindex="-1"></a><span class="co"># Arial font</span></span>
<span id="cb99-30"><a href="#cb99-30" aria-hidden="true" tabindex="-1"></a>rcParams[<span class="st">'font.sans-serif'</span>] <span class="op">=</span> <span class="st">"Arial"</span></span>
<span id="cb99-31"><a href="#cb99-31" aria-hidden="true" tabindex="-1"></a>rcParams[<span class="st">'font.family'</span>] <span class="op">=</span> <span class="st">"sans-serif"</span></span>
<span id="cb99-32"><a href="#cb99-32" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb99-33"><a href="#cb99-33" aria-hidden="true" tabindex="-1"></a><span class="co"># assign the colors to the adata_control_3p</span></span>
<span id="cb99-34"><a href="#cb99-34" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb99-35"><a href="#cb99-35" aria-hidden="true" tabindex="-1"></a><span class="co"># add diffmap to adata object</span></span>
<span id="cb99-36"><a href="#cb99-36" aria-hidden="true" tabindex="-1"></a><span class="co"># diffmap_csv = pd.read_csv("/g/huber/users/kharuk/DECODE/adata_notch_control_3p_diffmap.csv", index_col=0)</span></span>
<span id="cb99-37"><a href="#cb99-37" aria-hidden="true" tabindex="-1"></a><span class="co"># diffmap from the local machine</span></span>
<span id="cb99-38"><a href="#cb99-38" aria-hidden="true" tabindex="-1"></a>diffmap_csv <span class="op">=</span> pd.read_csv(<span class="st">"/Users/sviatoslavkharuk/Documents/StudyBioinformatics/Bioinformatics_EMBL/Bioinformatics_project_EMBL/scRNA-seq-RNA-velocity-CRISPR-DECODE/Quarto-notebook-DECODE/data/adata_notch_control_3p_diffmap.csv"</span>, index_col<span class="op">=</span><span class="dv">0</span>)</span>
<span id="cb99-39"><a href="#cb99-39" aria-hidden="true" tabindex="-1"></a>adata_control_3p.obsm[<span class="st">"X_diffmap"</span>] <span class="op">=</span> diffmap_csv.loc[adata_control_3p.obs_names].values</span>
<span id="cb99-40"><a href="#cb99-40" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb99-41"><a href="#cb99-41" aria-hidden="true" tabindex="-1"></a><span class="co"># plot the streams of subsetted data on entire</span></span>
<span id="cb99-42"><a href="#cb99-42" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> scv.pl.scatter(adata_control_3p, basis<span class="op">=</span><span class="st">"diffmap"</span>, color<span class="op">=</span><span class="st">"high_res_annotation"</span>, legend_loc<span class="op">=</span><span class="st">"right_margin"</span>, legend_fontsize<span class="op">=</span><span class="dv">6</span>, components<span class="op">=</span><span class="st">"1,2"</span>, dpi<span class="op">=</span><span class="dv">300</span>, alpha<span class="op">=</span><span class="fl">0.2</span>, size<span class="op">=</span><span class="dv">35</span>, show<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb99-43"><a href="#cb99-43" aria-hidden="true" tabindex="-1"></a>scv.pl.velocity_embedding_grid(adata_control_3p, color<span class="op">=</span><span class="st">"high_res_annotation"</span>, basis<span class="op">=</span><span class="st">"diffmap"</span>,components<span class="op">=</span><span class="st">"1,2"</span>, alpha<span class="op">=</span><span class="dv">0</span>, ax<span class="op">=</span>ax, dpi<span class="op">=</span><span class="dv">300</span>, arrow_size<span class="op">=</span><span class="fl">1.25</span>, min_mass<span class="op">=</span><span class="dv">10</span>, title<span class="op">=</span><span class="st">"emb_2_new, alpha=0.2, point_size=35, min_mass=10, arrow_size=1.25 </span><span class="ch">\n</span><span class="st"> control, moments default"</span>)</span>
<span id="cb99-44"><a href="#cb99-44" aria-hidden="true" tabindex="-1"></a>adata_control_3p.uns[<span class="st">"high_res_annotation_colors"</span>]</span>
<span id="cb99-45"><a href="#cb99-45" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb99-46"><a href="#cb99-46" aria-hidden="true" tabindex="-1"></a><span class="co"># add diffmap to adata object</span></span>
<span id="cb99-47"><a href="#cb99-47" aria-hidden="true" tabindex="-1"></a><span class="co"># diffmap_csv = pd.read_csv("/g/huber/users/kharuk/DECODE/adata_notch_control_3p_diffmap.csv", index_col=0)</span></span>
<span id="cb99-48"><a href="#cb99-48" aria-hidden="true" tabindex="-1"></a><span class="co"># diffmap from the local machine</span></span>
<span id="cb99-49"><a href="#cb99-49" aria-hidden="true" tabindex="-1"></a>fig, ax <span class="op">=</span> plt.subplots(figsize<span class="op">=</span>(<span class="dv">8</span>, <span class="dv">6</span>))</span>
<span id="cb99-50"><a href="#cb99-50" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb99-51"><a href="#cb99-51" aria-hidden="true" tabindex="-1"></a>diffmap_csv <span class="op">=</span> pd.read_csv(<span class="st">"/Users/sviatoslavkharuk/Documents/StudyBioinformatics/Bioinformatics_EMBL/Bioinformatics_project_EMBL/scRNA-seq-RNA-velocity-CRISPR-DECODE/Quarto-notebook-DECODE/data/adata_notch_control_3p_diffmap.csv"</span>, index_col<span class="op">=</span><span class="dv">0</span>)</span>
<span id="cb99-52"><a href="#cb99-52" aria-hidden="true" tabindex="-1"></a>adata_control_3p.obsm[<span class="st">"X_diffmap"</span>] <span class="op">=</span> diffmap_csv.loc[adata_control_3p.obs_names].values</span>
<span id="cb99-53"><a href="#cb99-53" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb99-54"><a href="#cb99-54" aria-hidden="true" tabindex="-1"></a><span class="co"># assign the colors to the adata_control_3p</span></span>
<span id="cb99-55"><a href="#cb99-55" aria-hidden="true" tabindex="-1"></a>adata_control_3p.uns[<span class="st">"high_res_annotation_colors"</span>] <span class="op">=</span> [<span class="st">'#1f77b4'</span>, <span class="st">'#ff7f0e'</span>, <span class="st">'#279e68'</span>, <span class="st">'#d62728'</span>, <span class="st">'#aa40fc'</span>, <span class="st">'#8c564b'</span>,</span>
<span id="cb99-56"><a href="#cb99-56" aria-hidden="true" tabindex="-1"></a>                                                      <span class="st">'#b5bd61'</span>, <span class="st">'#17becf'</span>, <span class="st">'#ffbb78'</span>, <span class="st">'#98df8a'</span>, <span class="st">'#ff9896'</span>, <span class="st">'#c5b0d5'</span>,</span>
<span id="cb99-57"><a href="#cb99-57" aria-hidden="true" tabindex="-1"></a>                                                      <span class="st">'#c49c94'</span>, <span class="st">'#f7b6d2'</span>]</span>
<span id="cb99-58"><a href="#cb99-58" aria-hidden="true" tabindex="-1"></a><span class="co"># plot the streams of subsetted data on entire</span></span>
<span id="cb99-59"><a href="#cb99-59" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> scv.pl.scatter(adata_control_3p, basis<span class="op">=</span><span class="st">"diffmap"</span>, color<span class="op">=</span><span class="st">"high_res_annotation"</span>, legend_loc<span class="op">=</span><span class="st">"right_margin"</span>, legend_fontsize<span class="op">=</span><span class="dv">6</span>, components<span class="op">=</span><span class="st">"1,2"</span>, dpi<span class="op">=</span><span class="dv">300</span>, alpha<span class="op">=</span><span class="fl">0.2</span>, size<span class="op">=</span><span class="dv">35</span>, show<span class="op">=</span><span class="va">False</span>, ax<span class="op">=</span>ax)</span>
<span id="cb99-60"><a href="#cb99-60" aria-hidden="true" tabindex="-1"></a>scv.pl.velocity_embedding_grid(adata_control_3p, color<span class="op">=</span><span class="st">"high_res_annotation"</span>, basis<span class="op">=</span><span class="st">"diffmap"</span>,components<span class="op">=</span><span class="st">"1,2"</span>, alpha<span class="op">=</span><span class="dv">0</span>, ax<span class="op">=</span>ax, dpi<span class="op">=</span><span class="dv">300</span>, arrow_size<span class="op">=</span><span class="fl">1.25</span>, min_mass<span class="op">=</span><span class="dv">10</span>, title<span class="op">=</span><span class="st">"Control, all cell types"</span>)</span>
<span id="cb99-61"><a href="#cb99-61" aria-hidden="true" tabindex="-1"></a>fig.savefig(<span class="st">"/Users/sviatoslavkharuk/Documents/StudyBioinformatics/Bioinformatics_EMBL/Bioinformatics_project_EMBL/scRNA-seq-RNA-velocity-CRISPR-DECODE/Quarto-notebook-DECODE/data/DECODE_streamlines_control.pdf"</span>, dpi<span class="op">=</span><span class="dv">300</span>, bbox_inches<span class="op">=</span><span class="st">'tight'</span>)</span>
<span id="cb99-62"><a href="#cb99-62" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
 Applying min_shared_counts filter 

Filtered out 9834 genes that are detected 20 counts (shared).
WARNING: Did not normalize X as it looks processed already. To enforce normalization, set `enforce=True`.
Normalized count data: spliced, unspliced.
If you pass `n_top_genes`, all cutoffs are ignored.
extracting highly variable genes
    finished (0:00:00)
--&gt; added
    'highly_variable', boolean vector (adata.var)
    'means', float vector (adata.var)
    'dispersions', float vector (adata.var)
    'dispersions_norm', float vector (adata.var)
WARNING: adata.X seems to be already log-transformed.
'
 Computing PCA, knn and UMAP 
 
computing PCA
    on highly variable genes
    with n_comps=50
    finished (0:00:01)
computing neighbors
    finished: added to `.uns['neighbors']`
    `.obsp['distances']`, distances for each pair of neighbors
    `.obsp['connectivities']`, weighted adjacency matrix (0:00:01)
computing UMAP
    finished: added
    'X_umap', UMAP coordinates (adata.obsm) (0:00:08)

 Computing RNA velocity 
 
computing neighbors
    finished (0:00:01) --&gt; added 
    'distances' and 'connectivities', weighted adjacency matrices (adata.obsp)
computing moments based on connectivities
    finished (0:00:01) --&gt; added 
    'Ms' and 'Mu', moments of un/spliced abundances (adata.layers)

 Applying dynamical mode 

recovering dynamics (using 1/8 cores)
    finished (0:06:05) --&gt; added 
    'fit_pars', fitted parameters for splicing dynamics (adata.var)
computing velocities
    finished (0:00:02) --&gt; added 
    'velocity', velocity vectors for each individual cell (adata.layers)
computing velocity graph (using 1/8 cores)
    finished (0:00:05) --&gt; added 
    'velocity_graph', sparse matrix with cosine correlations (adata.uns)
computing velocity embedding
    finished (0:00:01) --&gt; added
    'velocity_diffmap', embedded velocity vectors (adata.obsm)</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="main_analyses_files/figure-html/cell-73-output-2.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="main_analyses_files/figure-html/cell-73-output-3.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="main_analyses_files/figure-html/cell-73-output-4.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="main_analyses_files/figure-html/cell-73-output-5.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="main_analyses_files/figure-html/cell-73-output-6.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<div id="cell-82" class="cell" data-execution_count="38">
<div class="sourceCode cell-code" id="cb101"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb101-1"><a href="#cb101-1" aria-hidden="true" tabindex="-1"></a>fig, ax <span class="op">=</span> plt.subplots(figsize<span class="op">=</span>(<span class="dv">8</span>, <span class="dv">6</span>))</span>
<span id="cb101-2"><a href="#cb101-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb101-3"><a href="#cb101-3" aria-hidden="true" tabindex="-1"></a><span class="co"># plot the streams of subsetted data on entire</span></span>
<span id="cb101-4"><a href="#cb101-4" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> scv.pl.scatter(adata_control_3p, basis<span class="op">=</span><span class="st">"diffmap"</span>, color<span class="op">=</span><span class="st">"high_res_annotation"</span>, legend_loc<span class="op">=</span><span class="st">"right_margin"</span>, legend_fontsize<span class="op">=</span><span class="dv">6</span>, components<span class="op">=</span><span class="st">"1,2"</span>, dpi<span class="op">=</span><span class="dv">300</span>, alpha<span class="op">=</span><span class="fl">0.2</span>, size<span class="op">=</span><span class="dv">35</span>, show<span class="op">=</span><span class="va">False</span>, ax<span class="op">=</span>ax)</span>
<span id="cb101-5"><a href="#cb101-5" aria-hidden="true" tabindex="-1"></a>scv.pl.velocity_embedding_grid(adata_control_3p, color<span class="op">=</span><span class="st">"high_res_annotation"</span>, basis<span class="op">=</span><span class="st">"diffmap"</span>,components<span class="op">=</span><span class="st">"1,2"</span>, alpha<span class="op">=</span><span class="dv">0</span>, ax<span class="op">=</span>ax, dpi<span class="op">=</span><span class="dv">300</span>, arrow_size<span class="op">=</span><span class="fl">2.5</span>, min_mass<span class="op">=</span><span class="dv">10</span>, title<span class="op">=</span><span class="st">"Control, all cell types"</span>)</span>
<span id="cb101-6"><a href="#cb101-6" aria-hidden="true" tabindex="-1"></a>fig.savefig(<span class="st">"/Users/sviatoslavkharuk/Documents/StudyBioinformatics/Bioinformatics_EMBL/Bioinformatics_project_EMBL/scRNA-seq-RNA-velocity-CRISPR-DECODE/Quarto-notebook-DECODE/figures/DECODE_streamlines_control_wrong.pdf"</span>, dpi<span class="op">=</span><span class="dv">300</span>, bbox_inches<span class="op">=</span><span class="st">'tight'</span>)</span>
<span id="cb101-7"><a href="#cb101-7" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="main_analyses_files/figure-html/cell-74-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<div id="cell-83" class="cell">
<div class="sourceCode cell-code" id="cb102"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb102-1"><a href="#cb102-1" aria-hidden="true" tabindex="-1"></a><span class="co">"""Notch KO"""</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="cell-84" class="cell" data-execution_count="39">
<div class="sourceCode cell-code" id="cb103"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb103-1"><a href="#cb103-1" aria-hidden="true" tabindex="-1"></a><span class="co">"""Fit with ISC instead of EB"""</span></span>
<span id="cb103-2"><a href="#cb103-2" aria-hidden="true" tabindex="-1"></a><span class="co"># now subset the cells for the fitting and for Notch condition</span></span>
<span id="cb103-3"><a href="#cb103-3" aria-hidden="true" tabindex="-1"></a>adata_notch_3p <span class="op">=</span> adata[((adata.obs[<span class="st">"perturbation"</span>] <span class="op">==</span> <span class="st">"Notch gRNA"</span>) <span class="op">&amp;</span></span>
<span id="cb103-4"><a href="#cb103-4" aria-hidden="true" tabindex="-1"></a>                                ((adata.obs[<span class="st">"sample_name"</span>] <span class="op">==</span> <span class="st">"TX22_D"</span>) <span class="op">|</span></span>
<span id="cb103-5"><a href="#cb103-5" aria-hidden="true" tabindex="-1"></a>                                (adata.obs[<span class="st">"sample_name"</span>] <span class="op">==</span> <span class="st">"TX23_D"</span>)))].copy()</span>
<span id="cb103-6"><a href="#cb103-6" aria-hidden="true" tabindex="-1"></a><span class="co"># add diffmap to adata object</span></span>
<span id="cb103-7"><a href="#cb103-7" aria-hidden="true" tabindex="-1"></a><span class="co"># diffmap_csv = pd.read_csv("/g/huber/users/kharuk/DECODE/adata_notch_control_3p_diffmap.csv", index_col=0)</span></span>
<span id="cb103-8"><a href="#cb103-8" aria-hidden="true" tabindex="-1"></a>diffmap_csv <span class="op">=</span> pd.read_csv(<span class="st">"/Users/sviatoslavkharuk/Documents/StudyBioinformatics/Bioinformatics_EMBL/Bioinformatics_project_EMBL/scRNA-seq-RNA-velocity-CRISPR-DECODE/Quarto-notebook-DECODE/data/adata_notch_control_3p_diffmap.csv"</span>, index_col<span class="op">=</span><span class="dv">0</span>)</span>
<span id="cb103-9"><a href="#cb103-9" aria-hidden="true" tabindex="-1"></a>adata_notch_3p.obsm[<span class="st">"X_diffmap"</span>] <span class="op">=</span> diffmap_csv.loc[adata_notch_3p.obs_names].values</span>
<span id="cb103-10"><a href="#cb103-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb103-11"><a href="#cb103-11" aria-hidden="true" tabindex="-1"></a><span class="co"># keep these cell types for inferring</span></span>
<span id="cb103-12"><a href="#cb103-12" aria-hidden="true" tabindex="-1"></a><span class="co"># velocity_inferring(adata_notch_3p_filtered, mode="dynamical", min_shared_counts=20, n_top_genes=500, annotation="high_res_annotation", groups_for_fit=None, condition="Notch")</span></span>
<span id="cb103-13"><a href="#cb103-13" aria-hidden="true" tabindex="-1"></a>velocity_inferring_local(adata_notch_3p, mode<span class="op">=</span><span class="st">"dynamical"</span>, min_shared_counts<span class="op">=</span><span class="dv">20</span>, n_top_genes<span class="op">=</span><span class="dv">500</span>, annotation<span class="op">=</span><span class="st">"high_res_annotation"</span>, groups_for_fit<span class="op">=</span><span class="va">None</span>, condition<span class="op">=</span><span class="st">"Notch"</span>)</span>
<span id="cb103-14"><a href="#cb103-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb103-15"><a href="#cb103-15" aria-hidden="true" tabindex="-1"></a><span class="co"># plot streams on the entire embedding</span></span>
<span id="cb103-16"><a href="#cb103-16" aria-hidden="true" tabindex="-1"></a>adata_notch_3p <span class="op">=</span> adata_notch_3p[<span class="op">~</span>adata_notch_3p.obs[<span class="st">"high_res_annotation"</span>].isin([<span class="st">"Malpighian tubule"</span>])].copy()</span>
<span id="cb103-17"><a href="#cb103-17" aria-hidden="true" tabindex="-1"></a>adata_notch_3p.uns[<span class="st">"high_res_annotation_colors"</span>] <span class="op">=</span> [<span class="st">'#1f77b4'</span>, <span class="st">'#ff7f0e'</span>, <span class="st">'#279e68'</span>, <span class="st">'#d62728'</span>, <span class="st">'#aa40fc'</span>, <span class="st">'#8c564b'</span>, <span class="st">'#b5bd61'</span>, <span class="st">'#17becf'</span>, <span class="st">'#ffbb78'</span>, <span class="st">'#98df8a'</span>, <span class="st">'#c5b0d5'</span>, <span class="st">'#c49c94'</span>, <span class="st">'#f7b6d2'</span>]</span>
<span id="cb103-18"><a href="#cb103-18" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> scv.pl.scatter(adata_notch_3p, basis<span class="op">=</span><span class="st">"diffmap"</span>, color<span class="op">=</span><span class="st">"high_res_annotation"</span>, legend_loc<span class="op">=</span><span class="st">"right_margin"</span>, legend_fontsize<span class="op">=</span><span class="dv">6</span>, components<span class="op">=</span><span class="st">"1,2"</span>, dpi<span class="op">=</span><span class="dv">300</span>, alpha<span class="op">=</span><span class="fl">0.2</span>, size<span class="op">=</span><span class="dv">35</span>, show<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb103-19"><a href="#cb103-19" aria-hidden="true" tabindex="-1"></a>scv.pl.velocity_embedding_grid(adata_notch_3p, color<span class="op">=</span><span class="st">"high_res_annotation"</span>, basis<span class="op">=</span><span class="st">"diffmap"</span>,components<span class="op">=</span><span class="st">"1,2"</span>, alpha<span class="op">=</span><span class="dv">0</span>, ax<span class="op">=</span>ax, dpi<span class="op">=</span><span class="dv">300</span>, arrow_size<span class="op">=</span><span class="fl">1.25</span>, min_mass<span class="op">=</span><span class="dv">10</span>, title<span class="op">=</span><span class="st">"emb_2_new, alpha=0.2, point_size=35, min_mass=10, arrow_size=1.25 </span><span class="ch">\n</span><span class="st"> Notch KO, moments default"</span>)</span>
<span id="cb103-20"><a href="#cb103-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb103-21"><a href="#cb103-21" aria-hidden="true" tabindex="-1"></a>fig, ax <span class="op">=</span> plt.subplots(figsize<span class="op">=</span>(<span class="dv">8</span>, <span class="dv">6</span>))</span>
<span id="cb103-22"><a href="#cb103-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb103-23"><a href="#cb103-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb103-24"><a href="#cb103-24" aria-hidden="true" tabindex="-1"></a><span class="co"># assign the colors to the adata_notch_3p</span></span>
<span id="cb103-25"><a href="#cb103-25" aria-hidden="true" tabindex="-1"></a>adata_notch_3p.uns[<span class="st">"high_res_annotation_colors"</span>] <span class="op">=</span> [<span class="st">'#1f77b4'</span>, <span class="st">'#ff7f0e'</span>, <span class="st">'#279e68'</span>, <span class="st">'#d62728'</span>, <span class="st">'#aa40fc'</span>, <span class="st">'#8c564b'</span>, <span class="st">'#b5bd61'</span>, <span class="st">'#17becf'</span>, <span class="st">'#ffbb78'</span>, <span class="st">'#98df8a'</span>, <span class="st">'#c5b0d5'</span>, <span class="st">'#c49c94'</span>, <span class="st">'#f7b6d2'</span>]</span>
<span id="cb103-26"><a href="#cb103-26" aria-hidden="true" tabindex="-1"></a>adata_notch_3p <span class="op">=</span> adata_notch_3p[<span class="op">~</span>adata_notch_3p.obs[<span class="st">"high_res_annotation"</span>].isin([<span class="st">"Malpighian tubule"</span>])].copy()</span>
<span id="cb103-27"><a href="#cb103-27" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> scv.pl.scatter(adata_notch_3p, basis<span class="op">=</span><span class="st">"diffmap"</span>, color<span class="op">=</span><span class="st">"high_res_annotation"</span>, legend_loc<span class="op">=</span><span class="st">"right_margin"</span>, legend_fontsize<span class="op">=</span><span class="dv">6</span>, components<span class="op">=</span><span class="st">"1,2"</span>, dpi<span class="op">=</span><span class="dv">300</span>, alpha<span class="op">=</span><span class="fl">0.2</span>, size<span class="op">=</span><span class="dv">35</span>, show<span class="op">=</span><span class="va">False</span>, ax<span class="op">=</span>ax)</span>
<span id="cb103-28"><a href="#cb103-28" aria-hidden="true" tabindex="-1"></a>scv.pl.velocity_embedding_grid(adata_notch_3p, color<span class="op">=</span><span class="st">"high_res_annotation"</span>, basis<span class="op">=</span><span class="st">"diffmap"</span>,components<span class="op">=</span><span class="st">"1,2"</span>, alpha<span class="op">=</span><span class="dv">0</span>, ax<span class="op">=</span>ax, dpi<span class="op">=</span><span class="dv">300</span>, arrow_size<span class="op">=</span><span class="fl">1.25</span>, min_mass<span class="op">=</span><span class="dv">10</span>, title<span class="op">=</span><span class="st">"Notch KO, all cell types"</span>)</span>
<span id="cb103-29"><a href="#cb103-29" aria-hidden="true" tabindex="-1"></a>fig.savefig(<span class="st">"/Users/sviatoslavkharuk/Documents/StudyBioinformatics/Bioinformatics_EMBL/Bioinformatics_project_EMBL/scRNA-seq-RNA-velocity-CRISPR-DECODE/Quarto-notebook-DECODE/data/DECODE_streamlines_notch.pdf"</span>, dpi<span class="op">=</span><span class="dv">300</span>, bbox_inches<span class="op">=</span><span class="st">'tight'</span>)</span>
<span id="cb103-30"><a href="#cb103-30" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
 Applying min_shared_counts filter 

Filtered out 9256 genes that are detected 20 counts (shared).
WARNING: Did not normalize X as it looks processed already. To enforce normalization, set `enforce=True`.
Normalized count data: spliced, unspliced.
If you pass `n_top_genes`, all cutoffs are ignored.
extracting highly variable genes
    finished (0:00:00)
--&gt; added
    'highly_variable', boolean vector (adata.var)
    'means', float vector (adata.var)
    'dispersions', float vector (adata.var)
    'dispersions_norm', float vector (adata.var)
WARNING: adata.X seems to be already log-transformed.
'
 Computing PCA, knn and UMAP 
 
computing PCA
    on highly variable genes
    with n_comps=50
    finished (0:00:00)
computing neighbors
    finished: added to `.uns['neighbors']`
    `.obsp['distances']`, distances for each pair of neighbors
    `.obsp['connectivities']`, weighted adjacency matrix (0:00:01)
computing UMAP
    finished: added
    'X_umap', UMAP coordinates (adata.obsm) (0:00:07)

 Computing RNA velocity 
 
computing neighbors
    finished (0:00:01) --&gt; added 
    'distances' and 'connectivities', weighted adjacency matrices (adata.obsp)
computing moments based on connectivities
    finished (0:00:02) --&gt; added 
    'Ms' and 'Mu', moments of un/spliced abundances (adata.layers)

 Applying dynamical mode 

recovering dynamics (using 1/8 cores)
    finished (0:07:45) --&gt; added 
    'fit_pars', fitted parameters for splicing dynamics (adata.var)
computing velocities
    finished (0:00:03) --&gt; added 
    'velocity', velocity vectors for each individual cell (adata.layers)
computing velocity graph (using 1/8 cores)
    finished (0:00:05) --&gt; added 
    'velocity_graph', sparse matrix with cosine correlations (adata.uns)
computing velocity embedding
    finished (0:00:01) --&gt; added
    'velocity_diffmap', embedded velocity vectors (adata.obsm)</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="main_analyses_files/figure-html/cell-76-output-2.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="main_analyses_files/figure-html/cell-76-output-3.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="main_analyses_files/figure-html/cell-76-output-4.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="main_analyses_files/figure-html/cell-76-output-5.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="main_analyses_files/figure-html/cell-76-output-6.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<div id="cell-85" class="cell" data-execution_count="40">
<div class="sourceCode cell-code" id="cb105"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb105-1"><a href="#cb105-1" aria-hidden="true" tabindex="-1"></a>fig, ax <span class="op">=</span> plt.subplots(figsize<span class="op">=</span>(<span class="dv">8</span>, <span class="dv">6</span>))</span>
<span id="cb105-2"><a href="#cb105-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb105-3"><a href="#cb105-3" aria-hidden="true" tabindex="-1"></a><span class="co"># assign the colors to the adata_notch_3p</span></span>
<span id="cb105-4"><a href="#cb105-4" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> scv.pl.scatter(adata_notch_3p, basis<span class="op">=</span><span class="st">"diffmap"</span>, color<span class="op">=</span><span class="st">"high_res_annotation"</span>, legend_loc<span class="op">=</span><span class="st">"right_margin"</span>, legend_fontsize<span class="op">=</span><span class="dv">6</span>, components<span class="op">=</span><span class="st">"1,2"</span>, dpi<span class="op">=</span><span class="dv">300</span>, alpha<span class="op">=</span><span class="fl">0.2</span>, size<span class="op">=</span><span class="dv">35</span>, show<span class="op">=</span><span class="va">False</span>, ax<span class="op">=</span>ax)</span>
<span id="cb105-5"><a href="#cb105-5" aria-hidden="true" tabindex="-1"></a>scv.pl.velocity_embedding_grid(adata_notch_3p, color<span class="op">=</span><span class="st">"high_res_annotation"</span>, basis<span class="op">=</span><span class="st">"diffmap"</span>,components<span class="op">=</span><span class="st">"1,2"</span>, alpha<span class="op">=</span><span class="dv">0</span>, ax<span class="op">=</span>ax, dpi<span class="op">=</span><span class="dv">300</span>, arrow_size<span class="op">=</span><span class="fl">2.5</span>, min_mass<span class="op">=</span><span class="dv">10</span>, title<span class="op">=</span><span class="st">"Notch KO, all cell types"</span>)</span>
<span id="cb105-6"><a href="#cb105-6" aria-hidden="true" tabindex="-1"></a>fig.savefig(<span class="st">"/Users/sviatoslavkharuk/Documents/StudyBioinformatics/Bioinformatics_EMBL/Bioinformatics_project_EMBL/scRNA-seq-RNA-velocity-CRISPR-DECODE/Quarto-notebook-DECODE/figures/DECODE_streamlines_notch_wrong.pdf"</span>, dpi<span class="op">=</span><span class="dv">300</span>, bbox_inches<span class="op">=</span><span class="st">'tight'</span>)</span>
<span id="cb105-7"><a href="#cb105-7" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="main_analyses_files/figure-html/cell-77-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="abundance-in-cell-types-is-different-in-conditions" class="level1">
<h1>3.8 Abundance in cell types is different in conditions</h1>
<div id="cell-87" class="cell" data-execution_count="41">
<div class="sourceCode cell-code" id="cb106"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb106-1"><a href="#cb106-1" aria-hidden="true" tabindex="-1"></a><span class="op">%</span>load_ext autoreload</span>
<span id="cb106-2"><a href="#cb106-2" aria-hidden="true" tabindex="-1"></a><span class="op">%</span>autoreload <span class="dv">2</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="cell-88" class="cell">
<div class="sourceCode cell-code" id="cb107"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb107-1"><a href="#cb107-1" aria-hidden="true" tabindex="-1"></a>pwd</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>'/Users/sviatoslavkharuk/Documents/StudyBioinformatics/Bioinformatics_EMBL/Bioinformatics_project_EMBL/scRNA-seq-RNA-velocity-CRISPR-DECODE/Quarto-notebook-DECODE/notebooks'</code></pre>
</div>
</div>
<div id="cell-89" class="cell" data-execution_count="42">
<div class="sourceCode cell-code" id="cb109"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb109-1"><a href="#cb109-1" aria-hidden="true" tabindex="-1"></a><span class="op">%</span>load_ext autoreload</span>
<span id="cb109-2"><a href="#cb109-2" aria-hidden="true" tabindex="-1"></a><span class="op">%</span>autoreload <span class="dv">2</span></span>
<span id="cb109-3"><a href="#cb109-3" aria-hidden="true" tabindex="-1"></a><span class="op">%</span>load_ext rpy2.ipython</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>The autoreload extension is already loaded. To reload it, use:
  %reload_ext autoreload</code></pre>
</div>
</div>
<div id="cell-90" class="cell" data-execution_count="43">
<div class="sourceCode cell-code" id="cb111"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb111-1"><a href="#cb111-1" aria-hidden="true" tabindex="-1"></a><span class="op">%%</span>R</span>
<span id="cb111-2"><a href="#cb111-2" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="op">!</span>require(<span class="st">"BiocManager"</span>, quietly <span class="op">=</span> TRUE))</span>
<span id="cb111-3"><a href="#cb111-3" aria-hidden="true" tabindex="-1"></a>    install.packages(<span class="st">"BiocManager"</span>)</span>
<span id="cb111-4"><a href="#cb111-4" aria-hidden="true" tabindex="-1"></a>BiocManager::install(<span class="st">"edgeR"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Update all/some/none? [a/s/n]: </code></pre>
</div>
<div class="cell-output cell-output-display">
<pre><code>Bioconductor version 3.19 (BiocManager 1.30.25), R 4.4.1 (2024-06-14)
Bioconductor version '3.19' is out-of-date; the current release version '3.20'
  is available with R version '4.4'; see https://bioconductor.org/install
Bioconductor version 3.19 (BiocManager 1.30.25), R 4.4.1 (2024-06-14)
Old packages: 'broom', 'ps', 'rstudioapi', 'FNN', 'MASS', 'Matrix', 'Rcpp',
  'RcppArmadillo', 'abind', 'askpass', 'bit', 'bit64', 'bitops', 'boot',
  'commonmark', 'curl', 'data.table', 'dotCall64', 'evaluate', 'fontawesome',
  'foreign', 'fs', 'future.apply', 'glue', 'gplots', 'gtable', 'igraph',
  'jsonlite', 'knitr', 'later', 'nlme', 'openssl', 'parallelly', 'patchwork',
  'progressr', 'promises', 'reticulate', 'rmarkdown', 'spam', 'spatstat.data',
  'spatstat.explore', 'spatstat.geom', 'spatstat.random', 'spatstat.univar',
  'spatstat.utils', 'survival', 'sys', 'tinytex', 'withr', 'xfun'
In addition: Warning message:
package(s) not installed when version(s) same as or greater than current; use
  `force = TRUE` to re-install: 'edgeR' </code></pre>
</div>
</div>
<div id="cell-91" class="cell" data-execution_count="44">
<div class="sourceCode cell-code" id="cb114"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb114-1"><a href="#cb114-1" aria-hidden="true" tabindex="-1"></a><span class="op">%%</span>R</span>
<span id="cb114-2"><a href="#cb114-2" aria-hidden="true" tabindex="-1"></a>BiocManager::install(<span class="st">"statmod"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Update all/some/none? [a/s/n]: </code></pre>
</div>
<div class="cell-output cell-output-display">
<pre><code>Bioconductor version 3.19 (BiocManager 1.30.25), R 4.4.1 (2024-06-14)
Old packages: 'broom', 'ps', 'rstudioapi', 'FNN', 'MASS', 'Matrix', 'Rcpp',
  'RcppArmadillo', 'abind', 'askpass', 'bit', 'bit64', 'bitops', 'boot',
  'commonmark', 'curl', 'data.table', 'dotCall64', 'evaluate', 'fontawesome',
  'foreign', 'fs', 'future.apply', 'glue', 'gplots', 'gtable', 'igraph',
  'jsonlite', 'knitr', 'later', 'nlme', 'openssl', 'parallelly', 'patchwork',
  'progressr', 'promises', 'reticulate', 'rmarkdown', 'spam', 'spatstat.data',
  'spatstat.explore', 'spatstat.geom', 'spatstat.random', 'spatstat.univar',
  'spatstat.utils', 'survival', 'sys', 'tinytex', 'withr', 'xfun'
In addition: Warning message:
package(s) not installed when version(s) same as or greater than current; use
  `force = TRUE` to re-install: 'statmod' </code></pre>
</div>
</div>
<div id="cell-92" class="cell" data-execution_count="45">
<div class="sourceCode cell-code" id="cb117"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb117-1"><a href="#cb117-1" aria-hidden="true" tabindex="-1"></a><span class="op">%%</span>R</span>
<span id="cb117-2"><a href="#cb117-2" aria-hidden="true" tabindex="-1"></a>library(edgeR)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>Loading required package: limma</code></pre>
</div>
</div>
<div id="cell-93" class="cell" data-execution_count="51">
<div class="sourceCode cell-code" id="cb119"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb119-1"><a href="#cb119-1" aria-hidden="true" tabindex="-1"></a><span class="co"># upload the data</span></span>
<span id="cb119-2"><a href="#cb119-2" aria-hidden="true" tabindex="-1"></a><span class="co">#adata_notch_control_3p_DE = sc.read_h5ad("../data/adata_notch_control_3p_diffmap_local_machine.h5ad")</span></span>
<span id="cb119-3"><a href="#cb119-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-4"><a href="#cb119-4" aria-hidden="true" tabindex="-1"></a><span class="co"># upload the data from the cluster on local machine</span></span>
<span id="cb119-5"><a href="#cb119-5" aria-hidden="true" tabindex="-1"></a>adata_notch_control_3p_DE <span class="op">=</span> sc.read_h5ad(<span class="st">"../data/a_v_data_merged_DECODE_v1_Notch_Control_diffmap.h5ad"</span>)</span>
<span id="cb119-6"><a href="#cb119-6" aria-hidden="true" tabindex="-1"></a><span class="co"># rvelo-milo env</span></span>
<span id="cb119-7"><a href="#cb119-7" aria-hidden="true" tabindex="-1"></a>adata_notch_control_3p_DE, milo, mdata <span class="op">=</span> differential_abundance(adata_notch_control_3p_DE)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>computing PCA
    on highly variable genes
    with n_comps=50
    finished (0:00:06)
computing neighbors
    finished: added to `.uns['neighbors_harmony_batch']`
    `.obsp['neighbors_harmony_batch_distances']`, distances for each pair of neighbors
    `.obsp['neighbors_harmony_batch_connectivities']`, weighted adjacency matrix (0:00:01)
computing UMAP
    finished: added
    'X_umap', UMAP coordinates (adata.obsm) (0:00:15)
computing neighbors
    finished: added to `.uns['neighbors']`
    `.obsp['distances']`, distances for each pair of neighbors
    `.obsp['connectivities']`, weighted adjacency matrix (0:00:03)</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>2024-12-01 15:43:47,082 - harmonypy - INFO - Computing initial centroids with sklearn.KMeans...
2024-12-01 15:43:59,445 - harmonypy - INFO - sklearn.KMeans initialization complete.
2024-12-01 15:43:59,599 - harmonypy - INFO - Iteration 1 of 50
2024-12-01 15:44:09,475 - harmonypy - INFO - Iteration 2 of 50
2024-12-01 15:44:16,780 - harmonypy - INFO - Iteration 3 of 50
2024-12-01 15:44:25,739 - harmonypy - INFO - Iteration 4 of 50
2024-12-01 15:44:33,771 - harmonypy - INFO - Iteration 5 of 50
2024-12-01 15:44:42,649 - harmonypy - INFO - Iteration 6 of 50
2024-12-01 15:44:50,353 - harmonypy - INFO - Iteration 7 of 50
2024-12-01 15:44:59,508 - harmonypy - INFO - Iteration 8 of 50
2024-12-01 15:45:05,664 - harmonypy - INFO - Iteration 9 of 50
2024-12-01 15:45:09,108 - harmonypy - INFO - Converged after 9 iterations</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="main_analyses_files/figure-html/cell-84-output-3.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="main_analyses_files/figure-html/cell-84-output-4.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="main_analyses_files/figure-html/cell-84-output-5.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="main_analyses_files/figure-html/cell-84-output-6.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="main_analyses_files/figure-html/cell-84-output-7.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="main_analyses_files/figure-html/cell-84-output-8.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="main_analyses_files/figure-html/cell-84-output-9.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<div id="cell-94" class="cell" data-execution_count="52">
<div class="sourceCode cell-code" id="cb122"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb122-1"><a href="#cb122-1" aria-hidden="true" tabindex="-1"></a><span class="co">### Visualize results on embedding</span></span>
<span id="cb122-2"><a href="#cb122-2" aria-hidden="true" tabindex="-1"></a>milo.build_nhood_graph(mdata)</span>
<span id="cb122-3"><a href="#cb122-3" aria-hidden="true" tabindex="-1"></a>plt.rcParams[<span class="st">"figure.figsize"</span>] <span class="op">=</span> [<span class="fl">7.5</span>, <span class="dv">7</span>]</span>
<span id="cb122-4"><a href="#cb122-4" aria-hidden="true" tabindex="-1"></a>milo.plot_nhood_graph(</span>
<span id="cb122-5"><a href="#cb122-5" aria-hidden="true" tabindex="-1"></a>    mdata,</span>
<span id="cb122-6"><a href="#cb122-6" aria-hidden="true" tabindex="-1"></a>    alpha<span class="op">=</span><span class="fl">0.1</span>,  <span class="co">## SpatialFDR level (10%)</span></span>
<span id="cb122-7"><a href="#cb122-7" aria-hidden="true" tabindex="-1"></a>    min_size<span class="op">=</span><span class="dv">1</span>,  <span class="co">## Size of smallest dot</span></span>
<span id="cb122-8"><a href="#cb122-8" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="main_analyses_files/figure-html/cell-85-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>